{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e41cdaa",
   "metadata": {},
   "source": [
    "### LLM interaction paper\n",
    "Running fulltexts through chatgpt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "483a400e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b93c1b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Parv\\Doc\\Nexus_project\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Parv\\anaconda3\\envs\\pytorch_gpu\\Lib\\site-packages\\IPython\\core\\magics\\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "%cd c:\\\\Users\\\\Parv\\\\Doc\\\\Nexus_project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebda6f34",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:1: SyntaxWarning: invalid escape sequence '\\o'\n",
      "<>:1: SyntaxWarning: invalid escape sequence '\\o'\n",
      "C:\\Users\\Parv\\AppData\\Local\\Temp\\ipykernel_31412\\1461466291.py:1: SyntaxWarning: invalid escape sequence '\\o'\n",
      "  with open('backend\\openai_api_key.txt', 'r') as f:\n"
     ]
    }
   ],
   "source": [
    "with open('backend\\openai_api_key.txt', 'r') as f:\n",
    "    openai_key = f.readlines()[0]\n",
    "\n",
    "os.environ['OPENAI_API_KEY'] = openai_key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "93982066",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6a0795d7b4c54b4cab33a819b171a0ac",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading https://raw.githubusercontent.com/stanfordnlp/stanza-resources/main/resources_1.10.0.json:   0%|  …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-05 16:28:26 INFO: Downloaded file to C:\\Users\\Parv\\stanza_resources\\resources.json\n",
      "2025-09-05 16:28:26 WARNING: Language en package default expects mwt, which has been added\n",
      "2025-09-05 16:28:26 INFO: Downloading these customized packages for language: en (English)...\n",
      "=======================================\n",
      "| Processor       | Package           |\n",
      "---------------------------------------\n",
      "| tokenize        | combined          |\n",
      "| mwt             | combined          |\n",
      "| pos             | combined_charlm   |\n",
      "| lemma           | combined_nocharlm |\n",
      "| depparse        | combined_charlm   |\n",
      "| pretrain        | conll17           |\n",
      "| backward_charlm | 1billion          |\n",
      "| forward_charlm  | 1billion          |\n",
      "=======================================\n",
      "\n",
      "2025-09-05 16:28:26 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\tokenize\\combined.pt\n",
      "2025-09-05 16:28:26 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\mwt\\combined.pt\n",
      "2025-09-05 16:28:26 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\pos\\combined_charlm.pt\n",
      "2025-09-05 16:28:26 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\lemma\\combined_nocharlm.pt\n",
      "2025-09-05 16:28:26 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\depparse\\combined_charlm.pt\n",
      "2025-09-05 16:28:26 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\pretrain\\conll17.pt\n",
      "2025-09-05 16:28:27 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\backward_charlm\\1billion.pt\n",
      "2025-09-05 16:28:27 INFO: File exists: C:\\Users\\Parv\\stanza_resources\\en\\forward_charlm\\1billion.pt\n",
      "2025-09-05 16:28:27 INFO: Finished downloading models and saved to C:\\Users\\Parv\\stanza_resources\n",
      "2025-09-05 16:28:27 INFO: Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1648097c02534c4bb9177b7d52acb60d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading https://raw.githubusercontent.com/stanfordnlp/stanza-resources/main/resources_1.10.0.json:   0%|  …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-05 16:28:27 INFO: Downloaded file to C:\\Users\\Parv\\stanza_resources\\resources.json\n",
      "2025-09-05 16:28:27 WARNING: Language en package default expects mwt, which has been added\n",
      "2025-09-05 16:28:27 INFO: Loading these models for language: en (English):\n",
      "=================================\n",
      "| Processor | Package           |\n",
      "---------------------------------\n",
      "| tokenize  | combined          |\n",
      "| mwt       | combined          |\n",
      "| pos       | combined_charlm   |\n",
      "| lemma     | combined_nocharlm |\n",
      "| depparse  | combined_charlm   |\n",
      "=================================\n",
      "\n",
      "2025-09-05 16:28:27 INFO: Using device: cuda\n",
      "2025-09-05 16:28:27 INFO: Loading: tokenize\n",
      "2025-09-05 16:28:28 INFO: Loading: mwt\n",
      "2025-09-05 16:28:28 INFO: Loading: pos\n",
      "2025-09-05 16:28:31 INFO: Loading: lemma\n",
      "2025-09-05 16:28:32 INFO: Loading: depparse\n",
      "2025-09-05 16:28:33 INFO: Done loading processors!\n"
     ]
    }
   ],
   "source": [
    "from backend.backend_function import search_for_papers, retrieve_paper, bulk_retrieve_papers, get_paper_pdf_urls\n",
    "from backend.text_extraction import extract_pdf_text_from_url"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de564b09",
   "metadata": {},
   "source": [
    "### Getting test paper case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5d666eca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Retrieving bulk papers returned status 429 for 20 papers\n",
      "ERROR:root:Retrieving bulk papers returned status 429 for 20 papers\n",
      "ERROR:root:Retrieving bulk papers returned status 429 for 20 papers\n",
      "ERROR:root:Retrieving bulk papers returned status 429 for 20 papers\n"
     ]
    }
   ],
   "source": [
    "paper = search_for_papers('Motif caller')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "99e550a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "paper = retrieve_paper(paper[0]['paperId'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "816defa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "refs = bulk_retrieve_papers([i['paperId'] for i in paper['references']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "23a3a155",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = [paper] + refs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "35e5f0be",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Error in extracting pdf url for {'paperId': '8289cb7b8806e4d4b345dc8fd02ff1f52fc4047f', 'externalIds': {'MAG': '1471723175', 'DBLP': 'conf/osdi/BalakrishnanBDEGHLOPR14', 'CorpusId': 417688}, 'corpusId': 417688, 'title': 'Pelican: A Building Block for Exascale Cold Data Storage', 'year': 2014, 'referenceCount': 19, 'citationCount': 82, 'influentialCitationCount': 14, 'openAccessPdf': {'url': '', 'status': None, 'license': None}, 'fieldsOfStudy': ['Computer Science'], 's2FieldsOfStudy': [{'category': 'Computer Science', 'source': 'external'}, {'category': 'Computer Science', 'source': 's2-fos-model'}, {'category': 'Engineering', 'source': 's2-fos-model'}], 'tldr': {'model': 'tldr@v2.0.0', 'text': 'Pelican performs well for cold workloads, providing high throughput with acceptable latency, and is compared against a traditional resource overprovisioned storage rack using a cross-validated simulator.'}, 'journal': {'pages': '351-365'}, 'authors': [{'authorId': '3074541', 'name': 'S. Balakrishnan'}, {'authorId': '144247681', 'name': 'Richard Black'}, {'authorId': '2793107', 'name': 'Austin Donnelly'}, {'authorId': '144304781', 'name': 'P. England'}, {'authorId': '2191057010', 'name': 'Adam Glass'}, {'authorId': '144218922', 'name': 'D. Harper'}, {'authorId': '3229837', 'name': 'Sergey Legtchenko'}, {'authorId': '34447742', 'name': 'Aaron Ogus'}, {'authorId': '2059365345', 'name': 'Eric Peterson'}, {'authorId': '1710239', 'name': 'A. Rowstron'}], 'references': None}\n",
      "ERROR:root:Error in extracting pdf url for None\n",
      "ERROR:root:Error in extracting pdf url for {'paperId': '11bcd76b45c8ffd2e93c15f7ef39377435fd5a60', 'externalIds': {'CorpusId': 8138244}, 'corpusId': 8138244, 'title': 'Cold Storage in the Cloud : Trends , Challenges , and Solutions', 'year': 2013, 'referenceCount': 4, 'citationCount': 17, 'influentialCitationCount': 1, 'openAccessPdf': {'url': '', 'status': None, 'license': None}, 'fieldsOfStudy': None, 's2FieldsOfStudy': [{'category': 'Computer Science', 'source': 's2-fos-model'}, {'category': 'Engineering', 'source': 's2-fos-model'}], 'tldr': {'model': 'tldr@v2.0.0', 'text': 'Compatibility with existing enterprise architecture is important for seamless access to data as well as to simplify manageability, which helps reduce total cost of ownership (TCO).'}, 'journal': None, 'authors': [{'authorId': '2053146712', 'name': 'Adam Mendoza'}], 'references': None}\n",
      "ERROR:root:Error in extracting pdf url for None\n",
      "ERROR:root:Error in extracting pdf url for None\n"
     ]
    }
   ],
   "source": [
    "pdf_urls = get_paper_pdf_urls([paper] + refs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670213ee",
   "metadata": {},
   "source": [
    "### Paper segmentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c1140704",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def segment_paper(text):\n",
    "    headers = [\"abstract\", \"introduction\", \"methods?\", \"results?\", \"discussion\", \n",
    "               \"conclusion\", \"references?\"]\n",
    "    # Put (?i) ONCE at the start for case-insensitive matching\n",
    "    pattern = r\"(?i)\\b(\" + \"|\".join(headers) + r\")\\b\"\n",
    "    sections = re.split(pattern, text)\n",
    "    return [s.strip() for s in sections if s.strip()]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8480def4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pdf extraction error HTTPSConnectionPool(host='advanced.onlinelibrary.wiley.com', port=443): Max retries exceeded with url: /doi/10.1002/advs.202411354 (Caused by SSLError(SSLEOFError(8, '[SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1000)')))\n"
     ]
    }
   ],
   "source": [
    "fulltexts = [extract_pdf_text_from_url(i) if i else None for i in pdf_urls]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "26d20a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "import regex as re\n",
    "import requests\n",
    "import fitz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9130e261",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_url = pdf_urls[5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a61f5c75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.nature.com/articles/s41598-024-58386-z.pdf'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdf_url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f26fd28f",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get(pdf_url)\n",
    "with open(\"temp.pdf\", \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "doc = fitz.open(\"temp.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e8e7f62b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'format': 'PDF 1.4',\n",
       " 'title': 'Efficient DNA-based data storage using shortmer combinatorial encoding',\n",
       " 'author': 'Inbal Preuss ',\n",
       " 'subject': 'Scientific Reports, https://doi.org/10.1038/s41598-024-58386-z',\n",
       " 'keywords': '',\n",
       " 'creator': 'Springer',\n",
       " 'producer': 'Adobe PDF Library 15.0; modified using iText® 5.3.5 ©2000-2012 1T3XT BVBA (SPRINGER SBM; licensed version)',\n",
       " 'creationDate': \"D:20240402072217+05'30'\",\n",
       " 'modDate': \"D:20240902151110+02'00'\",\n",
       " 'trapped': '',\n",
       " 'encryption': None}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doc.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d77f85a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def segment_paper(doc, toc):\n",
    "\n",
    "    def define_match_locations_per_page(match_locations):\n",
    "        \"\"\"\n",
    "        Based on matched section headings in the text (page numbers..), get the best estimate on where\n",
    "        the text begins and ends\n",
    "        \"\"\"\n",
    "\n",
    "        match_locs = []  # Final list\n",
    "\n",
    "        for ptr, loc in enumerate(match_locations):\n",
    "                \n",
    "            if len(loc) >= 1:\n",
    "                if len(loc) < 2:\n",
    "                    best_match_loc = loc[0]\n",
    "\n",
    "                else:\n",
    "                    if ptr == 0: # First section on the page\n",
    "                        best_match_loc = loc[0]\n",
    "                    else:\n",
    "                        # Select the last match as long as it is before the next section found\n",
    "                        best_match_loc = len(loc) - 1\n",
    "                        while True:\n",
    "                            if ptr < len(match_locations) - 2:\n",
    "                                next_match = match_locations[ptr + 1]\n",
    "                                if len(next_match) > 0:\n",
    "                                    if next_match[0] < loc[best_match_loc]:\n",
    "                                        best_match_loc -= 1\n",
    "                                    else:\n",
    "                                        break\n",
    "                                else:\n",
    "                                    ptr += 1\n",
    "                                    continue\n",
    "                            else:\n",
    "                                break\n",
    "            else:\n",
    "                best_match_loc = -1\n",
    "            \n",
    "            match_locs.append(best_match_loc)\n",
    "\n",
    "        return match_locs\n",
    "\n",
    "    per_page_dicts = []\n",
    "\n",
    "    for page_num, page in enumerate(doc):\n",
    "\n",
    "        text = page.get_text('text')\n",
    "\n",
    "        section_headings_in_page = [\n",
    "            i[1] for i in toc if i[2] == page_num + 1]  # Getting section headings from toc\n",
    "\n",
    "        match_locations = []\n",
    "        for ptr, section_heading in enumerate(section_headings_in_page):  # Getting all match locations\n",
    "            text_matches = [\n",
    "                    m.start() for m in re.finditer(section_heading, text, re.IGNORECASE)\n",
    "                ]\n",
    "            match_locations.append(text_matches)\n",
    "        per_page_dict = {}\n",
    "        final_match_locations = define_match_locations_per_page(match_locations)  # Final match locations for all section headings found\n",
    "        for i in range(len(section_headings_in_page)):\n",
    "            heading = section_headings_in_page[i]\n",
    "            if i == 0:\n",
    "                per_page_dict['prev'] = text[:final_match_locations[0]]\n",
    "            if i == len(section_headings_in_page) - 1:\n",
    "                per_page_dict[heading] = text[final_match_locations[i]:]\n",
    "            else:\n",
    "                per_page_dict[heading] = text[final_match_locations[i]: final_match_locations[i+1]]\n",
    "\n",
    "        per_page_dicts.append(per_page_dict)\n",
    "        \n",
    "    segmented_paper = {}\n",
    "    last_section = \"\"\n",
    "    held_key = \"\"\n",
    "    for i, j in enumerate(per_page_dicts):\n",
    "        for key, value in j.items():\n",
    "            if key == 'prev' and i > 0:\n",
    "                segmented_paper[last_section] += value\n",
    "            else:\n",
    "                segmented_paper[key] = value\n",
    "\n",
    "            held_key = key\n",
    "        last_section = held_key \n",
    "\n",
    "    return segmented_paper\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e2b82f4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▌         | 1/20 [00:00<00:02,  7.82it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 'Introduction', 1], [1, 'Motif-Based Synthesis', 2], [1, 'Basecalling & Motif Search - The Baseline', 3], [2, 'Zero-error (ZE) Search', 3], [2, 'Approximate-matching (AM) Search', 3], [1, 'Motif Caller', 4], [2, 'Training Procedure', 4], [2, 'Decoding', 5], [1, 'Dataset', 5], [1, 'Results', 6], [2, 'Motif accuracies per read', 6], [2, 'Information recovered per block with increasing reads', 7], [2, 'Decoding percentage for diluted reads', 7], [2, 'Generalization of performance to alternative sequencing runs', 8], [2, 'Finetuning and Quality Proportions', 8], [1, 'Discussion', 9], [1, 'References', 9]]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 15%|█▌        | 3/20 [00:00<00:03,  4.86it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "[[1, 'Introduction', 1], [1, 'Motif-Based DNA Storage', 2], [2, 'General Principle', 2], [2, 'Experimental Data Set', 2], [2, 'Channel Simulator', 3], [1, 'Coupon Collector Channels', 3], [2, 'Capacity of the CC(n,k,R) Channel', 4], [2, 'Erasure Version of the Coupon Collector Channel', 5], [2, 'Coupon Collector Channel with Interference', 6], [2, 'Read-Write Cost Trade-Off', 7], [1, 'Non-Binary LDPC Codes Over the Coupon Collector Channels', 8], [2, 'No Interference', 8], [2, 'Interference', 8], [2, 'Protograph-Based Non-Binary SC-LDPC Codes', 9], [2, 'Combinatorial Explosion in Decoding Complexity', 9], [1, 'Numerical Results', 10], [1, 'Conclusion', 10], [1, 'References', 11]]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 20%|██        | 4/20 [00:00<00:03,  5.27it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 'Introduction', 1], [1, 'Background', 2], [2, 'Errors due to Consensus Calling', 2], [2, 'Errors due to Random Access', 3], [1, 'Design', 4], [2, 'Write Pipeline', 5], [2, 'Read Pipeline', 6], [1, 'Evaluation', 8], [2, 'Small-Scale Wet-Lab Validation', 9], [3, 'Error Pattern', 9], [3, 'Data Recovery', 9], [2, 'Large-Scale Wet-lab Validation', 9], [3, 'Error Pattern', 10], [3, 'Coverage Bias', 10], [3, 'Data Recover per Extent', 10], [2, 'SOTA Comparison', 10], [2, 'Benefits of Column-Based Design', 11], [2, 'Simulation of Large Scale Random Access', 12], [1, 'Conclusion', 13], [1, 'References', 13]]\n",
      "[]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 30%|███       | 6/20 [00:09<00:34,  2.50s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 'Efficient DNA-based data storage using shortmer combinatorial encoding', 1], [2, 'Results', 2], [3, 'Design of shortmer combinatorial encoding for DNA storage', 2], [3, 'Binary and binomial combinatorial alphabets', 3], [3, 'Reconstruction probabilities for binomial encoding', 4], [4, 'Reconstruction of a single combinatorial letter', 4], [4, 'Reconstruction of a combinatorial sequence', 4], [3, 'An end-to-end combinatorial shortmer storage system', 4], [3, 'Experimental proof of concept', 7], [2, 'Discussion', 9], [2, 'Methods', 11], [3, 'Reconstruction probability of a binomial encoding letter', 11], [3, 'An end-to-end combinatorial storage system', 12], [4, 'Combinatorial encoding and padding', 12], [4, 'Error correction codes', 12], [4, 'Synthesis and sequencing simulation with errors', 13], [4, 'Reconstruction', 14], [4, 'Cost analysis', 14], [4, 'Proof of concept experiment', 14], [4, 'DNA assembly and sequencing', 14], [4, 'Decoding and analysis', 15], [3, 'Information capacities for selected encodings', 15], [3, 'Ethics declaration', 16], [2, 'References', 16], [2, 'Acknowledgements', 17]]\n",
      "pdf extraction error\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 40%|████      | 8/20 [00:11<00:20,  1.72s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 'Abstract', 1], [2, 'Background', 1], [2, 'Results', 1], [2, 'Conclusions', 1], [2, 'Keywords', 1], [1, 'Background', 1], [1, 'Results and discussion', 2], [2, 'Default model performance', 2], [2, 'Custom model performance', 4], [2, 'Consensus error profiles', 4], [2, 'Nanopolish performance', 4], [1, 'Conclusions', 6], [1, 'Methods', 7], [2, 'Custom model training', 7], [2, 'Read sets', 7], [2, 'Read accuracy', 8], [2, 'Consensus accuracy', 9], [2, 'Polishing', 9], [1, 'Additional files', 9], [2, 'Additional file 1', 9], [2, 'Additional file 2', 9], [2, 'Additional file 3', 9], [1, 'Acknowledgements', 9], [1, 'Review history', 9], [1, \"Authors' contributions\", 9], [1, 'Funding', 10], [1, 'Availability of data and materials', 10], [1, 'Ethics approval and consent to participate', 10], [1, 'Consent for publication', 10], [1, 'Competing interests', 10], [1, 'Author details', 10], [1, 'References', 10], [1, \"Publisher's Note\", 10]]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 55%|█████▌    | 11/20 [00:13<00:10,  1.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 'Abstract', 1], [2, 'Background', 1], [2, 'Results', 1], [2, 'Conclusions', 1], [2, 'Keywords', 1], [1, 'Background', 1], [1, 'Methods', 2], [2, 'The semi-global DP algorithm', 2], [2, 'Naïve difference recurrence relations', 3], [2, 'The proposed algorithm', 4], [2, 'Relation to other DP algorithms', 5], [2, 'Library implementation', 5], [3, 'Finding the cell with the maximum value', 5], [3, 'Traceback flags and generating alignment path strings', 7], [3, 'Computation of score profile vectors and prefetching sequences', 8], [3, 'Miscellaneous implementation techniques', 8], [4, 'Register usage:', 8], [4, 'Branch predictor consideration:', 9], [1, 'Results', 9], [2, 'Comparison to existing implementations', 11], [1, 'Discussion', 13], [1, 'Conclusions', 14], [1, 'Additional files', 14], [2, 'Additional file 1', 14], [2, 'Additional file 2', 14], [1, 'Abbreviations', 14], [1, 'Acknowledgments', 14], [1, 'Funding', 14], [1, 'Availability of data and materials', 14], [1, 'About this supplement', 14], [1, \"Authors' contributions\", 14], [1, 'Ethics approval and consent to participate', 14], [1, 'Consent for publication', 14], [1, 'Competing interests', 14], [1, \"Publisher's Note\", 14], [1, 'References', 14]]\n",
      "[]\n",
      "pdf extraction error\n",
      "pdf extraction error\n",
      "pdf extraction error\n",
      "pdf extraction error\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 75%|███████▌  | 15/20 [00:15<00:03,  1.37it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:16<00:00,  1.23it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 'Not Found', 1]]\n",
      "pdf extraction error\n",
      "pdf extraction error\n",
      "pdf extraction error\n",
      "pdf extraction error\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "segmented_papers = []\n",
    "for pdf_url in tqdm(pdf_urls):\n",
    "    try:\n",
    "        response = requests.get(pdf_url)\n",
    "        with open(\"temp.pdf\", \"wb\") as f:\n",
    "            f.write(response.content)\n",
    "        doc = fitz.open(\"temp.pdf\")\n",
    "        toc = doc.get_toc(simple=True)\n",
    "        print(toc)\n",
    "    except Exception as e:\n",
    "        print(\"pdf extraction error\")\n",
    "    seg_paper = segment_paper(doc, toc)\n",
    "    segmented_papers.append(seg_paper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1f695dd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "segmented_papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "28aee847",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'prev': '1\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  \\n| https://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports\\nEfficient DNA‑based data storage \\nusing shortmer combinatorial \\nencoding\\nInbal\\xa0Preuss\\u200a1,3*, Michael\\xa0Rosenberg\\u200a2, Zohar\\xa0Yakhini\\u200a1,3\\xa0& Leon\\xa0Anavy\\u200a1,3\\nData storage in DNA has recently emerged as a promising archival solution,\\xa0offering space-efficient \\nand long-lasting digital storage solutions. Recent studies suggest leveraging the inherent redundancy \\nof synthesis and sequencing technologies by using composite DNA alphabets. A major challenge \\nof this approach involves the noisy inference process, obstructing\\xa0large composite alphabets. This \\npaper introduces a novel approach for DNA-based data storage, offering, in some implementations, \\na 6.5-fold increase in logical density over standard DNA-based storage systems, with near-zero \\nreconstruction error. Combinatorial DNA encoding uses a set of clearly distinguishable DNA shortmers \\nto construct large combinatorial alphabets, where each letter consists of a subset of shortmers. We \\nformally define various combinatorial encoding schemes and investigate their theoretical properties. \\nThese include information density and reconstruction probabilities, as well as required synthesis \\nand sequencing multiplicities. We then propose an end-to-end design for a combinatorial DNA-\\nbased data storage system, including encoding schemes, two-dimensional (2D) error correction \\ncodes, and reconstruction algorithms, under different error regimes. We performed simulations and \\nshow, for example, that\\xa0the use of 2D Reed-Solomon error correction has significantly improved \\nreconstruction rates. We validated\\xa0our approach by constructing two combinatorial sequences using \\nGibson assembly, imitating a 4-cycle combinatorial synthesis process. We confirmed the successful \\nreconstruction, and established the robustness of our approach for different error types. Subsampling \\nexperiments supported the important role of sampling rate and its effect on the overall performance. \\nOur work demonstrates the potential of combinatorial shortmer encoding for DNA-based data storage \\nand describes some theoretical research questions and technical challenges. Combining combinatorial \\nprinciples with error-correcting strategies, and investing in the development of DNA synthesis \\ntechnologies that efficiently support combinatorial synthesis, can pave the way to efficient, error-\\nresilient DNA-based storage solutions.\\nDNA is a promising media storage candidate for long-term data archiving, due to its high information density, \\nlong-term stability, and robustness. In recent years, several studies have demonstrated the use of synthetic DNA \\nfor storing digital information on a megabyte scale, exceeding the physical density of current magnetic tape-\\nbased systems by roughly six orders of \\xadmagnitude1,2. Physical density is one of several quantitative metrics for \\nevaluating the efficiency of DNA-based storage systems, measured by the data unit per gram of DNA. Another \\nperformance metric, which was introduced \\xadin3, is called logical density, refering to the amount of data encoded in \\neach synthesis cycle. Since DNA synthesis is the main cost component in DNA-based storage systems, increasing \\nthe logical density is the main focus of this work.\\nResearch efforts in the field of DNA-based storage systems have mainly focused on the application of various \\nencoding schemes, while relying on standard DNA synthesis and sequencing technologies. These include the \\ndevelopment of error-correcting codes for the unique information channel of DNA-based data \\xadstorage4–8. Random \\naccess capabilities for reading specific information stored in DNA also require advanced coding \\xadschemes9–11. Yet, \\ndespite the enormous benefits potentially associated with capacity, robustness, and size, existing DNA-based \\nstorage technologies are characterized by inherent information redundancy. This is due to the nature of DNA \\nsynthesis and sequencing methodologies, which process multiple molecules that represent the same information \\nbits in parallel. Recent studies suggest exploiting this redundancy to increase the logical density of the system, by \\nextending the standard DNA alphabet using composite letters (also referred to as degenerate bases), and thereby \\nOPEN\\n1School of Computer Science, Reichman University, 4610101\\xa0Herzliya, Israel. 2Institute of Nanotechnology and \\nAdvanced Materials, The Mina and Everard Goodman Faculty of Life Sciences, Bar-Ilan University, 5290002\\xa0Ramat \\nGan, Israel. 3Faculty of Computer Science, Technion, 3200003\\xa0Haifa, Israel. *email: inbalpreuss@gmail.com',\n",
       " 'Efficient DNA-based data storage using shortmer combinatorial encoding': '\\n2\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\nencoding more than 2 bits per \\xadletter12,13. In this approach, a composite DNA letter uses all four DNA bases (A, C, \\nG, and T), combined or mixed in a specified predetermined ratio σ = (σA, σC, σG, σT)\\u200a. A resolution parameter \\nk = σA + σC + σG + σT is defined, for controlling the alphabet size. The full composite alphabet of resolution \\nk\\u200a, denoted \\x1fk\\u200a, is the set of all composite letters, so that \\x1fi∈(A,C,G,T}σi = k\\u200a. Writing a composite letter is done by \\nusing a mixture of the DNA bases, determined by the letter’s ratio in the DNA synthesis cycle. Current synthesis \\ntechnologies produce multiple copies, and by using the predetermined base mixture each copy will contain a \\ndifferent base, thus preserving the ratio of the bases at the sequence-population level.\\nWhile the use of numerical ratios supports higher logical density in composite synthesis, it also introduces \\nchallenges related to the synthesis and inference of exact ratios. Combinatorial approaches, which also consist \\nof mixtures, address these challenges in a different way. Studies by Roquet et\\xa0al. (2021) and Yan et\\xa0al. (2023) \\ncontribute significantly to advancing DNA-based data storage technology. To encode and store data, Roquet et\\xa0al. \\nfocus on a novel combinatorial assembly method for DNA. Yan et\\xa0al. extend the frontiers of this technology by \\nenhancing the logical density of DNA storage, using enzymatically-ligated composite \\xadmotifs13,14.\\nIn this paper, we present a novel approach for encoding information in DNA, using combinatorial encoding \\nand shortmer DNA synthesis, leading to an efficient sequence design and improved DNA synthesis and readout \\ninterpretation. The method described herein leverages the advantages of combinatorial encoding schemes while \\nrelying on existing DNA chemical synthesis methods with some modifications. Using shortmer DNA synthesis \\nalso minimizes the effect of synthesis and sequencing errors. We formally define shortmer-based combinatorial \\nencoding schemes, explore different designs, and analyze their performance. We use computer-based simulations \\nof an end-to-end DNA-based data storage system built on combinatorial shortmer encodings, and study its \\nperformance. To demonstrate the potential of our suggested approach and experimentally test its validity, we \\nperformed an assembly-based molecular implementation of the proposed combinatorial encoding scheme and \\nanalyzed the resulting data. Finally, we discuss the potential of combinatorial encoding schemes and the future \\nwork required to enable these schemes in large-scale DNA-based data storage systems and other DNA data \\napplications.\\n',\n",
       " 'Results': 'Results\\n',\n",
       " 'Design of shortmer combinatorial encoding for DNA storage': 'Design of shortmer combinatorial encoding for DNA storage\\nWe suggest a novel method to extend the DNA alphabet while ensuring near-zero error rates.\\nLet \\x1f be a set of DNA k-mers that will serve as building blocks for our encoding scheme. Denote the elements \\nin \\x1f as X1, . . . , XN\\u200a. Elements in \\x1f are designed to be sufficiently different from each other, to minimize mix-up \\nerror probability. Formally, the set is designed to satisfy d\\n\\x1f\\nXi, Xj\\n\\x1e\\n≥d; ∀i \\x1c= j\\u200a, with the minimal Hamming \\ndistance d serving as a tunable parameter.\\nOther design criteria can be applied to the shortmers in \\x1f\\u200a, taking into consideration the properties of DNA \\nsynthesis, manipulation, and sequencing. These may include minimal Levenshtein distance, GC context, and \\navoiding long homopolymers. Clearly, any such filtering process will result in reduced alphabet size and reduced \\nlogical density.\\nNote that N = |\\x1f| ≤4k\\u200a. The elements in \\x1f will be used as building blocks for combinatorial DNA synthesis \\nin a method similar to the one used for composite DNA \\xadsynthesis3. Examples of k-mer sets \\x1f are presented in \\nSupplementary Sect.\\xa08.3.\\nWe define a combinatorial alphabet \\x1f over \\x1f as follows. Each letter in the alphabet represents a non-empty \\nsubset of the elements in \\x1f\\u200a. Formally, a letter σ ∈\\x1e\\u200a, representing a subset S ⊆\\x1f/∅\\u200a, can be written as an \\nN-dimensional binary vector where the indices for which σi = 1 represents the k-mers from \\x1f included in the \\nsubset S. We denote the k-mers in S as member k-mers of the letter σ\\u200a. For example, σ = (0, 1, 0, 1, 1, 0) represents \\nS = {X2, X4, X5} and |\\x1f| = N = 6\\u200a. Figure\\xa01a,b illustrate an example of a combinatorial alphabet using N = 16\\u200a, \\nin which every letter represents a subset of size 5 of \\x1f\\u200a. In\\xa0Sect.\\xa0“Binary and binomial combinatorial alphabets” \\nincludes a description of the construction of different combinatorial alphabets.\\nTo write a combinatorial letter σ in a specific position, a mixture of the member k-mers of σ is synthesized. To \\ninfer a combinatorial letter σ\\u200a, a set of reads needs to be analyzed to determine which k-mers are observed in the \\nanalyzed position (See\\xa0Sects. “Binary and binomial combinatorial alphabets” and “Reconstruction probabilities \\nfor binomial encoding” for more details). This set of k-mers observed in the sequencing readout and used for \\ninferring σ is referred to as inferred member k-mers. While the synthesis output and the sequencing readout will \\ninclude different counts for the member k-mers, the determination of the set of inferred k-mers will force binary \\nassignment for each k-mer to fit into the design scheme of combinatorial encoding.\\nFrom a hardware/chemistry perspective, the combinatorial shortmer encoding scheme can potentially be \\nbased on using the standard phosphoramidite chemistry synthesis technology, with some alterations (See Fig.\\xa01b \\nand Supplementary Sect.\\xa08.1)15,16. First, DNA k-mers should be used as building blocks for the \\xadsynthesis17. Such \\nreagents are commercially available for DNA trimers and were used, for example, for the synthesis of codon \\noptimization DNA \\xadlibraries18,19. In addition, a mixing step should be added to each cycle of the DNA synthesis \\nto allow mixing of the member k-mers prior to their introduction to the elongating molecules. Such systems are \\nyet to be developed and current attempts for combinatorial DNA synthesis are based on enzymatic assembly of \\nlonger DNA \\xadfragments13,14.\\nSimilar to composite DNA encoding, combinatorial encoding requires the barcoding of the sequences using \\nunique barcodes composed of standard DNA barcodes. This design enables direct grouping of reads pertaining \\nto the same combinatorial sequence. These groups of reads are the input for the process of reconstructing the \\ncombinatorial letters.\\nThe extended combinatorial alphabets allow for a higher logical density of the DNA-based storage system, \\nwhile the binary nature of the alphabet minimizes error rates.\\n3\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       " 'Binary and binomial combinatorial alphabets': 'Binary and binomial combinatorial alphabets\\nThe main parameter that defines a combinatorial encoding scheme is the alphabet \\x1f\\u200a. More specifically, it is the \\nset of valid subsets of \\x1f that can be used as letters. We define two general approaches for the construction of \\x1f\\u200a. \\nNamely, the binomial encoding and the full binary encoding.\\nIn the binomial encoding scheme, only subsets of \\x1f of size exactly K represent valid letters in \\x1f\\u200a, so that every \\nletter σ ∈\\x1e consists of exactly K member k-mers. Therefore, all the letters in the alphabet have the same \\nHamming weight K\\u200a. w(σ) = K, ∀σ ∈\\x1c\\u200a. This yields an effective alphabet of size |\\x1f| =\\n\\x1f\\nN\\nK\\n\\x1e\\n letters, where each \\ncombinatorial letter encodes log2(|\\x1e|) = log2\\n\\x1f\\nN\\nK\\n\\x1e\\n bits. An r-bit binary message requires \\nr\\nlog2\\n\\x1f N\\nK\\n\\x1e synthesis \\ncycles (and a DNA molecular segment with length \\nkr\\nlog2\\n\\x1f N\\nK\\n\\x1e ). In practice, we would prefer working with alphabet \\nsizes that are powers of two, where each letter will encode for \\n\\x1f\\nlog2\\n\\x1e\\nN\\nK\\n\\x1d\\x1c\\n bits. Note that this calculation ignores \\nerror correction redundancy, random access primers, and barcodes, which are all required for message \\nreconstruction. See Supplementary Sect.\\xa08.2 and Fig.\\xa01a, which illustrate a trimer-based binomial alphabet with \\nN = 16 and K = 5\\u200a, resulting in an alphabet of size |\\x1f| =\\n\\x1f\\n16\\n5\\n\\x1e\\n= 4, 368 that allows to encode ⌊log2(4368)⌋= 12 \\nbits per letter or synthesis position.\\nIn the full binary encoding scheme, all possible nonempty subsets of \\x1f represent valid letters in the alphabet. \\nThis yields an effective alphabet of size |\\x1f| = 2N −1 letters, each encoding for \\n\\x1f\\nlog2 (|\\x1e|)\\n\\x1e\\n= N −1 bits.\\nFrom this point on, we focus on the binomial encoding.\\nFigure\\xa01.\\u2002 \\u2009Our combinatorial encoding and synthesis approach. (a) Schematic view of a combinatorial \\nalphabet (Encode legend). A set of 16 trimers, X1, . . . , X16\\u200a, is used to construct 4096 combinatorial letters, \\neach representing a subset of 5 trimers as indicated on the right and depicted in the grayed-out cells of the \\ntable. (b) A suggested approach for combinatorial shortmer synthesis. A modified synthesizer would include \\ndesignated containers for the 16-trimer building blocks and a mixing chamber. Standard DNA synthesis is used \\nfor the barcode sequence (1), while the combinatorial synthesis proceeds as follows: The trimers included in \\nthe synthesized combinatorial letter are injected into the mixing chamber and introduced into the elongating \\nmolecules (2). The process repeats for the next combinatorial letter (3), and finally, the resulting molecules are \\ncleaved and collected (4).\\n4\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       " 'Reconstruction probabilities for binomial encoding': 'Reconstruction probabilities for binomial encoding\\nIn this section, the performance characteristics of binomial encoding are investigated. Specifically, we present a \\nmathematical analysis of the probability of successfully reconstructing the intended message. In Sects.\\xa0“An end-\\nto-end combinatorial shortmer storage system” and \"Experimental proof of concept\", results are presented from \\nour simulations and a small-scale molecular implementation of the binomial encoding, respectively.\\n',\n",
       " 'Reconstruction of a single combinatorial letter': 'Reconstruction of a single combinatorial letter\\nSince every letter σ ∈\\x1e consists exactly of the K member k-mers, the required number of reads for observing at \\nleast one read of each member k-mer in a single letter follows the coupon collector \\xaddistribution20. The number \\nof reads required to achieve this goal can be described as a random variable R = \\x1fK\\ni=1Ri\\u200a, where R1 = 1 and \\nRi ∼Geom\\n\\x1f K−i+1\\nK\\n\\x1e\\n, i = 2, . . . , K\\u200a. Hence, the expected number of required reads, is:\\nwhere Har(·) is the harmonic number.\\nThe expected number of reads required for reconstructing a single combinatorial letter thus remains \\nreasonable for the relevant values of K\\u200a. For example, when using a binomial encoding with K = 5 the expected \\nnumber of reads required for reconstructing a single combinatorial letter is roughly 11.5\\u200a, which is very close to \\nthe experimental results presented in Sect. \"Experimental proof of concept\".\\nBy Chebyshev’s inequality (See\\xa0Sect.\\xa0\"Reconstruction probability of a binomial encoding letter\"), we can \\nderive a (loose) upper bound on the probability of requiring more than E[R] + cK reads to observe at least one \\nread of each member k-mer, where c > 1 is a parameter:\\nFor example, when using a binomial encoding with K = 5\\u200a, the probability of requiring more than 26.5 reads \\n(corresponding to c = 3\\u200a) is bounded by 0.18\\u200a, which is consistent with the experimental result shown in Fig.\\xa05d.\\n',\n",
       " 'Reconstruction of a combinatorial sequence': 'Reconstruction of a combinatorial sequence\\nWhen we examine an entire K-subset binomial encoded combinatorial sequence of length l\\u200a, we denote by R(l) \\nthe required number of reads to observe K distinct k-mers in every position. Assuming independence between \\ndifferent positions and not taking errors into account, we get the following relationship between c and any desired \\nconfidence level 1 −δ (See\\xa0Sect.\\xa0\"Reconstruction probability of a binomial encoding letter\" for details):\\nAnd therefore:\\nThe number of reads required to guarantee reconstruction of a binomial encoded message, at a 1 −δ \\nprobability, with K = 5, and l synthesized positions, is thus KHar(K) + cK when\\nSupplementary Table\\xa0S2 shows several examples of this upper bound. As demonstrated in the simulations \\nand the experimental results, this bound is not tight (See Sects.\\xa0“An end-to-end combinatorial shortmer storage \\nsystem” and \"Experimental proof of concept\").\\nNote that with an online sequencing technology (such as nanopore sequencing) the sequencing reaction can \\nbe stopped after K distinct k-mers are confidently observed.\\nTo take into account the probability of observing a k-mer that is not included in \\x1f (e.g., due to synthesis or \\nsequencing error), we can require that at least t > 1 reads of each of the K distinct k-mers will be observed. This \\nis experimentally examined in\\xa0Sect. \"Experimental proof of concept\", while the formal derivation of the number \\nof required reads is not as trivial, and will be addressed in future work.\\nThe above analysis is based only on oligo recovery, which depends solely on the sampling rate, ignoring \\npossible mix-up errors (i.e., incorrect k-mer readings). This assumption is based on the near-zero mix-up \\nprobability attained by the construction of \\x1f\\u200a, which maximizes the minimal Hamming distance between elements \\nin \\x1f\\u200a. In\\xa0Sect. \"Experimental proof of concept\", this analysis is compared to experimental results obtained from \\nusing synthetic combinatorial DNA.\\nAn end‑to‑end combinatorial shortmer storage system\\nWe suggest a complete end-to-end workflow for DNA-based data storage with the combinatorial shortmer \\nencoding presented in Fig.\\xa02. The workflow begins with encoding, followed by DNA synthesis, storage, and \\nsequencing, and culminates in a final decoding step. A 2D Reed-Solomon (RS) error correction scheme, which \\ncorrects errors in the letter reconstruction (for example, due to synthesis, sequencing, and sampling errors) and \\n(1)\\nE[R] =\\n\\x1fK\\ni=1E[Ri] = K\\n\\x1fK\\ni=1\\n1\\ni = KHar(K)\\n(2)\\nP(|R −KHar(K)| ≥cK) ≤π2\\n6c2\\n(3)\\nP(|R(l) −KHar(K)| ≥cK) ≤1 −\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n< δ\\n(4)\\nP(R(l) < KHar(K) + cK) ≥\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n≥1 −δ\\n(5)\\nc ≥\\n\\x1f\\n1/6π\\n\\x1e\\n1 −(1 −δ)1/l\\x1d−1/2',\n",
       " 'An end-to-end combinatorial shortmer storage system': '\\n7\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       " 'Experimental proof of concept': 'Experimental proof of concept\\nTo assess and establish the potential of large combinatorial alphabets, we performed a small-scale experimental \\nproof of concept study demonstrating the encoding and decoding of a 96-bit input message, which is equivalent \\nto the text “DNA Storage!”. Since combinatorial DNA synthesis technology is not yet available, we demonstrated \\nthe combinatorial approach using Gibson assembly as an ad-hoc imitation for combinatorial synthesis. We \\nconstructed two combinatorial sequences, each containing a barcode and four payload cycles over a binomial \\nalphabet with N = 16 and K = 5\\u200a. The assembly was performed using DNA fragments composed of a 20-mer \\ninformation sequence and an overlap of 20\\xa0bp between adjacent fragments, as shown in Fig.\\xa05a. The assembled \\nDNA was then stored and sequenced for analysis using Illumina Miseq (See Table\\xa03 and Sect.\\xa0\"Cost analysis\" for \\ndetails about the sequencing procedures).\\nThe sequencing output was then analyzed using the procedure described in\\xa0Sect.\\xa0“Decoding and analysis”. \\nBoth combinatorial sequences were successfully reconstructed from the sequencing reads, as presented in Fig.\\xa05b, \\nand\\xa0Supplementary Figs. S1, S2, and S3. The experiment also demonstrated the robustness of the binomial DNA \\nencoding for synthesis and sequencing errors, as described in Fig.\\xa05c. We observed a minor leakage between the \\ntwo synthesized sequences, which was overcome by the reconstruction pipeline (See Fig.\\xa05c, and\\xa0Supplementary \\nFigs.\\xa0S1, S2, and S3). Note that there is a partial overlap between the member k-mers of the two sequences.\\nFor comparison, a recent study \\xadby14 encoded the 84-bit phrase “HelloWord” using\\xa0a different encoding \\nand\\xa0synthesis approach. A comparison between the two experiments is shown in Table\\xa02. For example, while we \\nused Gibson assembly as our synthesis method, they introduced a new method called Bridge Oligonucleotide \\nAssembly. We encoded 12 bits per synthesis cycle and assembled four combinatorial fragments in each \\nsequence, \\xadwhile14 encoded 84 bits in a single combinatorial cycle. Our 96-bit were split and encoded using two \\ncombinatorial sequences, while they encoded the same 84-bits message, in its full format, on eight different \\na\\nb\\nc\\nd\\nText\\nPayload\\nPayload\\nPayload\\nRS\\nRS\\nRS\\nRS\\niv\\ni\\nii\\niii\\nText\\nPayload\\nPayload\\nPayload\\nRS\\nRS\\nRS\\nRS\\n1\\n2\\n3\\n8\\n7\\n6\\n5\\n4\\nFigure\\xa03.\\u2002 \\u2009Simulation of an end-to-end combinatorial shortmer encoding. (a) A schematic view of the \\nsimulation workflow. A text message is translated into a combinatorial message (1), and encoded using RS \\nerror correction on the barcode and payload (2). Each block is encoded using outer RS error correction (3). \\nDNA synthesis and sequencing are simulated under various error schemes, and the combinatorial letters are \\nreconstructed (4–5). RS decoding is performed on each block (6) and on each sequence (7) before translation \\nback to text (8). The Roman numerals (i-iv) represent the different error calculations. (b) Error rates in different \\nstages of the decoding process. Boxplot of the normalized Levenshtein distance (See Sect.\\xa0\"Reconstruction\") \\nfor the different stages in a simulation (30 runs) of sampling 100 reads, with an insertion error rate of 0.01. The \\nX-axis represents the stages of error correction (before 2D RS decoding (iv), after RS payload decoding (iii), and \\nafter 2D RS decoding (ii)). (c,d) Sampling rate effect on overall performance. Normalized Levenshtein distance \\nas a function of sampling rate before RS decoding (c) and after 2d RS decoding (ii). Different lines represent \\ndifferent error types (substitution, deletion, and insertion) introduced at a rate of 0.01.\\n9\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n140 reads to obtain 1 −δ = 0.99 probability of reconstruction. Clearly, this is well above the observed number \\nof 55.60 reads. Note, as explained, the calculated bound is a loose bound.\\nThe reconstruction procedure ends with a set of inferred k-mers that represent the inferred combinatorial \\nletter. This set is not guaranteed to be correct, especially when using t = 1\\u200a, which means that noisy reads may \\nresult in an incorrect k-mer included in the inferred letter. Figure\\xa05f depicts the rate of incorrect reconstructions \\nas a function of the number of required copies for each inferred k-mer (\\u200at = 1, 2, 3, 4\\u200a). Note that with t ≥3 results \\nin 100% successful reconstruction. This, however, comes with a price, where more reads must be analyzed.\\n',\n",
       " 'Discussion': 'Discussion\\nIn this study, we introduced combinatorial shortmer encoding for DNA-based data storage, which extends the \\napproach of composite DNA by resolving some of its sensitivity related issues. Combinatorial shortmer encoding \\nallows for increased logical density while maintaining low error rates and high reconstruction rates. We explored \\ntwo encoding schemes, binary and binomial, and evaluated some of their theoretical and practical characteristics. \\nThe inherent consistency of the binomial encoding scheme, where every letter in the sequence consists of exactly \\nK distinct member k-mers, ensures uniformity in the encoded DNA sequences. This approach not only simplifies \\nthe reading process, but also allows for a more streamlined decoding. For instance, technologies like nanopore \\nsequencing enable continuous sequencing until all k-mers at a given position are confidently observed.\\nOur suggested approach is designed to inherently overcome base substitution errors, which are the most \\ncommon errors expected in every DNA-based data storage system that includes DNA sequencing. This is \\nachieved by the selection of a set of N k-mer building blocks to be resilient to single-base substitutions. Other \\nconsiderations may also be incorporated in the selection of the set of valid k-mers, taking into account any \\nbiological, chemical, or technological constraints. This represents an inherent tradeoff in DNA-based data\\xa0storage \\nbetween sequence constraints and information density. Insertion and deletion errors, which usually originate \\nin the synthesis process, are more challenging to overcome. We introduced a 2D RS error correction scheme \\non the shortmer level, allowing for a successful message reconstruction even with error levels exceeding those \\nexpected in reality.\\nOur study highlights the significant effect of sampling rates on the overall performance of the system. The \\naccuracy and completeness of sequence reconstruction require each of the sequences to be observed with a \\nsufficiently high coverage. Our subsampling experiments underpin this observation, demonstrating the need \\nfor calibration of sampling rates to ensure the desired fidelity in DNA-based data storage. The crucial role of \\nthe sampling rate was also highlighted \\xadin3. However, while composite DNA uses mixed letter with varying \\nproportions of the different letters, the combinatorial encoding, studied in this current work, uses binary mixtures \\nand does not rely on proportions. This potentially allows scaling up the combinatorial encoding without a \\nsignificant effects on the required sampling rates.\\nCombinatorial DNA coding can potentially reduce the overall costs of DNA-based data storage. Considering \\nboth sequencing costs, which fluctuate, and synthesis costs, which consistently drop, the increase in the \\nalphabet size is accompanied by a decrease in overall cost. However, combinatorial DNA synthesis or assembly \\nis still unavailable for large-scale commercial use. Thus, further development of combinatorial DNA synthesis \\ntechnologies will continue to impose limitations and constraints on combinatorial encoding, and determine \\nthe overall costs.\\nWhile our proof-of-concept experiment showed success on a small scale, there are complexities to be \\naddressed in considering large-scale applications. These include synthesis efficiency, error correction, and \\ndecoding efficiency. Nonetheless, the resilience of our binomial DNA encoding for both synthesis and sequencing \\nerrors highlights its practical potential and scalability. One specific aspect is the effect of combinatorial encoding \\non possible sequence-related constraints. While sequences with unwanted compositions (e.g., containing \\nhomopolymers) will unavoidably be part of the synthesized mixtures, the uniform sampling of the combinatorial \\nshortmers in each position, together with the independence of the different positions, guarantees that only very \\nfew such sequences will be aythesized. In particular—these will not interfere with successful reconstruction. \\nAnother challenging aspect of scaling up combinatorial DNA systems is the need to use longer DNA k-mers to \\nconstruct larger sets with the desired constraints. This may make the combinatorial synthesis impractical and \\nwill require balancing the increase in logical density with the technological complexity.\\nSeveral future research directions emerge from our study. First, it is important to develop error correction \\nmethods for better handling insertion and deletion errors. One approach for achieving this goal, is to adjust \\nsampling rates: optimizing the sampling rate, especially in large-scale experiments, can lead to data retrieval at \\nhigh accuracy. While our study highlighted the role of sampling rates in achieving desired outcomes, delving \\ndeeper into the underlying theory will lead to more improvements. For example, based on theoretical bounds of \\nsampling rates, more concrete recommendations can be provided for real-world applications. The development \\nof error correction codes, designed specifically to overcome the error types that characterize combinatorial \\nencoding, is another important direction for future research. Most notably, transitioning from small-scale proof-\\nof-concept experiments to larger-scale implementations is an important next step. Evaluating the scalability of \\nour method across various scales and complexities will be enlightening, especially when considering synthesis \\nefficiency and error rates. Finally, the consideration of advanced sequencing technologies could redefine the \\npotential and efficacy of our proposed method, including its future practical implementation.\\nTo sum up, combinatorial DNA synthesis and sequence design are important beyond the scope of DNA-based \\ndata storage. Generating combinatorial DNA libraries is an efficient tool in synthetic biology, better supporting \\nlarge-scale experiments. DNA synthesis technologies that can incorporate a combinatorial synthesis of longer \\nDNA fragments will enable the design and generation of more DNA libraries with applications in data storage \\nand beyond.\\n11\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       " 'Methods': 'Methods\\n',\n",
       " 'Reconstruction probability of a binomial encoding letter': 'Reconstruction probability of a binomial encoding letter\\nLet the number of reads required for reconstruction be a random variable R = \\x1fK\\ni=1Ri where R1 = 1 and \\nRi ∼Geom\\n\\x1f K−i+1\\nK\\n\\x1e\\n, i = 2, . . . , K\\u200a. Hence, the expected number of required reads, is:\\nwhere Har(·) is the harmonic number.\\nUsing the independence of Ri\\u200a, the variance of R can be bound by \\xad(See25):\\nBy Chebyshev’s inequality, we get an upper bound (a loose bound) on the probability of requiring more than \\nE[R] + cK reads to observe at least one read of each member k-mer:\\nLet c = b π\\n√\\n6\\u200a, or b = c\\n√\\n6\\nπ \\u200a, and we obtain:\\nOr specifically:\\nWe now turn to address the reconstruction of an entire oligo of length l\\u200a. Let R(l) be the random variable \\nrepresenting the number of reads required to have observed all the K member k-mers in every position. Setting \\nany δ > 0\\u200a, if we show that P(R(l) > m) ≥1 −δ\\u200a, then we know that by accumulating m reads the probability of \\ncorrect full reconstruction is more than 1 −δ\\u200a. From Eq.\\xa0(11), and assuming independence of the positions (in \\nterms of observing all K member k-mers), we get Eq.\\xa0(12):\\nFrom which we can extract c\\u200a, so that:\\nWhich yields:\\nThis process allows us to evaluate the sequencing depth complexity. For example, consider l = 100 and \\nδ = 0.01\\u200a. We want to find c\\u200a, so that using KHar(K) + cK reads will reconstruct the entire sequence with 0.99 \\nprobability. We therefore set:\\n(6)\\nE[R] =\\n\\x1fK\\ni=1E[Ri] = K\\n\\x1fK\\ni=1\\n1\\ni = KHar(K)\\n(7)\\nVar(R) =\\n\\x1fK\\ni=1Var(Ri) < K2\\n\\x1e 1\\n12 + 1\\n22 + · · · + 1\\nK2\\n\\x1d\\n< π2\\n6 K2\\n(8)\\nP(|R −E(R)| ≥bσ) ≤1\\nb2\\n(9)\\nP\\n\\x1f\\n|R −E(R)| ≥b π\\n√\\n6\\nK\\n\\x1e\\n≤1\\nb2\\n(10)\\nP(|R −E[R]| ≥cK) ≤π2\\n6c2\\n(11)\\nP(|R −KHar(K)| ≥cK) ≤π2\\n6c2\\n(12)\\nP(R(l) < KHar(K) + cK) ≥\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n(13)\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n≥1 −δ\\n(14)\\nc ≥\\nπ\\n\\x1f\\n6\\n\\x1e\\n1 −(1 −δ)\\n1\\nl\\n\\x1d\\nFigure\\xa05.\\u2002 \\u2009Experiment analysis. (a) A schematic view of the Gibson assembly. Each combinatorial sequence \\nconsists of a barcode segment and four payload segments (denoted as cycles 1–4). (b) Reconstruction results \\nof the two combinatorial sequences. The color indicates read frequency, and the member k-mers are marked \\nwith orange boxes. (c) The distribution of reads over the 16\\xa0k-mers in an example combinatorial letter. \\nOverlaid histograms represent the percentage of reads for each of the 16\\xa0k-mers for the same position in our \\ntwo combinatorial sequences. This, in fact, is an enlarged view of the two c4 columns of panel b. (d) Required \\nnumber of reads for reconstructing a single combinatorial letter. A histogram of the number of reads required \\nto observe at least t = 1, 2 reads from K = 5 inferred k-mers. The results are based on resampling the reads 500 \\ntimes, the data represents cycle 4. (e) Required number of reads for reconstructing a four-letter combinatorial \\nsequence. Similar to d. (f), Reconstruction failure rate as a function of the required multiplicity t\\u200a. Erroenous \\nreconstruction rate shown for different values of required copies to observe each inferred k-mer (\\u200at = 1, 2, 3, 4\\u200a). \\nThe mean required number of reads for reconstruction is displayed using a secondary Y-axis in the dashed lines.\\n◂\\n12\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\nAnd get:\\nAnd therefore, using 128 reads guarantees reconstruction with 0.99 probability.\\nAn end‑to‑end combinatorial storage system\\nIn Sect.\\xa0“An end-to-end combinatorial shortmer storage system” we propose an end-to-end combinatorial storage \\nsystem, as follows.\\nCombinatorial encoding and padding\\nA binary message is encoded using a large k-mer combinatorial alphabet (e.g., trimer-based alphabet of size \\n|\\x1f| = 4096 letters, with N = |\\x1f| = 16\\u200a), resulting in r = 12 bits per combinatorial letter. The binary message is \\nzero-padded to ensure its length is divisible by r prior to the combinatorial encoding. The complete message is \\nbroken into sequences of set length l = 120\\u200a, each sequence is then marked with a standard DNA barcode and \\ntranslated using the table presented in the Encode legend (See\\xa0Supplementary Sect.\\xa08.2).\\nThe length of the complete combinatorial sequence must be divisible by the payload size l and by the block \\nsize B\\u200a. As described in Fig.\\xa06, this is ensured using another padding step, and the padding information is included \\nin the final combinatorial sequence.\\nError correction codes\\nThe 2D error correction scheme includes the use of three \\xadRS26 encodings: on each barcode, on the payload part \\nof each sequence, and an outer error correction code on each block of sequences.\\n•\\t Each barcode is encoded using a systematic RS(6,8) code over GF(24)\\u200a, transforming the unique 12nt barcode \\ninto a 16nt sequence.\\n•\\t Each 120 combinatorial letter payload sequence is encoded using an RS(120,134) code over GF(212)\\u200a, resulting \\nin a sequence of length 134 combinatorial letters.\\n•\\t To protect against sequence dropouts, an outer error correction code is used on the columns of the matrix \\n(See Fig.\\xa06). Each block of B = 42 sequences, is encoded using a RS(42,48) RS code GF\\n\\x1f\\n212\\x1e\\n\\u200a. This is applied \\nin each column separately.\\nFor simplicity, Fig.\\xa06 demonstrates the encoding of\\u2009~\\u20090.1\\xa0KB using shorter messages with simpler error \\ncorrection codes. The following parameters are used:\\n•\\t A barcode length of 6nt encoded using RS(3,5) code over GF\\n\\x1f\\n24\\x1e\\n to get 10nt.\\n•\\t A payload length of l = 12 encoded using RS(12,18) over GF\\n\\x1f\\n29\\x1e\\n for the \\n\\x1f\\n16\\n3\\n\\x1e\\n binomial alphabet.\\n•\\t A 10-sequence block encoded, column-wise, using a (10,15) RS code over GF(29).\\n(15)\\n\\x1f\\n1 −π2\\n6c2\\n\\x1e100\\n≥0.99\\n(16)\\nc ≥\\nπ\\n\\x1f\\n6\\n\\x1e\\n1 −(0.99)0.01\\x1d = 127.94\\nTable\\u202f2.\\u2002 \\u2009Comparison of our experiment with data \\xadfrom14.\\nThis work\\nYan et\\xa0al.14\\nMessage length (bits)\\n96\\n84\\nLogical density (bits/ Synthesis Cycle)\\n12\\n84\\nSequence length (Cycles)\\n4\\n1\\nNumber of sequences\\n2\\n8\\nNumber of unique k-mers (N)\\n16\\n96\\nNumber of unique k-mers in each letter (K)\\n5\\n32\\nSequence length\\n220\\xa0bp\\n75\\xa0bp\\nBarcode length\\n20\\xa0bp\\n25\\xa0bp\\nSpace / Adjacent fragments\\n20\\xa0bp\\n25\\xa0bp, same in all BC\\nPayload length\\n20\\xa0bp\\n25\\xa0bp\\nSequencing method\\nIlumina Miseq\\nNanopore\\nSynthesis\\nGibson Assembly\\nBridge Oligonucleotide Assembly\\nAmount of space / Adjacent fragments\\n6\\n1',\n",
       " 'An end-to-end combinatorial storage system': '',\n",
       " 'Combinatorial encoding and padding': 'Combinatorial encoding and padding\\nA binary message is encoded using a large k-mer combinatorial alphabet (e.g., trimer-based alphabet of size \\n|\\x1f| = 4096 letters, with N = |\\x1f| = 16\\u200a), resulting in r = 12 bits per combinatorial letter. The binary message is \\nzero-padded to ensure its length is divisible by r prior to the combinatorial encoding. The complete message is \\nbroken into sequences of set length l = 120\\u200a, each sequence is then marked with a standard DNA barcode and \\ntranslated using the table presented in the Encode legend (See\\xa0Supplementary Sect.\\xa08.2).\\nThe length of the complete combinatorial sequence must be divisible by the payload size l and by the block \\nsize B\\u200a. As described in Fig.\\xa06, this is ensured using another padding step, and the padding information is included \\nin the final combinatorial sequence.\\n',\n",
       " 'Error correction codes': 'Error correction codes\\nThe 2D error correction scheme includes the use of three \\xadRS26 encodings: on each barcode, on the payload part \\nof each sequence, and an outer error correction code on each block of sequences.\\n•\\t Each barcode is encoded using a systematic RS(6,8) code over GF(24)\\u200a, transforming the unique 12nt barcode \\ninto a 16nt sequence.\\n•\\t Each 120 combinatorial letter payload sequence is encoded using an RS(120,134) code over GF(212)\\u200a, resulting \\nin a sequence of length 134 combinatorial letters.\\n•\\t To protect against sequence dropouts, an outer error correction code is used on the columns of the matrix \\n(See Fig.\\xa06). Each block of B = 42 sequences, is encoded using a RS(42,48) RS code GF\\n\\x1f\\n212\\x1e\\n\\u200a. This is applied \\nin each column separately.\\nFor simplicity, Fig.\\xa06 demonstrates the encoding of\\u2009~\\u20090.1\\xa0KB using shorter messages with simpler error \\ncorrection codes. The following parameters are used:\\n•\\t A barcode length of 6nt encoded using RS(3,5) code over GF\\n\\x1f\\n24\\x1e\\n to get 10nt.\\n•\\t A payload length of l = 12 encoded using RS(12,18) over GF\\n\\x1f\\n29\\x1e\\n for the \\n\\x1f\\n16\\n3\\n\\x1e\\n binomial alphabet.\\n•\\t A 10-sequence block encoded, column-wise, using a (10,15) RS code over GF(29).\\n(15)\\n\\x1f\\n1 −π2\\n6c2\\n\\x1e100\\n≥0.99\\n(16)\\nc ≥\\nπ\\n\\x1f\\n6\\n\\x1e\\n1 −(0.99)0.01\\x1d = 127.94\\nTable\\u202f2.\\u2002 \\u2009Comparison of our experiment with data \\xadfrom14.\\nThis work\\nYan et\\xa0al.14\\nMessage length (bits)\\n96\\n84\\nLogical density (bits/ Synthesis Cycle)\\n12\\n84\\nSequence length (Cycles)\\n4\\n1\\nNumber of sequences\\n2\\n8\\nNumber of unique k-mers (N)\\n16\\n96\\nNumber of unique k-mers in each letter (K)\\n5\\n32\\nSequence length\\n220\\xa0bp\\n75\\xa0bp\\nBarcode length\\n20\\xa0bp\\n25\\xa0bp\\nSpace / Adjacent fragments\\n20\\xa0bp\\n25\\xa0bp, same in all BC\\nPayload length\\n20\\xa0bp\\n25\\xa0bp\\nSequencing method\\nIlumina Miseq\\nNanopore\\nSynthesis\\nGibson Assembly\\nBridge Oligonucleotide Assembly\\nAmount of space / Adjacent fragments\\n6\\n1\\n13\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\nThe 824 bits are first padded to be 828 = 92 ∗9\\u200a. The 92 combinatorial letter message is split into 7 sequences of \\n12 letters and an additional sequence of 8 letters. Finally, a complete block of 12 sequences (total of 10 ∗12 = 120 \\nletters) is created by padding with one additional sequence of 12 letters and including the padding information \\nas the last sequence.\\n',\n",
       " 'Synthesis and sequencing simulation with errors': 'Synthesis and sequencing simulation with errors\\n•\\t Simulating the synthesis process. DNA molecules pertaining to the designed sequences are synthesized \\nusing combinatorial k-mer DNA synthesis (See Fig.\\xa01b). For each combinatorial sequence, we first determine \\nthe number of synthesized copies by sampling from X ∼N(µ = 1000, σ 2 = 100)\\u200a. Let x be the number of \\ncopies for a specific sequence. Next, for every position in the sequence, we uniformly sample x independent \\nk-mers from the set of member k-mers of the combinatorial letter in the specific position. We concatenate \\nthe sampled k-mers to the already existing x synthesized molecules.\\n•\\t Error simulation. Synthesis and sequencing errors are simulated as follows. Error probabilities for deletion, \\ninsertion, and substitution are given as parameters denoted as Pd, PI, and Ps respectively. Deletion and \\ninsertion errors are assumed to occur during k-mer synthesis and thus implemented on the k-mer level (i.e., \\nan entire k-mer is deleted or inserted in a specific position during the synthesis simulation). Substitution \\nerrors are assumed to be sequencing errors and hence implemented on a single base level (i.e., a single letter \\nis substituted, disregarding the position within the k-mer).\\n•\\t Mixing. Post synthesis, molecules undergo mixing to mirror genuine molecular combinations. This is \\nachieved through a randomized data line shuffle using an SQLite database, enabling shuffle processes even \\nfor sizable input \\xadfiles27.\\nFigure\\xa06.\\u2002 \\u2009Example of message coding, including padding and RS error correction. Encoding of a\\u2009~\\u20090.1\\xa0KB \\nmessage into a 512 letter binomial alphabet (\\u200aN = 16, K = 3)\\u200a. First, bit padding is added, included here in \\nthe letter 1\\nσ257\\u200a. Next, block padding is added, included here in 2\\nσ1 and 3\\nσ1\\u200a. Padding information is included in the \\nlast sequence of all blocks. The last sequence holds the number of padding binary bits. In this example, 4\\nσ149 \\nrepresents 148 bits of padding, composed of 4 + (4 ∗9) + (12 ∗9)bits\\u200a, 4 bits from 1\\nσ257\\u200a, 4 letters from 2\\nσ1 and 12 \\nletters from 3\\nσ1.\\n14\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n•\\t Reading and sampling. From the simulated synthesized molecule set, a subsample of predefined size \\nS ∗numberofsynthesizedseqeunces is drawn, simulating the sampling effect of the sequencing process.\\n',\n",
       " 'Reconstruction': 'Reconstruction\\n•\\t Barcode decoding The barcode sequence of each read is decoded using the RS(6,8) code.\\n•\\t Grouping by barcode The reads are then grouped by their barcode sequence to allow the reconstruction of \\nthe combinatorial sequences.\\n•\\t Filtering of read groups Barcodes (set of reads) with less than 10% of the sampling rate S reads are discarded.\\n•\\t Combinatorial reconstruction For each set of reads, every position is analyzed separately. The K most \\ncommon k-mers are identified and used to determine the combinatorial letter σ in this position. Let \\n\\x1f be the difference between the length of the analyzed reads and the length of the designed sequence. \\n\\x1f = l −len(read)\\u200a. Reads with |\\x1f| > k −1 are discarded from the analysis. Invalid k-mers (not in \\x1f) are \\nreplaced by a dummy k-mer Xdummy.\\n•\\t Missing barcodes Missing barcodes are replaced with dummy sequences to enable correct outer RS decoding.\\n•\\t Normalized Levenshtein distance Levenshtein distance between the observed sequence O and the expected \\nsequence E is \\xadcalculated28,29. Normalized Levenshtein distance is calculated by dividing the distance by the \\nlength of the expected sequence:\\n',\n",
       " 'Cost analysis': 'Cost analysis\\nSynthesis cost estimation was performed using the logical density calculation presented in Supplementary \\nSect.\\xa08.5 and Supplementary Table\\xa0S1. To calculate the sequencing costs, we used the coupon collector model \\npresented \\xadin24 to assess the required sequencing depth given the combinatorial alphabet. Figure\\xa04b indicates the \\ntotal number of reads required for reconstructing the sequences, calculated as the required sequencing depth \\nmultiplied by the number of sequences from Supplementary Sect.\\xa08.5 and Supplementary Table\\xa0S1. The analysis \\nwas performed on the following set of combinatorial alphabets: Standard DNA, \\n\\x1f\\n8\\n4\\n\\x1e\\n,\\n\\x1f\\n16\\n3\\n\\x1e\\n,\\n\\x1f\\n16\\n5\\n\\x1e\\n,\\n\\x1f\\n16\\n7\\n\\x1e\\n,\\n\\x1f\\n32\\n10\\n\\x1e\\n,\\n\\x1f\\n32\\n16\\n\\x1e\\n,\\n\\x1f\\n64\\n32\\n\\x1e\\n,\\n\\x1f\\n96\\n32\\n\\x1e\\n.\\n',\n",
       " 'Proof of concept experiment': 'Proof of concept experiment\\nThe proof-of-concept experiment was performed by imitating combinatorial synthesis using Gibson assembly \\nof larger DNA fragments. Each DNA fragment was composed of a 20-mer information sequence and an overlap \\nof 20\\xa0bp between adjacent fragments, as depicted in Fig.\\xa05a. Two combinatorial sequences were designed, each \\ncomposed of a barcode fragment, 4 payload fragments, and Illumina\\xa0Miseq P5 and P7 anchors at the ends. The \\ninformation fragments included in each combinatorial position were chosen from a set of 16 sequences with \\nsufficient pair-wise distance. The full list of DNA sequences and the design of combinatorial sequences is listed \\nin Supplementary Sect.\\xa08.6.\\n',\n",
       " 'DNA assembly and sequencing': 'DNA assembly and sequencing\\nPayload, barcode, and P7 anchor fragments with 20\\xa0bp overlaps for the purpose of Gibson assembly were \\nproduced by annealing complementary oligonucleotides manufactured by Integrated DNA Technologies (IDT). \\nOligos were dissolved in Duplex Buffer (100\\xa0mM Potassium Acetate; 30\\xa0mM HEPES, pH 7.5; available from \\nIDT) to the final concentration of 100 micromolar. For annealing, 25\\xa0µl of each oligo in a pair were combined \\nto the final concentration of 50 micromolars. The oligo mixes were incubated for 2\\xa0min at \\xad940 C, and gradually \\ncooled down to room temperature. The annealed payload oligos that belonged to the same cycle (5 oligos total) \\nwere mixed to the final concentration of 1 micromolar per oligo—a total of 5 micromolar, by adding 2\\xa0µl of each \\nannealed oligo into the 90\\xa0µl of nuclease-free water—a final volume of 100\\xa0µl. Annealed barcode and P7 anchor \\noligos were also diluted to the final concentration of 5 micromolar in nuclease-free water, after thorough mixing \\nby vortexing. The diluted oligos were stored at −20\\xa0°C.\\nImmediately prior to the Gibson assembly, payload oligo mixes, barcode, and P7 anchor oligos were further \\ndiluted 100-fold to the final working dilution of 0.05\\xa0pmol/microliter in nuclease-free water. Gibson reaction \\nwas assembled by adding 1\\xa0µl (0.05\\xa0pmol) of barcode, 4\\u2009×\\u2009cycle mixes, and P7 anchor to the 4\\xa0µl of nuclease-free \\nwater and supplemented with 10\\xa0µl of NEBuilder HiFi DNA assembly master mix (New England Biolabs (NEB)) \\nto the final volume of 20\\xa0µl, according to the manufacturer instructions. The reactions were incubated for 1\\xa0h \\nat 50\\xa0°C, and purified with AmpPure Beads (Thermo Scientific) at 0.8X ratio (16\\xa0µl of beads per 20\\xa0µl Gibson \\nreaction) to remove free oligos / incomplete assembly products. After adding beads and thorough mixing, \\nthe reactions were incubated for 10\\xa0min at room temperature and then placed on a magnet for 5\\xa0min at room \\ntemperature. After removing the sup, the beads were washed twice with 100\\xa0µl of 80% ethanol. The remaining \\nwashing solution was further removed by a 20\\xa0µl tip, and the beads dried for 3\\xa0min on the magnet with an open \\nlid. After removing from the magnet, the beads were resuspended in 22\\xa0µl of IDTE buffer (IDT), incubated for \\n5\\xa0min at room temperature, and then placed back on the magnet.\\n20\\xa0µl of eluate were transferred into the separate 1.7\\xa0ml tube. 5\\xa0µl of the eluted DNA were used as a template \\nfor PCR amplification combined with 23\\xa0µl of nuclease-free water, 1\\xa0µl of 20 micromolar indexing primer 5, \\n(17)\\nNormalized Levenshtein distance (O, E) = Levenshtein distance (O, E)\\n|O|\\n15\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n1\\xa0µl of 20 micromolar indexing primer 7, and 10\\xa0µl of rhAMPseq master mix v8.1—a total of 40\\xa0µl. After initial \\ndenaturation of 3\\xa0min at 95\\xa0°C, the PCR reaction proceeded with 50 cycles of 15\\xa0s at 95\\xa0°C, 30\\xa0s at 60\\xa0°C, and 30\\xa0s \\nat 72\\xa0°C, followed by final elongation of 1\\xa0min at 72\\xa0°C and hold at 4\\xa0°C. The PCR reactions were purified with \\nAmpure beads at 0.8X ratio (32\\xa0µl beads per 40\\xa0µl of PCR reaction) as outlined above, and eluted in 22\\xa0µl IDTE \\nbuffer. The concentration and the average size of the eluted product were determined by Qubit High Sensitivity \\nDNA kit and Agilent 2200 TapeStation system with D1000 high-sensitivity screen tape respectively. The eluted \\nproduct was diluted to 4\\xa0nM concentration, and used as an input for denatured sequencing library preparation, \\nper manufacturer instructions. The sequencing was performed on Illumina Miseq apparatus (V2 chemistry, \\n2\\u2009×\\u2009150\\xa0bp reads) using 6 picomolar denatured library supplemented with 40% PhiX sequencing control.\\n',\n",
       " 'Decoding and analysis': 'Decoding and analysis\\nThis section outlines the key steps involved in our sequencing analysis pipeline, aimed at effectively processing \\nand interpreting sequenced reads. The analysis pipeline gets the sequencing output file containing raw reads in \\n“.fastq” format and a design file containing the combinatorial sequences.\\nAnalysis steps:\\n1.\\t Length filtering. We saved reads that were 220\\xa0bp in length, retaining only those corresponding to our \\ndesigned read length.\\n2.\\t Read retrieval. We carefully checked each read for the presence of BCs, universals, and payloads. To keep \\nour data accurate, we discarded reads where the BCs, universals, or payloads had a Hamming distance of \\nmore than 3 errors.\\n3.\\t Identifying inferred k-mers. For every BC and each cycle, we counted the K most common k-mers. We \\nthen compared these with the design file to quantify those matching (Fig.\\xa05b) (See\\xa0Table\\xa03).\\n',\n",
       " 'Information capacities for selected encodings': 'Information capacities for selected encodings\\nTable\\xa01 illustrates the logical densities derived from encoding a 1\\xa0GB binary message using oligonucleotides with \\na 12nt barcode and an additional 4nt for standard DNA RS error correction, and a 120 letters payload with 14 \\nextra RS for the payload in combinatorial encoding schemes with parameters N and K.\\nThe densities were calculated as follows:\\n(18)\\nBits per Letter =\\n\\x1f\\nlog2\\n\\x1e\\x1e\\nN\\nk\\n\\x1d\\x1d\\x1c\\n(19)\\nAlphabet Size = 2Bits per Letter\\n(20)\\nBits per Sequence = Bits per Letter × Payload length.\\n(21)\\nNumber of Sequences = ⌈Message Size in bits\\nBits per Sequence ⌉\\n(22)\\nNumber of sequences padded : Total number of sequences after padding for the block size\\n(23)\\nPadding is = Block Size −\\n\\x1f\\nNumber of Sequences%Block Size\\n\\x1e\\n(24)\\nNumber of sequences with RS = Number of sequences padded\\nBlock Size\\n× Block Size AfterRS\\nTable\\u202f3.\\u2002 \\u2009Summary of sequencing reads analyzed in the study. The table shows the total number of reads \\nobtained, the number filtered by length (220 bases) for analysis, and the counts of reads associated with BC1, \\nBC2, and those that did not have any recognizable barcode (No BC).\\nReads\\nCount\\nTotal PF reads\\n2,634,683\\nReads of length 220\\n2,139,071\\nBC1\\n1,365,295\\nBC2\\n768,755\\nNo BC\\n5021\\n16\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       " 'Ethics declaration': 'Ethics declaration\\nNo animal or human subjects were involved in the study.\\nData availability\\nThe raw data is available in ENA (European Nucleotide Archive). The datasets generated and/or analyzed \\nduring the current\\xa0study are available in the ENA (European Nucleotide Archive) repository, Accession \\nNumber—ERR12364864.\\nCode Availability\\nImplementation of the algorithms and instructions on how to use them can be found in the GitHub repository\\xa0in \\nthe following links: https://\\u200bgithub.\\u200bcom/\\u200bInbal\\u200bPreuss/\\u200bdna_\\u200bstora\\u200bge_\\u200bshort\\u200bmer_\\u200bsimul\\u200bation,\\xa0https://\\u200bgithub.\\u200bcom/\\u200bInbal\\u200b\\nPreuss/\\u200bdna_\\u200bstora\\u200bge_\\u200bexper\\u200biment.\\nReceived: 11 December 2023; Accepted: 28 March 2024\\n',\n",
       " 'References': 'References\\n\\t 1.\\t Church, G., Gao, Y. & Kosuri, S. Next-generation digital information storage in DNA. Science 337, 1628 (2012).\\n\\t 2.\\t Goldman, N. et al. Towards practical, high-capacity, low-maintenance information storage in synthesized DNA. Nature 494, 77–80 \\n(2013).\\n\\t 3.\\t Anavy, L., Vaknin, I., Atar, O., Amit, R. & Yakhini, Z. Data storage in DNA with fewer synthesis cycles using composite DNA \\nletters. Nat. Biotechnol. 37, 1229–1236 (2019).\\n\\t 4.\\t Erlich, Y. & Zielinski, D. DNA Fountain enables a robust and efficient storage architecture. Science 355, 950–954 (2017).\\n\\t 5.\\t Gabrys, R., Kiah, H., & Milenkovic, O. Asymmetric lee distance codes for DNA-based storage. In 2015 IEEE International \\nSymposium on Information Theory (ISIT) (2015).\\n\\t 6.\\t NallappaBhavithran, G., & Selvakumar, R. Indel Error Correction Codes for DNA Digital Data Storage and Retrieval. ArXiv \\nabs/2302.1467 (2023).\\n\\t 7.\\t Wang, C. et al. Mainstream encoding–decoding methods of DNA data. CCF Trans. High Perform. Comput. 4, 23–22 (2022).\\n\\t 8.\\t Boruchvosky, A., Bar-Lev, D., & Yaakobi, E. DNA-Correcting Codes: End-to-end Correction in DNA Storage Systems. ArXiv, \\nabs/2304.0391 (2023).\\n\\t 9.\\t Bornholt, J. et al. Toward a DNA-based archival storage system. IEEE Micro 37, 98–104 (2017).\\n\\t10.\\t Yazdi, S., Yuan, Y., Ma, J., Zhao, H. & Milenkovic, O. A rewritable, random-access DNA-based storage system. Sci. Rep. 5, 1–10 \\n(2015).\\n\\t11.\\t Organick, L. et al. Random access in large-scale DNA data storage. Biotechnol. 36, 242–248 (2018).\\n\\t12.\\t Choi, Y. et al. High information capacity DNA-based data storage with augmented encoding characters using degenerate bases. \\nSci. Rep. 9, 6582 (2019).\\n\\t13.\\t Roquet, N., Bhatia, S., Flickinger, S., Mihm, S., Norsworthy, M., Leake, D., & Park, H. DNA-based data storage via combinatorial \\nassembly. 20 April 2021 (online). https://\\u200bdoi.\\u200borg/\\u200b10.\\u200b1101/\\u200b2021.\\u200b04.\\u200b20.\\u200b44019\\u200b4v1.\\n\\t14.\\t Yan, Y., Pinnamaneni, N., Chalapati, S., Crosbie, C. & Appuswamy, R. Scaling logical density of DNA storage with enzymatically-\\nligated composite motifs. Sci. Rep. 13, 15978 (2023).\\n\\t15.\\t LeProust, E. et al. Synthesis of high-quality libraries of long (105mer) oligonucleotides by a nover depurination controlled process. \\nNucl. Acids Res. 38, 2522–2540 (2019).\\n\\t16.\\t Barrett, M. et al. Comparative genomic hybridization using oligonucleotide microarrays and total genomic DNA. Proc. Natl Acad. \\nSci. USA 101, 17765–17770 (2004).\\n\\t17.\\t Eleuteri, A., Capaldi, D., Douglas, L. & Ravikumar, V. Oligodeoxyribonucleotide phosphorothioates: Substantial reduction of \\n(N-1)-mer content through the use of trimeric phosphoramidite synthons. Nucleosides Nucleotides 3, 475–483 (1999).\\n\\t18.\\t Yagodkin, A. et al. Improved synthesis of trinucleotide phosphoramidites and generation of randomized oligonucleotide libraries. \\nNucleosides Nucleotides Nucl. Acids 26(5), 473–497 (2007).\\n\\t19.\\t Randolph, J., Yagodkin, A. & Mackie, H. Codon-based Mutagenesis. Nucl. Acids Symp. Ser. 52, 479 (2008).\\n\\t20.\\t Ferrante, M., & Saltalamacchia, M. The Coupon Collector’s Problem, p 35 (2014).\\n\\t21.\\t Press, W. et al. HEDGES error-correcting code for DNA storage corrects indels and allows sequence constraints. Proc. Natl. Acad. \\nSci. 117(31), 18489–18496 (2020).\\n\\t22.\\t Haoling, Z., et al. SPIDER-WEB generates coding algorithms with superior error tolerance and real-time information retrieval \\ncapacity. arXiv preprint arXiv 2204.02855 (2022).\\n\\t23.\\t Sabary, O., Orlev, Y., Shafir, R. & Anavy, L. SOLQC: Synthetic oligo library quality control tool. Bioinformatics 2, 740 (2020).\\n\\t24.\\t Preuss, I., Galili, B., Yakhini, Z., & Anavy, Z. Sequencing coverage analysis for combinatorial DNA-based storage systems. biorxiv \\n(2024).\\n\\t25.\\t Ayoub, R. Euler and the zeta function. Am. Math. Mon. 81, 1067–1086 (1974).\\n\\t26.\\t Reed, I. & Solomon, G. Polynomial codes over certain finite fields. J. Soc. Ind. Appl. Math. 8, 300–304 (1960).\\n\\t27.\\t Hipp, R. D. SQLite (2020) (Online). https://\\u200bwww.\\u200bsqlite.\\u200borg/\\u200bindex.\\u200bhtml.\\n\\t28.\\t Levenshtein, V. Binary codes capable of correcting spurious insertions and deletions of ones. Problems Inf. Transm. 1, 8–17 (1965).\\n\\t29.\\t Levenshtein, V. Binary codes capable of correcting deletion, insertions and reversals. Soviet Physics Doklady 10(8), 707–710 (1966).\\n(25)\\nSynthesis Cycles = Number of sequences with RS × Full sequence length with RS\\n(26)\\nBits per Synthesis Cycle = Message Size in bits\\nSynthesis Cycle\\n(27)\\nFold Increase =\\nBits per Synthesis Cycle\\nBits per Synthesis Cycle of Standard Scheme\\n17\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       " 'Acknowledgements': 'Acknowledgements\\nThis project has received funding from the European Union’s Horizon Europe research and innovation \\nprogramme under grant\\xa0agreement No. 101115134 (DiDAX project). The authors of this paper thank the Yakhini \\nresearch group for the fruitful\\xa0discussions. The authors also thank Eland Nagar for his support and problem-\\nsolving approach regarding the experimental\\xa0proof of concept.\\nAuthor contributions\\nI.P., Z.Y. and L.A conceived the idea, designed the experiments, interpreted and analyzed the data. I.P performed \\nall computational work, including the simulations and the data analysis. M.R. was solely responsible for \\nconducting the experimental work. All authors wrote the manuscript.\\nFunding\\nEuropean Union’s Horizon Europe Research and Innovation Programme, 101115134.\\nCompeting interests\\u2002\\nZohar Yakhini, Leon Anavy have competing interests as defined by Nature Research. Z. Yakhini and L. Anavy are \\nnamed as inventors on a patent related to the content of this paper: L. Anavy, Z. Yakhini, and R. Amit, \"Molecular \\ndata storage systems and methods\". United States of America Patent US20210141568A1, 2021.\\nAdditional information\\nSupplementary Information The online version contains supplementary material available at https://\\u200bdoi.\\u200borg/\\u200b\\n10.\\u200b1038/\\u200bs41598-\\u200b024-\\u200b58386-z.\\nCorrespondence and requests for materials should be addressed to I.P.\\nReprints and permissions information is available at www.nature.com/reprints.\\nPublisher’s note\\u2002 Springer Nature remains neutral with regard to jurisdictional claims in published maps and \\ninstitutional affiliations.\\nOpen Access\\u2002  This article is licensed under a Creative Commons Attribution 4.0 International \\nLicense, which permits use, sharing, adaptation, distribution and reproduction in any medium or \\nformat, as long as you give appropriate credit to the original author(s) and the source, provide a link to the \\nCreative Commons licence, and indicate if changes were made. The images or other third party material in this \\narticle are included in the article’s Creative Commons licence, unless indicated otherwise in a credit line to the \\nmaterial. If material is not included in the article’s Creative Commons licence and your intended use is not \\npermitted by statutory regulation or exceeds the permitted use, you will need to obtain permission directly from \\nthe copyright holder. To view a copy of this licence, visit http://\\u200bcreat\\u200biveco\\u200bmmons.\\u200borg/\\u200blicen\\u200bses/\\u200bby/4.\\u200b0/.\\n© The Author(s) 2024\\n'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seg_paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "78e7c509",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'prev': '1\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  \\n| https://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports\\nEfficient DNA‑based data storage \\nusing shortmer combinatorial \\nencoding\\nInbal\\xa0Preuss\\u200a1,3*, Michael\\xa0Rosenberg\\u200a2, Zohar\\xa0Yakhini\\u200a1,3\\xa0& Leon\\xa0Anavy\\u200a1,3\\nData storage in DNA has recently emerged as a promising archival solution,\\xa0offering space-efficient \\nand long-lasting digital storage solutions. Recent studies suggest leveraging the inherent redundancy \\nof synthesis and sequencing technologies by using composite DNA alphabets. A major challenge \\nof this approach involves the noisy inference process, obstructing\\xa0large composite alphabets. This \\npaper introduces a novel approach for DNA-based data storage, offering, in some implementations, \\na 6.5-fold increase in logical density over standard DNA-based storage systems, with near-zero \\nreconstruction error. Combinatorial DNA encoding uses a set of clearly distinguishable DNA shortmers \\nto construct large combinatorial alphabets, where each letter consists of a subset of shortmers. We \\nformally define various combinatorial encoding schemes and investigate their theoretical properties. \\nThese include information density and reconstruction probabilities, as well as required synthesis \\nand sequencing multiplicities. We then propose an end-to-end design for a combinatorial DNA-\\nbased data storage system, including encoding schemes, two-dimensional (2D) error correction \\ncodes, and reconstruction algorithms, under different error regimes. We performed simulations and \\nshow, for example, that\\xa0the use of 2D Reed-Solomon error correction has significantly improved \\nreconstruction rates. We validated\\xa0our approach by constructing two combinatorial sequences using \\nGibson assembly, imitating a 4-cycle combinatorial synthesis process. We confirmed the successful \\nreconstruction, and established the robustness of our approach for different error types. Subsampling \\nexperiments supported the important role of sampling rate and its effect on the overall performance. \\nOur work demonstrates the potential of combinatorial shortmer encoding for DNA-based data storage \\nand describes some theoretical research questions and technical challenges. Combining combinatorial \\nprinciples with error-correcting strategies, and investing in the development of DNA synthesis \\ntechnologies that efficiently support combinatorial synthesis, can pave the way to efficient, error-\\nresilient DNA-based storage solutions.\\nDNA is a promising media storage candidate for long-term data archiving, due to its high information density, \\nlong-term stability, and robustness. In recent years, several studies have demonstrated the use of synthetic DNA \\nfor storing digital information on a megabyte scale, exceeding the physical density of current magnetic tape-\\nbased systems by roughly six orders of \\xadmagnitude1,2. Physical density is one of several quantitative metrics for \\nevaluating the efficiency of DNA-based storage systems, measured by the data unit per gram of DNA. Another \\nperformance metric, which was introduced \\xadin3, is called logical density, refering to the amount of data encoded in \\neach synthesis cycle. Since DNA synthesis is the main cost component in DNA-based storage systems, increasing \\nthe logical density is the main focus of this work.\\nResearch efforts in the field of DNA-based storage systems have mainly focused on the application of various \\nencoding schemes, while relying on standard DNA synthesis and sequencing technologies. These include the \\ndevelopment of error-correcting codes for the unique information channel of DNA-based data \\xadstorage4–8. Random \\naccess capabilities for reading specific information stored in DNA also require advanced coding \\xadschemes9–11. Yet, \\ndespite the enormous benefits potentially associated with capacity, robustness, and size, existing DNA-based \\nstorage technologies are characterized by inherent information redundancy. This is due to the nature of DNA \\nsynthesis and sequencing methodologies, which process multiple molecules that represent the same information \\nbits in parallel. Recent studies suggest exploiting this redundancy to increase the logical density of the system, by \\nextending the standard DNA alphabet using composite letters (also referred to as degenerate bases), and thereby \\nOPEN\\n1School of Computer Science, Reichman University, 4610101\\xa0Herzliya, Israel. 2Institute of Nanotechnology and \\nAdvanced Materials, The Mina and Everard Goodman Faculty of Life Sciences, Bar-Ilan University, 5290002\\xa0Ramat \\nGan, Israel. 3Faculty of Computer Science, Technion, 3200003\\xa0Haifa, Israel. *email: inbalpreuss@gmail.com',\n",
       "  'Efficient DNA-based data storage using shortmer combinatorial encoding': '\\n'},\n",
       " {'prev': '2\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\nencoding more than 2 bits per \\xadletter12,13. In this approach, a composite DNA letter uses all four DNA bases (A, C, \\nG, and T), combined or mixed in a specified predetermined ratio σ = (σA, σC, σG, σT)\\u200a. A resolution parameter \\nk = σA + σC + σG + σT is defined, for controlling the alphabet size. The full composite alphabet of resolution \\nk\\u200a, denoted \\x1fk\\u200a, is the set of all composite letters, so that \\x1fi∈(A,C,G,T}σi = k\\u200a. Writing a composite letter is done by \\nusing a mixture of the DNA bases, determined by the letter’s ratio in the DNA synthesis cycle. Current synthesis \\ntechnologies produce multiple copies, and by using the predetermined base mixture each copy will contain a \\ndifferent base, thus preserving the ratio of the bases at the sequence-population level.\\nWhile the use of numerical ratios supports higher logical density in composite synthesis, it also introduces \\nchallenges related to the synthesis and inference of exact ratios. Combinatorial approaches, which also consist \\nof mixtures, address these challenges in a different way. Studies by Roquet et\\xa0al. (2021) and Yan et\\xa0al. (2023) \\ncontribute significantly to advancing DNA-based data storage technology. To encode and store data, Roquet et\\xa0al. \\nfocus on a novel combinatorial assembly method for DNA. Yan et\\xa0al. extend the frontiers of this technology by \\nenhancing the logical density of DNA storage, using enzymatically-ligated composite \\xadmotifs13,14.\\nIn this paper, we present a novel approach for encoding information in DNA, using combinatorial encoding \\nand shortmer DNA synthesis, leading to an efficient sequence design and improved DNA synthesis and readout \\ninterpretation. The method described herein leverages the advantages of combinatorial encoding schemes while \\nrelying on existing DNA chemical synthesis methods with some modifications. Using shortmer DNA synthesis \\nalso minimizes the effect of synthesis and sequencing errors. We formally define shortmer-based combinatorial \\nencoding schemes, explore different designs, and analyze their performance. We use computer-based simulations \\nof an end-to-end DNA-based data storage system built on combinatorial shortmer encodings, and study its \\nperformance. To demonstrate the potential of our suggested approach and experimentally test its validity, we \\nperformed an assembly-based molecular implementation of the proposed combinatorial encoding scheme and \\nanalyzed the resulting data. Finally, we discuss the potential of combinatorial encoding schemes and the future \\nwork required to enable these schemes in large-scale DNA-based data storage systems and other DNA data \\napplications.\\n',\n",
       "  'Results': 'Results\\n',\n",
       "  'Design of shortmer combinatorial encoding for DNA storage': 'Design of shortmer combinatorial encoding for DNA storage\\nWe suggest a novel method to extend the DNA alphabet while ensuring near-zero error rates.\\nLet \\x1f be a set of DNA k-mers that will serve as building blocks for our encoding scheme. Denote the elements \\nin \\x1f as X1, . . . , XN\\u200a. Elements in \\x1f are designed to be sufficiently different from each other, to minimize mix-up \\nerror probability. Formally, the set is designed to satisfy d\\n\\x1f\\nXi, Xj\\n\\x1e\\n≥d; ∀i \\x1c= j\\u200a, with the minimal Hamming \\ndistance d serving as a tunable parameter.\\nOther design criteria can be applied to the shortmers in \\x1f\\u200a, taking into consideration the properties of DNA \\nsynthesis, manipulation, and sequencing. These may include minimal Levenshtein distance, GC context, and \\navoiding long homopolymers. Clearly, any such filtering process will result in reduced alphabet size and reduced \\nlogical density.\\nNote that N = |\\x1f| ≤4k\\u200a. The elements in \\x1f will be used as building blocks for combinatorial DNA synthesis \\nin a method similar to the one used for composite DNA \\xadsynthesis3. Examples of k-mer sets \\x1f are presented in \\nSupplementary Sect.\\xa08.3.\\nWe define a combinatorial alphabet \\x1f over \\x1f as follows. Each letter in the alphabet represents a non-empty \\nsubset of the elements in \\x1f\\u200a. Formally, a letter σ ∈\\x1e\\u200a, representing a subset S ⊆\\x1f/∅\\u200a, can be written as an \\nN-dimensional binary vector where the indices for which σi = 1 represents the k-mers from \\x1f included in the \\nsubset S. We denote the k-mers in S as member k-mers of the letter σ\\u200a. For example, σ = (0, 1, 0, 1, 1, 0) represents \\nS = {X2, X4, X5} and |\\x1f| = N = 6\\u200a. Figure\\xa01a,b illustrate an example of a combinatorial alphabet using N = 16\\u200a, \\nin which every letter represents a subset of size 5 of \\x1f\\u200a. In\\xa0Sect.\\xa0“Binary and binomial combinatorial alphabets” \\nincludes a description of the construction of different combinatorial alphabets.\\nTo write a combinatorial letter σ in a specific position, a mixture of the member k-mers of σ is synthesized. To \\ninfer a combinatorial letter σ\\u200a, a set of reads needs to be analyzed to determine which k-mers are observed in the \\nanalyzed position (See\\xa0Sects. “Binary and binomial combinatorial alphabets” and “Reconstruction probabilities \\nfor binomial encoding” for more details). This set of k-mers observed in the sequencing readout and used for \\ninferring σ is referred to as inferred member k-mers. While the synthesis output and the sequencing readout will \\ninclude different counts for the member k-mers, the determination of the set of inferred k-mers will force binary \\nassignment for each k-mer to fit into the design scheme of combinatorial encoding.\\nFrom a hardware/chemistry perspective, the combinatorial shortmer encoding scheme can potentially be \\nbased on using the standard phosphoramidite chemistry synthesis technology, with some alterations (See Fig.\\xa01b \\nand Supplementary Sect.\\xa08.1)15,16. First, DNA k-mers should be used as building blocks for the \\xadsynthesis17. Such \\nreagents are commercially available for DNA trimers and were used, for example, for the synthesis of codon \\noptimization DNA \\xadlibraries18,19. In addition, a mixing step should be added to each cycle of the DNA synthesis \\nto allow mixing of the member k-mers prior to their introduction to the elongating molecules. Such systems are \\nyet to be developed and current attempts for combinatorial DNA synthesis are based on enzymatic assembly of \\nlonger DNA \\xadfragments13,14.\\nSimilar to composite DNA encoding, combinatorial encoding requires the barcoding of the sequences using \\nunique barcodes composed of standard DNA barcodes. This design enables direct grouping of reads pertaining \\nto the same combinatorial sequence. These groups of reads are the input for the process of reconstructing the \\ncombinatorial letters.\\nThe extended combinatorial alphabets allow for a higher logical density of the DNA-based storage system, \\nwhile the binary nature of the alphabet minimizes error rates.\\n'},\n",
       " {'prev': '3\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       "  'Binary and binomial combinatorial alphabets': 'Binary and binomial combinatorial alphabets\\nThe main parameter that defines a combinatorial encoding scheme is the alphabet \\x1f\\u200a. More specifically, it is the \\nset of valid subsets of \\x1f that can be used as letters. We define two general approaches for the construction of \\x1f\\u200a. \\nNamely, the binomial encoding and the full binary encoding.\\nIn the binomial encoding scheme, only subsets of \\x1f of size exactly K represent valid letters in \\x1f\\u200a, so that every \\nletter σ ∈\\x1e consists of exactly K member k-mers. Therefore, all the letters in the alphabet have the same \\nHamming weight K\\u200a. w(σ) = K, ∀σ ∈\\x1c\\u200a. This yields an effective alphabet of size |\\x1f| =\\n\\x1f\\nN\\nK\\n\\x1e\\n letters, where each \\ncombinatorial letter encodes log2(|\\x1e|) = log2\\n\\x1f\\nN\\nK\\n\\x1e\\n bits. An r-bit binary message requires \\nr\\nlog2\\n\\x1f N\\nK\\n\\x1e synthesis \\ncycles (and a DNA molecular segment with length \\nkr\\nlog2\\n\\x1f N\\nK\\n\\x1e ). In practice, we would prefer working with alphabet \\nsizes that are powers of two, where each letter will encode for \\n\\x1f\\nlog2\\n\\x1e\\nN\\nK\\n\\x1d\\x1c\\n bits. Note that this calculation ignores \\nerror correction redundancy, random access primers, and barcodes, which are all required for message \\nreconstruction. See Supplementary Sect.\\xa08.2 and Fig.\\xa01a, which illustrate a trimer-based binomial alphabet with \\nN = 16 and K = 5\\u200a, resulting in an alphabet of size |\\x1f| =\\n\\x1f\\n16\\n5\\n\\x1e\\n= 4, 368 that allows to encode ⌊log2(4368)⌋= 12 \\nbits per letter or synthesis position.\\nIn the full binary encoding scheme, all possible nonempty subsets of \\x1f represent valid letters in the alphabet. \\nThis yields an effective alphabet of size |\\x1f| = 2N −1 letters, each encoding for \\n\\x1f\\nlog2 (|\\x1e|)\\n\\x1e\\n= N −1 bits.\\nFrom this point on, we focus on the binomial encoding.\\nFigure\\xa01.\\u2002 \\u2009Our combinatorial encoding and synthesis approach. (a) Schematic view of a combinatorial \\nalphabet (Encode legend). A set of 16 trimers, X1, . . . , X16\\u200a, is used to construct 4096 combinatorial letters, \\neach representing a subset of 5 trimers as indicated on the right and depicted in the grayed-out cells of the \\ntable. (b) A suggested approach for combinatorial shortmer synthesis. A modified synthesizer would include \\ndesignated containers for the 16-trimer building blocks and a mixing chamber. Standard DNA synthesis is used \\nfor the barcode sequence (1), while the combinatorial synthesis proceeds as follows: The trimers included in \\nthe synthesized combinatorial letter are injected into the mixing chamber and introduced into the elongating \\nmolecules (2). The process repeats for the next combinatorial letter (3), and finally, the resulting molecules are \\ncleaved and collected (4).\\n'},\n",
       " {'prev': '4\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       "  'Reconstruction probabilities for binomial encoding': 'Reconstruction probabilities for binomial encoding\\nIn this section, the performance characteristics of binomial encoding are investigated. Specifically, we present a \\nmathematical analysis of the probability of successfully reconstructing the intended message. In Sects.\\xa0“An end-\\nto-end combinatorial shortmer storage system” and \"Experimental proof of concept\", results are presented from \\nour simulations and a small-scale molecular implementation of the binomial encoding, respectively.\\n',\n",
       "  'Reconstruction of a single combinatorial letter': 'Reconstruction of a single combinatorial letter\\nSince every letter σ ∈\\x1e consists exactly of the K member k-mers, the required number of reads for observing at \\nleast one read of each member k-mer in a single letter follows the coupon collector \\xaddistribution20. The number \\nof reads required to achieve this goal can be described as a random variable R = \\x1fK\\ni=1Ri\\u200a, where R1 = 1 and \\nRi ∼Geom\\n\\x1f K−i+1\\nK\\n\\x1e\\n, i = 2, . . . , K\\u200a. Hence, the expected number of required reads, is:\\nwhere Har(·) is the harmonic number.\\nThe expected number of reads required for reconstructing a single combinatorial letter thus remains \\nreasonable for the relevant values of K\\u200a. For example, when using a binomial encoding with K = 5 the expected \\nnumber of reads required for reconstructing a single combinatorial letter is roughly 11.5\\u200a, which is very close to \\nthe experimental results presented in Sect. \"Experimental proof of concept\".\\nBy Chebyshev’s inequality (See\\xa0Sect.\\xa0\"Reconstruction probability of a binomial encoding letter\"), we can \\nderive a (loose) upper bound on the probability of requiring more than E[R] + cK reads to observe at least one \\nread of each member k-mer, where c > 1 is a parameter:\\nFor example, when using a binomial encoding with K = 5\\u200a, the probability of requiring more than 26.5 reads \\n(corresponding to c = 3\\u200a) is bounded by 0.18\\u200a, which is consistent with the experimental result shown in Fig.\\xa05d.\\n',\n",
       "  'Reconstruction of a combinatorial sequence': 'Reconstruction of a combinatorial sequence\\nWhen we examine an entire K-subset binomial encoded combinatorial sequence of length l\\u200a, we denote by R(l) \\nthe required number of reads to observe K distinct k-mers in every position. Assuming independence between \\ndifferent positions and not taking errors into account, we get the following relationship between c and any desired \\nconfidence level 1 −δ (See\\xa0Sect.\\xa0\"Reconstruction probability of a binomial encoding letter\" for details):\\nAnd therefore:\\nThe number of reads required to guarantee reconstruction of a binomial encoded message, at a 1 −δ \\nprobability, with K = 5, and l synthesized positions, is thus KHar(K) + cK when\\nSupplementary Table\\xa0S2 shows several examples of this upper bound. As demonstrated in the simulations \\nand the experimental results, this bound is not tight (See Sects.\\xa0“An end-to-end combinatorial shortmer storage \\nsystem” and \"Experimental proof of concept\").\\nNote that with an online sequencing technology (such as nanopore sequencing) the sequencing reaction can \\nbe stopped after K distinct k-mers are confidently observed.\\nTo take into account the probability of observing a k-mer that is not included in \\x1f (e.g., due to synthesis or \\nsequencing error), we can require that at least t > 1 reads of each of the K distinct k-mers will be observed. This \\nis experimentally examined in\\xa0Sect. \"Experimental proof of concept\", while the formal derivation of the number \\nof required reads is not as trivial, and will be addressed in future work.\\nThe above analysis is based only on oligo recovery, which depends solely on the sampling rate, ignoring \\npossible mix-up errors (i.e., incorrect k-mer readings). This assumption is based on the near-zero mix-up \\nprobability attained by the construction of \\x1f\\u200a, which maximizes the minimal Hamming distance between elements \\nin \\x1f\\u200a. In\\xa0Sect. \"Experimental proof of concept\", this analysis is compared to experimental results obtained from \\nusing synthetic combinatorial DNA.\\nAn end‑to‑end combinatorial shortmer storage system\\nWe suggest a complete end-to-end workflow for DNA-based data storage with the combinatorial shortmer \\nencoding presented in Fig.\\xa02. The workflow begins with encoding, followed by DNA synthesis, storage, and \\nsequencing, and culminates in a final decoding step. A 2D Reed-Solomon (RS) error correction scheme, which \\ncorrects errors in the letter reconstruction (for example, due to synthesis, sequencing, and sampling errors) and \\n(1)\\nE[R] =\\n\\x1fK\\ni=1E[Ri] = K\\n\\x1fK\\ni=1\\n1\\ni = KHar(K)\\n(2)\\nP(|R −KHar(K)| ≥cK) ≤π2\\n6c2\\n(3)\\nP(|R(l) −KHar(K)| ≥cK) ≤1 −\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n< δ\\n(4)\\nP(R(l) < KHar(K) + cK) ≥\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n≥1 −δ\\n(5)\\nc ≥\\n\\x1f\\n1/6π\\n\\x1e\\n1 −(1 −δ)1/l\\x1d−1/2',\n",
       "  'An end-to-end combinatorial shortmer storage system': '\\n'},\n",
       " {},\n",
       " {},\n",
       " {'prev': '7\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       "  'Experimental proof of concept': 'Experimental proof of concept\\nTo assess and establish the potential of large combinatorial alphabets, we performed a small-scale experimental \\nproof of concept study demonstrating the encoding and decoding of a 96-bit input message, which is equivalent \\nto the text “DNA Storage!”. Since combinatorial DNA synthesis technology is not yet available, we demonstrated \\nthe combinatorial approach using Gibson assembly as an ad-hoc imitation for combinatorial synthesis. We \\nconstructed two combinatorial sequences, each containing a barcode and four payload cycles over a binomial \\nalphabet with N = 16 and K = 5\\u200a. The assembly was performed using DNA fragments composed of a 20-mer \\ninformation sequence and an overlap of 20\\xa0bp between adjacent fragments, as shown in Fig.\\xa05a. The assembled \\nDNA was then stored and sequenced for analysis using Illumina Miseq (See Table\\xa03 and Sect.\\xa0\"Cost analysis\" for \\ndetails about the sequencing procedures).\\nThe sequencing output was then analyzed using the procedure described in\\xa0Sect.\\xa0“Decoding and analysis”. \\nBoth combinatorial sequences were successfully reconstructed from the sequencing reads, as presented in Fig.\\xa05b, \\nand\\xa0Supplementary Figs. S1, S2, and S3. The experiment also demonstrated the robustness of the binomial DNA \\nencoding for synthesis and sequencing errors, as described in Fig.\\xa05c. We observed a minor leakage between the \\ntwo synthesized sequences, which was overcome by the reconstruction pipeline (See Fig.\\xa05c, and\\xa0Supplementary \\nFigs.\\xa0S1, S2, and S3). Note that there is a partial overlap between the member k-mers of the two sequences.\\nFor comparison, a recent study \\xadby14 encoded the 84-bit phrase “HelloWord” using\\xa0a different encoding \\nand\\xa0synthesis approach. A comparison between the two experiments is shown in Table\\xa02. For example, while we \\nused Gibson assembly as our synthesis method, they introduced a new method called Bridge Oligonucleotide \\nAssembly. We encoded 12 bits per synthesis cycle and assembled four combinatorial fragments in each \\nsequence, \\xadwhile14 encoded 84 bits in a single combinatorial cycle. Our 96-bit were split and encoded using two \\ncombinatorial sequences, while they encoded the same 84-bits message, in its full format, on eight different \\na\\nb\\nc\\nd\\nText\\nPayload\\nPayload\\nPayload\\nRS\\nRS\\nRS\\nRS\\niv\\ni\\nii\\niii\\nText\\nPayload\\nPayload\\nPayload\\nRS\\nRS\\nRS\\nRS\\n1\\n2\\n3\\n8\\n7\\n6\\n5\\n4\\nFigure\\xa03.\\u2002 \\u2009Simulation of an end-to-end combinatorial shortmer encoding. (a) A schematic view of the \\nsimulation workflow. A text message is translated into a combinatorial message (1), and encoded using RS \\nerror correction on the barcode and payload (2). Each block is encoded using outer RS error correction (3). \\nDNA synthesis and sequencing are simulated under various error schemes, and the combinatorial letters are \\nreconstructed (4–5). RS decoding is performed on each block (6) and on each sequence (7) before translation \\nback to text (8). The Roman numerals (i-iv) represent the different error calculations. (b) Error rates in different \\nstages of the decoding process. Boxplot of the normalized Levenshtein distance (See Sect.\\xa0\"Reconstruction\") \\nfor the different stages in a simulation (30 runs) of sampling 100 reads, with an insertion error rate of 0.01. The \\nX-axis represents the stages of error correction (before 2D RS decoding (iv), after RS payload decoding (iii), and \\nafter 2D RS decoding (ii)). (c,d) Sampling rate effect on overall performance. Normalized Levenshtein distance \\nas a function of sampling rate before RS decoding (c) and after 2d RS decoding (ii). Different lines represent \\ndifferent error types (substitution, deletion, and insertion) introduced at a rate of 0.01.\\n'},\n",
       " {},\n",
       " {'prev': '9\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n140 reads to obtain 1 −δ = 0.99 probability of reconstruction. Clearly, this is well above the observed number \\nof 55.60 reads. Note, as explained, the calculated bound is a loose bound.\\nThe reconstruction procedure ends with a set of inferred k-mers that represent the inferred combinatorial \\nletter. This set is not guaranteed to be correct, especially when using t = 1\\u200a, which means that noisy reads may \\nresult in an incorrect k-mer included in the inferred letter. Figure\\xa05f depicts the rate of incorrect reconstructions \\nas a function of the number of required copies for each inferred k-mer (\\u200at = 1, 2, 3, 4\\u200a). Note that with t ≥3 results \\nin 100% successful reconstruction. This, however, comes with a price, where more reads must be analyzed.\\n',\n",
       "  'Discussion': 'Discussion\\nIn this study, we introduced combinatorial shortmer encoding for DNA-based data storage, which extends the \\napproach of composite DNA by resolving some of its sensitivity related issues. Combinatorial shortmer encoding \\nallows for increased logical density while maintaining low error rates and high reconstruction rates. We explored \\ntwo encoding schemes, binary and binomial, and evaluated some of their theoretical and practical characteristics. \\nThe inherent consistency of the binomial encoding scheme, where every letter in the sequence consists of exactly \\nK distinct member k-mers, ensures uniformity in the encoded DNA sequences. This approach not only simplifies \\nthe reading process, but also allows for a more streamlined decoding. For instance, technologies like nanopore \\nsequencing enable continuous sequencing until all k-mers at a given position are confidently observed.\\nOur suggested approach is designed to inherently overcome base substitution errors, which are the most \\ncommon errors expected in every DNA-based data storage system that includes DNA sequencing. This is \\nachieved by the selection of a set of N k-mer building blocks to be resilient to single-base substitutions. Other \\nconsiderations may also be incorporated in the selection of the set of valid k-mers, taking into account any \\nbiological, chemical, or technological constraints. This represents an inherent tradeoff in DNA-based data\\xa0storage \\nbetween sequence constraints and information density. Insertion and deletion errors, which usually originate \\nin the synthesis process, are more challenging to overcome. We introduced a 2D RS error correction scheme \\non the shortmer level, allowing for a successful message reconstruction even with error levels exceeding those \\nexpected in reality.\\nOur study highlights the significant effect of sampling rates on the overall performance of the system. The \\naccuracy and completeness of sequence reconstruction require each of the sequences to be observed with a \\nsufficiently high coverage. Our subsampling experiments underpin this observation, demonstrating the need \\nfor calibration of sampling rates to ensure the desired fidelity in DNA-based data storage. The crucial role of \\nthe sampling rate was also highlighted \\xadin3. However, while composite DNA uses mixed letter with varying \\nproportions of the different letters, the combinatorial encoding, studied in this current work, uses binary mixtures \\nand does not rely on proportions. This potentially allows scaling up the combinatorial encoding without a \\nsignificant effects on the required sampling rates.\\nCombinatorial DNA coding can potentially reduce the overall costs of DNA-based data storage. Considering \\nboth sequencing costs, which fluctuate, and synthesis costs, which consistently drop, the increase in the \\nalphabet size is accompanied by a decrease in overall cost. However, combinatorial DNA synthesis or assembly \\nis still unavailable for large-scale commercial use. Thus, further development of combinatorial DNA synthesis \\ntechnologies will continue to impose limitations and constraints on combinatorial encoding, and determine \\nthe overall costs.\\nWhile our proof-of-concept experiment showed success on a small scale, there are complexities to be \\naddressed in considering large-scale applications. These include synthesis efficiency, error correction, and \\ndecoding efficiency. Nonetheless, the resilience of our binomial DNA encoding for both synthesis and sequencing \\nerrors highlights its practical potential and scalability. One specific aspect is the effect of combinatorial encoding \\non possible sequence-related constraints. While sequences with unwanted compositions (e.g., containing \\nhomopolymers) will unavoidably be part of the synthesized mixtures, the uniform sampling of the combinatorial \\nshortmers in each position, together with the independence of the different positions, guarantees that only very \\nfew such sequences will be aythesized. In particular—these will not interfere with successful reconstruction. \\nAnother challenging aspect of scaling up combinatorial DNA systems is the need to use longer DNA k-mers to \\nconstruct larger sets with the desired constraints. This may make the combinatorial synthesis impractical and \\nwill require balancing the increase in logical density with the technological complexity.\\nSeveral future research directions emerge from our study. First, it is important to develop error correction \\nmethods for better handling insertion and deletion errors. One approach for achieving this goal, is to adjust \\nsampling rates: optimizing the sampling rate, especially in large-scale experiments, can lead to data retrieval at \\nhigh accuracy. While our study highlighted the role of sampling rates in achieving desired outcomes, delving \\ndeeper into the underlying theory will lead to more improvements. For example, based on theoretical bounds of \\nsampling rates, more concrete recommendations can be provided for real-world applications. The development \\nof error correction codes, designed specifically to overcome the error types that characterize combinatorial \\nencoding, is another important direction for future research. Most notably, transitioning from small-scale proof-\\nof-concept experiments to larger-scale implementations is an important next step. Evaluating the scalability of \\nour method across various scales and complexities will be enlightening, especially when considering synthesis \\nefficiency and error rates. Finally, the consideration of advanced sequencing technologies could redefine the \\npotential and efficacy of our proposed method, including its future practical implementation.\\nTo sum up, combinatorial DNA synthesis and sequence design are important beyond the scope of DNA-based \\ndata storage. Generating combinatorial DNA libraries is an efficient tool in synthetic biology, better supporting \\nlarge-scale experiments. DNA synthesis technologies that can incorporate a combinatorial synthesis of longer \\nDNA fragments will enable the design and generation of more DNA libraries with applications in data storage \\nand beyond.\\n'},\n",
       " {},\n",
       " {'prev': '11\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       "  'Methods': 'Methods\\n',\n",
       "  'Reconstruction probability of a binomial encoding letter': 'Reconstruction probability of a binomial encoding letter\\nLet the number of reads required for reconstruction be a random variable R = \\x1fK\\ni=1Ri where R1 = 1 and \\nRi ∼Geom\\n\\x1f K−i+1\\nK\\n\\x1e\\n, i = 2, . . . , K\\u200a. Hence, the expected number of required reads, is:\\nwhere Har(·) is the harmonic number.\\nUsing the independence of Ri\\u200a, the variance of R can be bound by \\xad(See25):\\nBy Chebyshev’s inequality, we get an upper bound (a loose bound) on the probability of requiring more than \\nE[R] + cK reads to observe at least one read of each member k-mer:\\nLet c = b π\\n√\\n6\\u200a, or b = c\\n√\\n6\\nπ \\u200a, and we obtain:\\nOr specifically:\\nWe now turn to address the reconstruction of an entire oligo of length l\\u200a. Let R(l) be the random variable \\nrepresenting the number of reads required to have observed all the K member k-mers in every position. Setting \\nany δ > 0\\u200a, if we show that P(R(l) > m) ≥1 −δ\\u200a, then we know that by accumulating m reads the probability of \\ncorrect full reconstruction is more than 1 −δ\\u200a. From Eq.\\xa0(11), and assuming independence of the positions (in \\nterms of observing all K member k-mers), we get Eq.\\xa0(12):\\nFrom which we can extract c\\u200a, so that:\\nWhich yields:\\nThis process allows us to evaluate the sequencing depth complexity. For example, consider l = 100 and \\nδ = 0.01\\u200a. We want to find c\\u200a, so that using KHar(K) + cK reads will reconstruct the entire sequence with 0.99 \\nprobability. We therefore set:\\n(6)\\nE[R] =\\n\\x1fK\\ni=1E[Ri] = K\\n\\x1fK\\ni=1\\n1\\ni = KHar(K)\\n(7)\\nVar(R) =\\n\\x1fK\\ni=1Var(Ri) < K2\\n\\x1e 1\\n12 + 1\\n22 + · · · + 1\\nK2\\n\\x1d\\n< π2\\n6 K2\\n(8)\\nP(|R −E(R)| ≥bσ) ≤1\\nb2\\n(9)\\nP\\n\\x1f\\n|R −E(R)| ≥b π\\n√\\n6\\nK\\n\\x1e\\n≤1\\nb2\\n(10)\\nP(|R −E[R]| ≥cK) ≤π2\\n6c2\\n(11)\\nP(|R −KHar(K)| ≥cK) ≤π2\\n6c2\\n(12)\\nP(R(l) < KHar(K) + cK) ≥\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n(13)\\n\\x1f\\n1 −π2\\n6c2\\n\\x1el\\n≥1 −δ\\n(14)\\nc ≥\\nπ\\n\\x1f\\n6\\n\\x1e\\n1 −(1 −δ)\\n1\\nl\\n\\x1d\\nFigure\\xa05.\\u2002 \\u2009Experiment analysis. (a) A schematic view of the Gibson assembly. Each combinatorial sequence \\nconsists of a barcode segment and four payload segments (denoted as cycles 1–4). (b) Reconstruction results \\nof the two combinatorial sequences. The color indicates read frequency, and the member k-mers are marked \\nwith orange boxes. (c) The distribution of reads over the 16\\xa0k-mers in an example combinatorial letter. \\nOverlaid histograms represent the percentage of reads for each of the 16\\xa0k-mers for the same position in our \\ntwo combinatorial sequences. This, in fact, is an enlarged view of the two c4 columns of panel b. (d) Required \\nnumber of reads for reconstructing a single combinatorial letter. A histogram of the number of reads required \\nto observe at least t = 1, 2 reads from K = 5 inferred k-mers. The results are based on resampling the reads 500 \\ntimes, the data represents cycle 4. (e) Required number of reads for reconstructing a four-letter combinatorial \\nsequence. Similar to d. (f), Reconstruction failure rate as a function of the required multiplicity t\\u200a. Erroenous \\nreconstruction rate shown for different values of required copies to observe each inferred k-mer (\\u200at = 1, 2, 3, 4\\u200a). \\nThe mean required number of reads for reconstruction is displayed using a secondary Y-axis in the dashed lines.\\n◂\\n'},\n",
       " {'prev': '12\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\nAnd get:\\nAnd therefore, using 128 reads guarantees reconstruction with 0.99 probability.\\nAn end‑to‑end combinatorial storage system\\nIn Sect.\\xa0“An end-to-end combinatorial shortmer storage system” we propose an end-to-end combinatorial storage \\nsystem, as follows.\\nCombinatorial encoding and padding\\nA binary message is encoded using a large k-mer combinatorial alphabet (e.g., trimer-based alphabet of size \\n|\\x1f| = 4096 letters, with N = |\\x1f| = 16\\u200a), resulting in r = 12 bits per combinatorial letter. The binary message is \\nzero-padded to ensure its length is divisible by r prior to the combinatorial encoding. The complete message is \\nbroken into sequences of set length l = 120\\u200a, each sequence is then marked with a standard DNA barcode and \\ntranslated using the table presented in the Encode legend (See\\xa0Supplementary Sect.\\xa08.2).\\nThe length of the complete combinatorial sequence must be divisible by the payload size l and by the block \\nsize B\\u200a. As described in Fig.\\xa06, this is ensured using another padding step, and the padding information is included \\nin the final combinatorial sequence.\\nError correction codes\\nThe 2D error correction scheme includes the use of three \\xadRS26 encodings: on each barcode, on the payload part \\nof each sequence, and an outer error correction code on each block of sequences.\\n•\\t Each barcode is encoded using a systematic RS(6,8) code over GF(24)\\u200a, transforming the unique 12nt barcode \\ninto a 16nt sequence.\\n•\\t Each 120 combinatorial letter payload sequence is encoded using an RS(120,134) code over GF(212)\\u200a, resulting \\nin a sequence of length 134 combinatorial letters.\\n•\\t To protect against sequence dropouts, an outer error correction code is used on the columns of the matrix \\n(See Fig.\\xa06). Each block of B = 42 sequences, is encoded using a RS(42,48) RS code GF\\n\\x1f\\n212\\x1e\\n\\u200a. This is applied \\nin each column separately.\\nFor simplicity, Fig.\\xa06 demonstrates the encoding of\\u2009~\\u20090.1\\xa0KB using shorter messages with simpler error \\ncorrection codes. The following parameters are used:\\n•\\t A barcode length of 6nt encoded using RS(3,5) code over GF\\n\\x1f\\n24\\x1e\\n to get 10nt.\\n•\\t A payload length of l = 12 encoded using RS(12,18) over GF\\n\\x1f\\n29\\x1e\\n for the \\n\\x1f\\n16\\n3\\n\\x1e\\n binomial alphabet.\\n•\\t A 10-sequence block encoded, column-wise, using a (10,15) RS code over GF(29).\\n(15)\\n\\x1f\\n1 −π2\\n6c2\\n\\x1e100\\n≥0.99\\n(16)\\nc ≥\\nπ\\n\\x1f\\n6\\n\\x1e\\n1 −(0.99)0.01\\x1d = 127.94\\nTable\\u202f2.\\u2002 \\u2009Comparison of our experiment with data \\xadfrom14.\\nThis work\\nYan et\\xa0al.14\\nMessage length (bits)\\n96\\n84\\nLogical density (bits/ Synthesis Cycle)\\n12\\n84\\nSequence length (Cycles)\\n4\\n1\\nNumber of sequences\\n2\\n8\\nNumber of unique k-mers (N)\\n16\\n96\\nNumber of unique k-mers in each letter (K)\\n5\\n32\\nSequence length\\n220\\xa0bp\\n75\\xa0bp\\nBarcode length\\n20\\xa0bp\\n25\\xa0bp\\nSpace / Adjacent fragments\\n20\\xa0bp\\n25\\xa0bp, same in all BC\\nPayload length\\n20\\xa0bp\\n25\\xa0bp\\nSequencing method\\nIlumina Miseq\\nNanopore\\nSynthesis\\nGibson Assembly\\nBridge Oligonucleotide Assembly\\nAmount of space / Adjacent fragments\\n6\\n1',\n",
       "  'An end-to-end combinatorial storage system': '',\n",
       "  'Combinatorial encoding and padding': 'Combinatorial encoding and padding\\nA binary message is encoded using a large k-mer combinatorial alphabet (e.g., trimer-based alphabet of size \\n|\\x1f| = 4096 letters, with N = |\\x1f| = 16\\u200a), resulting in r = 12 bits per combinatorial letter. The binary message is \\nzero-padded to ensure its length is divisible by r prior to the combinatorial encoding. The complete message is \\nbroken into sequences of set length l = 120\\u200a, each sequence is then marked with a standard DNA barcode and \\ntranslated using the table presented in the Encode legend (See\\xa0Supplementary Sect.\\xa08.2).\\nThe length of the complete combinatorial sequence must be divisible by the payload size l and by the block \\nsize B\\u200a. As described in Fig.\\xa06, this is ensured using another padding step, and the padding information is included \\nin the final combinatorial sequence.\\n',\n",
       "  'Error correction codes': 'Error correction codes\\nThe 2D error correction scheme includes the use of three \\xadRS26 encodings: on each barcode, on the payload part \\nof each sequence, and an outer error correction code on each block of sequences.\\n•\\t Each barcode is encoded using a systematic RS(6,8) code over GF(24)\\u200a, transforming the unique 12nt barcode \\ninto a 16nt sequence.\\n•\\t Each 120 combinatorial letter payload sequence is encoded using an RS(120,134) code over GF(212)\\u200a, resulting \\nin a sequence of length 134 combinatorial letters.\\n•\\t To protect against sequence dropouts, an outer error correction code is used on the columns of the matrix \\n(See Fig.\\xa06). Each block of B = 42 sequences, is encoded using a RS(42,48) RS code GF\\n\\x1f\\n212\\x1e\\n\\u200a. This is applied \\nin each column separately.\\nFor simplicity, Fig.\\xa06 demonstrates the encoding of\\u2009~\\u20090.1\\xa0KB using shorter messages with simpler error \\ncorrection codes. The following parameters are used:\\n•\\t A barcode length of 6nt encoded using RS(3,5) code over GF\\n\\x1f\\n24\\x1e\\n to get 10nt.\\n•\\t A payload length of l = 12 encoded using RS(12,18) over GF\\n\\x1f\\n29\\x1e\\n for the \\n\\x1f\\n16\\n3\\n\\x1e\\n binomial alphabet.\\n•\\t A 10-sequence block encoded, column-wise, using a (10,15) RS code over GF(29).\\n(15)\\n\\x1f\\n1 −π2\\n6c2\\n\\x1e100\\n≥0.99\\n(16)\\nc ≥\\nπ\\n\\x1f\\n6\\n\\x1e\\n1 −(0.99)0.01\\x1d = 127.94\\nTable\\u202f2.\\u2002 \\u2009Comparison of our experiment with data \\xadfrom14.\\nThis work\\nYan et\\xa0al.14\\nMessage length (bits)\\n96\\n84\\nLogical density (bits/ Synthesis Cycle)\\n12\\n84\\nSequence length (Cycles)\\n4\\n1\\nNumber of sequences\\n2\\n8\\nNumber of unique k-mers (N)\\n16\\n96\\nNumber of unique k-mers in each letter (K)\\n5\\n32\\nSequence length\\n220\\xa0bp\\n75\\xa0bp\\nBarcode length\\n20\\xa0bp\\n25\\xa0bp\\nSpace / Adjacent fragments\\n20\\xa0bp\\n25\\xa0bp, same in all BC\\nPayload length\\n20\\xa0bp\\n25\\xa0bp\\nSequencing method\\nIlumina Miseq\\nNanopore\\nSynthesis\\nGibson Assembly\\nBridge Oligonucleotide Assembly\\nAmount of space / Adjacent fragments\\n6\\n1\\n'},\n",
       " {'prev': '13\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\nThe 824 bits are first padded to be 828 = 92 ∗9\\u200a. The 92 combinatorial letter message is split into 7 sequences of \\n12 letters and an additional sequence of 8 letters. Finally, a complete block of 12 sequences (total of 10 ∗12 = 120 \\nletters) is created by padding with one additional sequence of 12 letters and including the padding information \\nas the last sequence.\\n',\n",
       "  'Synthesis and sequencing simulation with errors': 'Synthesis and sequencing simulation with errors\\n•\\t Simulating the synthesis process. DNA molecules pertaining to the designed sequences are synthesized \\nusing combinatorial k-mer DNA synthesis (See Fig.\\xa01b). For each combinatorial sequence, we first determine \\nthe number of synthesized copies by sampling from X ∼N(µ = 1000, σ 2 = 100)\\u200a. Let x be the number of \\ncopies for a specific sequence. Next, for every position in the sequence, we uniformly sample x independent \\nk-mers from the set of member k-mers of the combinatorial letter in the specific position. We concatenate \\nthe sampled k-mers to the already existing x synthesized molecules.\\n•\\t Error simulation. Synthesis and sequencing errors are simulated as follows. Error probabilities for deletion, \\ninsertion, and substitution are given as parameters denoted as Pd, PI, and Ps respectively. Deletion and \\ninsertion errors are assumed to occur during k-mer synthesis and thus implemented on the k-mer level (i.e., \\nan entire k-mer is deleted or inserted in a specific position during the synthesis simulation). Substitution \\nerrors are assumed to be sequencing errors and hence implemented on a single base level (i.e., a single letter \\nis substituted, disregarding the position within the k-mer).\\n•\\t Mixing. Post synthesis, molecules undergo mixing to mirror genuine molecular combinations. This is \\nachieved through a randomized data line shuffle using an SQLite database, enabling shuffle processes even \\nfor sizable input \\xadfiles27.\\nFigure\\xa06.\\u2002 \\u2009Example of message coding, including padding and RS error correction. Encoding of a\\u2009~\\u20090.1\\xa0KB \\nmessage into a 512 letter binomial alphabet (\\u200aN = 16, K = 3)\\u200a. First, bit padding is added, included here in \\nthe letter 1\\nσ257\\u200a. Next, block padding is added, included here in 2\\nσ1 and 3\\nσ1\\u200a. Padding information is included in the \\nlast sequence of all blocks. The last sequence holds the number of padding binary bits. In this example, 4\\nσ149 \\nrepresents 148 bits of padding, composed of 4 + (4 ∗9) + (12 ∗9)bits\\u200a, 4 bits from 1\\nσ257\\u200a, 4 letters from 2\\nσ1 and 12 \\nletters from 3\\nσ1.\\n'},\n",
       " {'prev': '14\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n•\\t Reading and sampling. From the simulated synthesized molecule set, a subsample of predefined size \\nS ∗numberofsynthesizedseqeunces is drawn, simulating the sampling effect of the sequencing process.\\n',\n",
       "  'Reconstruction': 'Reconstruction\\n•\\t Barcode decoding The barcode sequence of each read is decoded using the RS(6,8) code.\\n•\\t Grouping by barcode The reads are then grouped by their barcode sequence to allow the reconstruction of \\nthe combinatorial sequences.\\n•\\t Filtering of read groups Barcodes (set of reads) with less than 10% of the sampling rate S reads are discarded.\\n•\\t Combinatorial reconstruction For each set of reads, every position is analyzed separately. The K most \\ncommon k-mers are identified and used to determine the combinatorial letter σ in this position. Let \\n\\x1f be the difference between the length of the analyzed reads and the length of the designed sequence. \\n\\x1f = l −len(read)\\u200a. Reads with |\\x1f| > k −1 are discarded from the analysis. Invalid k-mers (not in \\x1f) are \\nreplaced by a dummy k-mer Xdummy.\\n•\\t Missing barcodes Missing barcodes are replaced with dummy sequences to enable correct outer RS decoding.\\n•\\t Normalized Levenshtein distance Levenshtein distance between the observed sequence O and the expected \\nsequence E is \\xadcalculated28,29. Normalized Levenshtein distance is calculated by dividing the distance by the \\nlength of the expected sequence:\\n',\n",
       "  'Cost analysis': 'Cost analysis\\nSynthesis cost estimation was performed using the logical density calculation presented in Supplementary \\nSect.\\xa08.5 and Supplementary Table\\xa0S1. To calculate the sequencing costs, we used the coupon collector model \\npresented \\xadin24 to assess the required sequencing depth given the combinatorial alphabet. Figure\\xa04b indicates the \\ntotal number of reads required for reconstructing the sequences, calculated as the required sequencing depth \\nmultiplied by the number of sequences from Supplementary Sect.\\xa08.5 and Supplementary Table\\xa0S1. The analysis \\nwas performed on the following set of combinatorial alphabets: Standard DNA, \\n\\x1f\\n8\\n4\\n\\x1e\\n,\\n\\x1f\\n16\\n3\\n\\x1e\\n,\\n\\x1f\\n16\\n5\\n\\x1e\\n,\\n\\x1f\\n16\\n7\\n\\x1e\\n,\\n\\x1f\\n32\\n10\\n\\x1e\\n,\\n\\x1f\\n32\\n16\\n\\x1e\\n,\\n\\x1f\\n64\\n32\\n\\x1e\\n,\\n\\x1f\\n96\\n32\\n\\x1e\\n.\\n',\n",
       "  'Proof of concept experiment': 'Proof of concept experiment\\nThe proof-of-concept experiment was performed by imitating combinatorial synthesis using Gibson assembly \\nof larger DNA fragments. Each DNA fragment was composed of a 20-mer information sequence and an overlap \\nof 20\\xa0bp between adjacent fragments, as depicted in Fig.\\xa05a. Two combinatorial sequences were designed, each \\ncomposed of a barcode fragment, 4 payload fragments, and Illumina\\xa0Miseq P5 and P7 anchors at the ends. The \\ninformation fragments included in each combinatorial position were chosen from a set of 16 sequences with \\nsufficient pair-wise distance. The full list of DNA sequences and the design of combinatorial sequences is listed \\nin Supplementary Sect.\\xa08.6.\\n',\n",
       "  'DNA assembly and sequencing': 'DNA assembly and sequencing\\nPayload, barcode, and P7 anchor fragments with 20\\xa0bp overlaps for the purpose of Gibson assembly were \\nproduced by annealing complementary oligonucleotides manufactured by Integrated DNA Technologies (IDT). \\nOligos were dissolved in Duplex Buffer (100\\xa0mM Potassium Acetate; 30\\xa0mM HEPES, pH 7.5; available from \\nIDT) to the final concentration of 100 micromolar. For annealing, 25\\xa0µl of each oligo in a pair were combined \\nto the final concentration of 50 micromolars. The oligo mixes were incubated for 2\\xa0min at \\xad940 C, and gradually \\ncooled down to room temperature. The annealed payload oligos that belonged to the same cycle (5 oligos total) \\nwere mixed to the final concentration of 1 micromolar per oligo—a total of 5 micromolar, by adding 2\\xa0µl of each \\nannealed oligo into the 90\\xa0µl of nuclease-free water—a final volume of 100\\xa0µl. Annealed barcode and P7 anchor \\noligos were also diluted to the final concentration of 5 micromolar in nuclease-free water, after thorough mixing \\nby vortexing. The diluted oligos were stored at −20\\xa0°C.\\nImmediately prior to the Gibson assembly, payload oligo mixes, barcode, and P7 anchor oligos were further \\ndiluted 100-fold to the final working dilution of 0.05\\xa0pmol/microliter in nuclease-free water. Gibson reaction \\nwas assembled by adding 1\\xa0µl (0.05\\xa0pmol) of barcode, 4\\u2009×\\u2009cycle mixes, and P7 anchor to the 4\\xa0µl of nuclease-free \\nwater and supplemented with 10\\xa0µl of NEBuilder HiFi DNA assembly master mix (New England Biolabs (NEB)) \\nto the final volume of 20\\xa0µl, according to the manufacturer instructions. The reactions were incubated for 1\\xa0h \\nat 50\\xa0°C, and purified with AmpPure Beads (Thermo Scientific) at 0.8X ratio (16\\xa0µl of beads per 20\\xa0µl Gibson \\nreaction) to remove free oligos / incomplete assembly products. After adding beads and thorough mixing, \\nthe reactions were incubated for 10\\xa0min at room temperature and then placed on a magnet for 5\\xa0min at room \\ntemperature. After removing the sup, the beads were washed twice with 100\\xa0µl of 80% ethanol. The remaining \\nwashing solution was further removed by a 20\\xa0µl tip, and the beads dried for 3\\xa0min on the magnet with an open \\nlid. After removing from the magnet, the beads were resuspended in 22\\xa0µl of IDTE buffer (IDT), incubated for \\n5\\xa0min at room temperature, and then placed back on the magnet.\\n20\\xa0µl of eluate were transferred into the separate 1.7\\xa0ml tube. 5\\xa0µl of the eluted DNA were used as a template \\nfor PCR amplification combined with 23\\xa0µl of nuclease-free water, 1\\xa0µl of 20 micromolar indexing primer 5, \\n(17)\\nNormalized Levenshtein distance (O, E) = Levenshtein distance (O, E)\\n|O|\\n'},\n",
       " {'prev': '15\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n1\\xa0µl of 20 micromolar indexing primer 7, and 10\\xa0µl of rhAMPseq master mix v8.1—a total of 40\\xa0µl. After initial \\ndenaturation of 3\\xa0min at 95\\xa0°C, the PCR reaction proceeded with 50 cycles of 15\\xa0s at 95\\xa0°C, 30\\xa0s at 60\\xa0°C, and 30\\xa0s \\nat 72\\xa0°C, followed by final elongation of 1\\xa0min at 72\\xa0°C and hold at 4\\xa0°C. The PCR reactions were purified with \\nAmpure beads at 0.8X ratio (32\\xa0µl beads per 40\\xa0µl of PCR reaction) as outlined above, and eluted in 22\\xa0µl IDTE \\nbuffer. The concentration and the average size of the eluted product were determined by Qubit High Sensitivity \\nDNA kit and Agilent 2200 TapeStation system with D1000 high-sensitivity screen tape respectively. The eluted \\nproduct was diluted to 4\\xa0nM concentration, and used as an input for denatured sequencing library preparation, \\nper manufacturer instructions. The sequencing was performed on Illumina Miseq apparatus (V2 chemistry, \\n2\\u2009×\\u2009150\\xa0bp reads) using 6 picomolar denatured library supplemented with 40% PhiX sequencing control.\\n',\n",
       "  'Decoding and analysis': 'Decoding and analysis\\nThis section outlines the key steps involved in our sequencing analysis pipeline, aimed at effectively processing \\nand interpreting sequenced reads. The analysis pipeline gets the sequencing output file containing raw reads in \\n“.fastq” format and a design file containing the combinatorial sequences.\\nAnalysis steps:\\n1.\\t Length filtering. We saved reads that were 220\\xa0bp in length, retaining only those corresponding to our \\ndesigned read length.\\n2.\\t Read retrieval. We carefully checked each read for the presence of BCs, universals, and payloads. To keep \\nour data accurate, we discarded reads where the BCs, universals, or payloads had a Hamming distance of \\nmore than 3 errors.\\n3.\\t Identifying inferred k-mers. For every BC and each cycle, we counted the K most common k-mers. We \\nthen compared these with the design file to quantify those matching (Fig.\\xa05b) (See\\xa0Table\\xa03).\\n',\n",
       "  'Information capacities for selected encodings': 'Information capacities for selected encodings\\nTable\\xa01 illustrates the logical densities derived from encoding a 1\\xa0GB binary message using oligonucleotides with \\na 12nt barcode and an additional 4nt for standard DNA RS error correction, and a 120 letters payload with 14 \\nextra RS for the payload in combinatorial encoding schemes with parameters N and K.\\nThe densities were calculated as follows:\\n(18)\\nBits per Letter =\\n\\x1f\\nlog2\\n\\x1e\\x1e\\nN\\nk\\n\\x1d\\x1d\\x1c\\n(19)\\nAlphabet Size = 2Bits per Letter\\n(20)\\nBits per Sequence = Bits per Letter × Payload length.\\n(21)\\nNumber of Sequences = ⌈Message Size in bits\\nBits per Sequence ⌉\\n(22)\\nNumber of sequences padded : Total number of sequences after padding for the block size\\n(23)\\nPadding is = Block Size −\\n\\x1f\\nNumber of Sequences%Block Size\\n\\x1e\\n(24)\\nNumber of sequences with RS = Number of sequences padded\\nBlock Size\\n× Block Size AfterRS\\nTable\\u202f3.\\u2002 \\u2009Summary of sequencing reads analyzed in the study. The table shows the total number of reads \\nobtained, the number filtered by length (220 bases) for analysis, and the counts of reads associated with BC1, \\nBC2, and those that did not have any recognizable barcode (No BC).\\nReads\\nCount\\nTotal PF reads\\n2,634,683\\nReads of length 220\\n2,139,071\\nBC1\\n1,365,295\\nBC2\\n768,755\\nNo BC\\n5021\\n'},\n",
       " {'prev': '16\\nVol:.(1234567890)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       "  'Ethics declaration': 'Ethics declaration\\nNo animal or human subjects were involved in the study.\\nData availability\\nThe raw data is available in ENA (European Nucleotide Archive). The datasets generated and/or analyzed \\nduring the current\\xa0study are available in the ENA (European Nucleotide Archive) repository, Accession \\nNumber—ERR12364864.\\nCode Availability\\nImplementation of the algorithms and instructions on how to use them can be found in the GitHub repository\\xa0in \\nthe following links: https://\\u200bgithub.\\u200bcom/\\u200bInbal\\u200bPreuss/\\u200bdna_\\u200bstora\\u200bge_\\u200bshort\\u200bmer_\\u200bsimul\\u200bation,\\xa0https://\\u200bgithub.\\u200bcom/\\u200bInbal\\u200b\\nPreuss/\\u200bdna_\\u200bstora\\u200bge_\\u200bexper\\u200biment.\\nReceived: 11 December 2023; Accepted: 28 March 2024\\n',\n",
       "  'References': 'References\\n\\t 1.\\t Church, G., Gao, Y. & Kosuri, S. Next-generation digital information storage in DNA. Science 337, 1628 (2012).\\n\\t 2.\\t Goldman, N. et al. Towards practical, high-capacity, low-maintenance information storage in synthesized DNA. Nature 494, 77–80 \\n(2013).\\n\\t 3.\\t Anavy, L., Vaknin, I., Atar, O., Amit, R. & Yakhini, Z. Data storage in DNA with fewer synthesis cycles using composite DNA \\nletters. Nat. Biotechnol. 37, 1229–1236 (2019).\\n\\t 4.\\t Erlich, Y. & Zielinski, D. DNA Fountain enables a robust and efficient storage architecture. Science 355, 950–954 (2017).\\n\\t 5.\\t Gabrys, R., Kiah, H., & Milenkovic, O. Asymmetric lee distance codes for DNA-based storage. In 2015 IEEE International \\nSymposium on Information Theory (ISIT) (2015).\\n\\t 6.\\t NallappaBhavithran, G., & Selvakumar, R. Indel Error Correction Codes for DNA Digital Data Storage and Retrieval. ArXiv \\nabs/2302.1467 (2023).\\n\\t 7.\\t Wang, C. et al. Mainstream encoding–decoding methods of DNA data. CCF Trans. High Perform. Comput. 4, 23–22 (2022).\\n\\t 8.\\t Boruchvosky, A., Bar-Lev, D., & Yaakobi, E. DNA-Correcting Codes: End-to-end Correction in DNA Storage Systems. ArXiv, \\nabs/2304.0391 (2023).\\n\\t 9.\\t Bornholt, J. et al. Toward a DNA-based archival storage system. IEEE Micro 37, 98–104 (2017).\\n\\t10.\\t Yazdi, S., Yuan, Y., Ma, J., Zhao, H. & Milenkovic, O. A rewritable, random-access DNA-based storage system. Sci. Rep. 5, 1–10 \\n(2015).\\n\\t11.\\t Organick, L. et al. Random access in large-scale DNA data storage. Biotechnol. 36, 242–248 (2018).\\n\\t12.\\t Choi, Y. et al. High information capacity DNA-based data storage with augmented encoding characters using degenerate bases. \\nSci. Rep. 9, 6582 (2019).\\n\\t13.\\t Roquet, N., Bhatia, S., Flickinger, S., Mihm, S., Norsworthy, M., Leake, D., & Park, H. DNA-based data storage via combinatorial \\nassembly. 20 April 2021 (online). https://\\u200bdoi.\\u200borg/\\u200b10.\\u200b1101/\\u200b2021.\\u200b04.\\u200b20.\\u200b44019\\u200b4v1.\\n\\t14.\\t Yan, Y., Pinnamaneni, N., Chalapati, S., Crosbie, C. & Appuswamy, R. Scaling logical density of DNA storage with enzymatically-\\nligated composite motifs. Sci. Rep. 13, 15978 (2023).\\n\\t15.\\t LeProust, E. et al. Synthesis of high-quality libraries of long (105mer) oligonucleotides by a nover depurination controlled process. \\nNucl. Acids Res. 38, 2522–2540 (2019).\\n\\t16.\\t Barrett, M. et al. Comparative genomic hybridization using oligonucleotide microarrays and total genomic DNA. Proc. Natl Acad. \\nSci. USA 101, 17765–17770 (2004).\\n\\t17.\\t Eleuteri, A., Capaldi, D., Douglas, L. & Ravikumar, V. Oligodeoxyribonucleotide phosphorothioates: Substantial reduction of \\n(N-1)-mer content through the use of trimeric phosphoramidite synthons. Nucleosides Nucleotides 3, 475–483 (1999).\\n\\t18.\\t Yagodkin, A. et al. Improved synthesis of trinucleotide phosphoramidites and generation of randomized oligonucleotide libraries. \\nNucleosides Nucleotides Nucl. Acids 26(5), 473–497 (2007).\\n\\t19.\\t Randolph, J., Yagodkin, A. & Mackie, H. Codon-based Mutagenesis. Nucl. Acids Symp. Ser. 52, 479 (2008).\\n\\t20.\\t Ferrante, M., & Saltalamacchia, M. The Coupon Collector’s Problem, p 35 (2014).\\n\\t21.\\t Press, W. et al. HEDGES error-correcting code for DNA storage corrects indels and allows sequence constraints. Proc. Natl. Acad. \\nSci. 117(31), 18489–18496 (2020).\\n\\t22.\\t Haoling, Z., et al. SPIDER-WEB generates coding algorithms with superior error tolerance and real-time information retrieval \\ncapacity. arXiv preprint arXiv 2204.02855 (2022).\\n\\t23.\\t Sabary, O., Orlev, Y., Shafir, R. & Anavy, L. SOLQC: Synthetic oligo library quality control tool. Bioinformatics 2, 740 (2020).\\n\\t24.\\t Preuss, I., Galili, B., Yakhini, Z., & Anavy, Z. Sequencing coverage analysis for combinatorial DNA-based storage systems. biorxiv \\n(2024).\\n\\t25.\\t Ayoub, R. Euler and the zeta function. Am. Math. Mon. 81, 1067–1086 (1974).\\n\\t26.\\t Reed, I. & Solomon, G. Polynomial codes over certain finite fields. J. Soc. Ind. Appl. Math. 8, 300–304 (1960).\\n\\t27.\\t Hipp, R. D. SQLite (2020) (Online). https://\\u200bwww.\\u200bsqlite.\\u200borg/\\u200bindex.\\u200bhtml.\\n\\t28.\\t Levenshtein, V. Binary codes capable of correcting spurious insertions and deletions of ones. Problems Inf. Transm. 1, 8–17 (1965).\\n\\t29.\\t Levenshtein, V. Binary codes capable of correcting deletion, insertions and reversals. Soviet Physics Doklady 10(8), 707–710 (1966).\\n(25)\\nSynthesis Cycles = Number of sequences with RS × Full sequence length with RS\\n(26)\\nBits per Synthesis Cycle = Message Size in bits\\nSynthesis Cycle\\n(27)\\nFold Increase =\\nBits per Synthesis Cycle\\nBits per Synthesis Cycle of Standard Scheme\\n'},\n",
       " {'prev': '17\\nVol.:(0123456789)\\nScientific Reports |         (2024) 14:7731  | \\nhttps://doi.org/10.1038/s41598-024-58386-z\\nwww.nature.com/scientificreports/\\n',\n",
       "  'Acknowledgements': 'Acknowledgements\\nThis project has received funding from the European Union’s Horizon Europe research and innovation \\nprogramme under grant\\xa0agreement No. 101115134 (DiDAX project). The authors of this paper thank the Yakhini \\nresearch group for the fruitful\\xa0discussions. The authors also thank Eland Nagar for his support and problem-\\nsolving approach regarding the experimental\\xa0proof of concept.\\nAuthor contributions\\nI.P., Z.Y. and L.A conceived the idea, designed the experiments, interpreted and analyzed the data. I.P performed \\nall computational work, including the simulations and the data analysis. M.R. was solely responsible for \\nconducting the experimental work. All authors wrote the manuscript.\\nFunding\\nEuropean Union’s Horizon Europe Research and Innovation Programme, 101115134.\\nCompeting interests\\u2002\\nZohar Yakhini, Leon Anavy have competing interests as defined by Nature Research. Z. Yakhini and L. Anavy are \\nnamed as inventors on a patent related to the content of this paper: L. Anavy, Z. Yakhini, and R. Amit, \"Molecular \\ndata storage systems and methods\". United States of America Patent US20210141568A1, 2021.\\nAdditional information\\nSupplementary Information The online version contains supplementary material available at https://\\u200bdoi.\\u200borg/\\u200b\\n10.\\u200b1038/\\u200bs41598-\\u200b024-\\u200b58386-z.\\nCorrespondence and requests for materials should be addressed to I.P.\\nReprints and permissions information is available at www.nature.com/reprints.\\nPublisher’s note\\u2002 Springer Nature remains neutral with regard to jurisdictional claims in published maps and \\ninstitutional affiliations.\\nOpen Access\\u2002  This article is licensed under a Creative Commons Attribution 4.0 International \\nLicense, which permits use, sharing, adaptation, distribution and reproduction in any medium or \\nformat, as long as you give appropriate credit to the original author(s) and the source, provide a link to the \\nCreative Commons licence, and indicate if changes were made. The images or other third party material in this \\narticle are included in the article’s Creative Commons licence, unless indicated otherwise in a credit line to the \\nmaterial. If material is not included in the article’s Creative Commons licence and your intended use is not \\npermitted by statutory regulation or exceeds the permitted use, you will need to obtain permission directly from \\nthe copyright holder. To view a copy of this licence, visit http://\\u200bcreat\\u200biveco\\u200bmmons.\\u200borg/\\u200blicen\\u200bses/\\u200bby/4.\\u200b0/.\\n© The Author(s) 2024\\n'}]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_page_dicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "ca4362cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'11\\ncapacity limits that are outside of the reach of the earlier\\nschemes. We show how to calculate those fundamental limits\\nand how to reason about the associated trade-off between the\\nread and write costs. We devise an error-correcting scheme\\nthat approaches these limits in practice and show how to\\ncircumvent the combinatorial complexity explosion in our\\ndecoding algorithm at the cost of some additional redundancy.\\nOverall, the paper offers a powerful framework to reason about\\nand incorporate error correcting coding in one of the most\\npromising approaches to DNA data storage.\\nREFERENCES\\n[1] L. Ceze, J. Nivala, and K. Strauss, “Molecular digital data storage using\\nDNA,” Nature Reviews Genetics, vol. 20, no. 8, pp. 456–466, May 2019.\\n[2] Y. Yan, N. Pinnamaneni, S. Chalapati, C. Crosbie, and R. Appuswamy,\\n“Scaling logical density of DNA storage with enzymatically-ligated\\ncomposite motifs,” Scientific Reports, vol. 13, no. 1, p. 15978, Sep.\\n2023.\\n[3] I. Preuss, M. Rosenberg, Z. Yakhini, and L. Anavy, “Efficient DNA-\\nbased data storage using shortmer combinatorial encoding,” Scientific\\nReports, vol. 14, no. 1, p. 7731, Apr. 2024.\\n[4] I. Preuss, B. Galili, Z. Yakhini, and L. Anavy, “Sequencing coverage\\nanalysis for combinatorial DNA-based storage systems,” IEEE Trans.\\nMol. Biol. Multi-Scale Commun., vol. 10, no. 2, pp. 297–316, Jun. 2024.\\n[5] T. Cohen and E. Yaakobi, “Optimizing the decoding probability and\\ncoverage ratio of composite DNA,” in Proc. IEEE Int. Symp. Inf. Theory\\n(ISIT), Athens, Greece, Jul. 2024.\\n[6] D. Bar-Lev, O. Sabary, R. Gabrys, and E. Yaakobi, “Cover your bases:\\nHow to minimize the sequencing coverage in DNA storage systems,” in\\nProc. IEEE Int. Symp. Inf. Theory (ISIT), Taipei, Taiwan, Jun. 2023, pp.\\n370–375.\\n[7] A. Gruica, D. Bar-Lev, A. Ravagnani, and E. Yaakobi, “A combinatorial\\nperspective on random access efficiency for DNA storage,” in Proc.\\nIEEE Int. Symp. Inf. Theory (ISIT), Athens, Greece, Jul. 2024.\\n[8] H. Abraham, R. Gabrys, and E. Yaakobi, “Covering all bases: The next\\ninning in DNA sequencing efficiency,” in Proc. IEEE Int. Symp. Inf.\\nTheory (ISIT), Athens, Greece, Jul. 2024.\\n[9] O. Sabary, I. Preuss, R. Gabrys, Z. Yakhini, L. Anavy, and E. Yaakobi,\\n“Error-correcting codes for combinatorial composite DNA,” in Proc.\\nIEEE Int. Symp. Inf. Theory (ISIT), Athens, Greece, Jul. 2024.\\n[10] R. Heckel, I. Shomorony, K. Ramchandran, and D. N. C. Tse, “Funda-\\nmental limits of DNA storage systems,” in Proc. IEEE Int. Symp. Inf.\\nTheory, Aachen, Germany, Jun. 2017, pp. 3130–3134.\\n[11] R. P. Stanley, Enumerative Combinatorics, 2nd ed.\\nCambridge Univer-\\nsity Press, 2011, vol. 1.\\n[12] R. G. Gallager, Information Theory and Reliable Communication. John\\nWiley & Sons, Inc., 1968.\\n[13] P. Flajolet and R. Sedgewick, Analytic combinatorics.\\nCambridge\\nUniversity Press, 2009.\\n[14] S. Chandak, K. Tatwawadi, B. Lau, M. Kubit, J. Mardia, J. Neu,\\nP. Griffin, M. Wootters, T. Weissman, and H. Ji, “Improved read/write\\ncost tradeoff in DNA-based data storage using LDPC codes,” in Proc.\\nAnnu. Allerton Conf. Commun., Control, Comput., Monticello, IL, USA,\\nSep. 2019, pp. 147–156.\\n[15] M. C. Davey and D. J. C. MacKay, “Low density parity check codes\\nover GF(q),” IEEE Commun. Lett., vol. 2, no. 6, pp. 165–167, Jun. 1998.\\n[16] W. E. Ryan and S. Lin, Channel Codes.\\nCambridge University Press,\\n2009.\\n[17] C. Atkins and J. Sayir, “Density evolution for SUDOKU codes on the\\nerasure channel,” in 2014 8th International Symposium on Turbo Codes\\nand Iterative Information Processing (ISTC), Bremen, Germany, Aug.\\n2014, pp. 233–237.\\n[18] A. Jimen´ez Feltstr¨om and K. S. Zigangirov, “Time-varying periodic\\nconvolutional codes with low-density parity-check matrix,” IEEE Trans.\\nInf. Theory, vol. 45, no. 6, pp. 2181–2191, Sep. 1999.\\n[19] M. Lentmaier, A. Sridharan, D. J. Costello, and K. S. Zigangirov,\\n“Iterative decoding threshold analysis for LDPC convolutional codes,”\\nIEEE Trans. Inf. Theory, vol. 56, no. 10, pp. 5274–5289, Oct. 2010.\\n[20] S. Kudekar, T. Richardson, and R. L. Urbanke, “Spatially coupled\\nensembles universally achieve capacity under belief propagation,” IEEE\\nTrans. Inf. Theory, vol. 59, no. 12, pp. 7761–7813, Dec. 2013.\\n[21] A. Sridharan, D. Truhachev, M. Lentmaier, D. J. Costello, and K. S.\\nZigangirov, “Distance bounds for an ensemble of LDPC convolutional\\ncodes,” IEEE Trans. Inf. Theory, vol. 53, no. 12, pp. 4537–4555, Dec.\\n2007.\\n[22] A. R. Iyengar, M. Papaleo, P. H. Siegel, J. K. Wolf, A. Vanelli-Coralli,\\nand G. E. Corazza, “Windowed decoding of protograph-based LDPC\\nconvolutional codes over erasure channels,” IEEE Trans. Inf. Theory,\\nvol. 58, no. 4, pp. 2303–2320, Apr. 2012.\\n'"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e75f9d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 'Introduction', 1]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toc[0][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "47b6f527",
   "metadata": {},
   "outputs": [],
   "source": [
    "block_dict = {}\n",
    "page_num = 1\n",
    "for page in doc: # Iterate all pages in the document\n",
    "    file_dict = page.get_text('html') # Get the page dictionary\n",
    "    #block = file_dict['blocks'] # Get the block information\n",
    "    #block_dict[page_num] = block # Store in block dictionary\n",
    "    #page_num += 1 # Increase the page value by 1\n",
    "\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "09bd4efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('page_doc.txt', 'w') as f:\n",
    "    f.write(file_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5e875999",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Font family: NimbusRomNo9L Font size: 7.0 Text: 1\n",
      "Font family: NimbusRomNo9L Font size: 23.9 Text: Coding Over Coupon Collector Channels\n",
      "Font family: NimbusRomNo9L Font size: 23.9 Text: for Combinatorial Motif-Based DNA Storage\n",
      "Font family: NimbusRomNo9L Font size: 11.0 Text: Roman Sokolovskii, Parv Agarwal, Luis Alberto Croquevielle, Zijian Zhou, and Thomas Heinis\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: Abstract\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: &#x2014;Encoding\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: information\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: in\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: combinations\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: of\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: pre-\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: synthesised deoxyribonucleic acid (DNA) strands (referred to as\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: motifs\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: ) is an interesting approach to DNA storage that could\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: potentially circumvent the prohibitive costs of nucleotide-by-\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: nucleotide DNA synthesis. Based on our analysis of an empirical\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: data set from HelixWorks, we propose two channel models for\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: this setup (with and without interference) and analyse their\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: fundamental limits. We propose a coding scheme that approaches\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: those limits by leveraging all information available at the output\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: of the channel, in contrast to earlier schemes developed for\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: a similar setup by Preuss\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text:  et al.\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text:  We highlight an important\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: connection between channel capacity curves and the fundamental\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: trade-off between synthesis (writing) and sequencing (reading)\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: costs, and offer a way to mitigate an exponential growth in\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: decoding complexity with the size of the motif library.\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: Index Terms&#x2014;\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: DNA storage, coupon collector&#x2019;s problem, spa-\n",
      "Font family: NimbusRomNo9L Font size: 9.0 Text: tially coupled LDPC codes.\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: I. I\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: NTRODUCTION\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: The density and durability of deoxyribonucleic acid (DNA)\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: make it an attractive medium for archival storage. However,\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: despite numerous demonstrations of the feasibility of DNA\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: storage (we refer the reader to [1] for a comprehensive\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: overview), prohibitive cost of DNA synthesis hinders its\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: widespread adoption. A promising alternative to the most com-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: monly used (and expensive) nucleotide-by-nucleotide DNA\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: synthesis is encoding information in combinations of pre-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: synthesised oligonucleotides (referred to as\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  motifs\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: ) from a\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: fixed motif library [2]. We refer to those combinations of mo-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: tifs as\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  combinatorial symbols\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: . During synthesis (writing), the\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: selected combinations of motifs are added to a pool of growing\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: DNA strands, with no control over which of those motifs is\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: attached to any given strand, and therefore over which of those\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: motifs will be encountered during sequencing (reading). The\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: exponential growth in the number of combinations allows for\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: a linear growth in information density with the size of the\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: motif library, at the cost of additional sequencing.\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: Combinatorial motif-based DNA storage is a refreshing\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: setup from a coding-theoretic perspective with little research\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: conducted to date. In [3], Preuss\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  et al.\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  emphasise an important\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: connection between the process of accumulating the motifs\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: during sequencing with the classical Coupon Collector&#x2019;s Prob-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: lem. The authors employ maximum distance separable (MDS)\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: codes and rely on fully accumulated symbols for decoding. In\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: other words, the coding scheme requires only a subset of the\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: This work was funded by MoSS (grant 101058035).\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: The authors are with the Department of Computing, Imperial College\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: London, SW7 2AZ London, UK (email: r.sokolovskii@imperial.ac.uk;\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: p.agarwal23@imperial.ac.uk;\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: a.croquevielle22@imperial.ac.uk;\n",
      "Font family: NimbusRomNo9L Font size: 8.0 Text: zijian.zhou18@imperial.ac.uk; t.heinis@imperial.ac.uk).\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: combinatorial symbols to be recovered in order to decode a\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: codeword, but these symbols must be recovered fully (i.e.,\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: all motifs constituting these symbols must be encountered\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: during sequencing) for decoding to be possible. This MDS-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: based scheme is subsequently analysed in detail in a follow-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: up paper [4] that proposes a computational framework for\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: choosing the parameters of the coding scheme based on\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: coverage and reliability constraints. For a similar setup, [5]\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: gives the expected number of reads required to reconstruct\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: information. We remark that in the case of standard (as op-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: posed to combinatorial motif-based) synthesis, the connection\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: between DNA sequencing and the Coupon Collector&#x2019;s Problem\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: is formally introduced in [6]; a random-access version of the\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: problem is further investigated in [7], [8].\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: The approach in [3], [4] has two important limitations: First,\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: as we elaborate in Section III, using only fully recovered\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: symbols during decoding reduces the pipeline to an erasure\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: channel; processing partial combinations offers a promising\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: way to improve decoding by leveraging all available informa-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: tion. Second, the setup in [4] mitigates possible\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  interference\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: (i.e., the detection of a motif that is not a part of the chosen\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: combination) by requiring several copies of the candidate\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: motifs to be encountered and declaring erasure otherwise. This\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: further increases the required coverage&#x2014;again, processing full\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: available probabilistic information could significantly improve\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: decoding performance.\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: The first work to make use of partial combinations is [9],\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: which studies adversarial code constructions that provably\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: recover from up to\n",
      "Font family: CMMI10 Font size: 10.0 Text:  e\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  missing motifs in up to\n",
      "Font family: CMMI10 Font size: 10.0 Text:  d\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  symbols\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: (the possibility of interference is not considered). The work\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: provides bounds on the size of such codes and offers an\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: explicit encoding scheme for correcting up to\n",
      "Font family: CMMI10 Font size: 10.0 Text:  e\n",
      "Font family: CMR10 Font size: 10.0 Text:  = 1\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  missing\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: motif (with\n",
      "Font family: CMMI10 Font size: 10.0 Text:  d\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  remaining a configurable parameter) when the\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: block length does not exceed the size of the motif library.\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: In this work, we consider error correction for combinatorial\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: motif-based DNA storage beyond the MDS-based schemes\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: proposed and analysed in [3], [4], aiming to use\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  all\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text:  information\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: available at the output of the channel. Based on our analysis of\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: the experimental data set provided by HelixWorks (Section II),\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: we formally introduce a channel model that mimics the empir-\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: ically observed behaviour, which we call the Coupon Collector\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: Channel, either with or without interference (Section III). We\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: derive the capacity for the interference-free Coupon Collector\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: Channel (Section III-A) and its erasure-channel counterpart\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: (Section III-B), as well as propose a computational method\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: to estimate the capacity of the channel with interference\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: (Section III-C). We demonstrate that relying only on fully\n",
      "Font family: NimbusRomNo9L Font size: 10.0 Text: accumulated combinations imposes a fundamental limit on\n",
      "Font family: Times New Roman Font size: 20.0 Text: arXiv:2406.04141v3  [cs.IT]  4 Dec 2024\n"
     ]
    }
   ],
   "source": [
    "\n",
    "split_page = file_dict.split('\\n')\n",
    "\n",
    "for line in split_page:\n",
    "\n",
    "    if line.startswith('<p'):\n",
    "        #print(line)\n",
    "\n",
    "        pattern = re.compile(\n",
    "            r'font-family:([^;,]+).*?font-size:([\\d.]+)pt.*?>(.*?)<',\n",
    "            re.DOTALL\n",
    "        )\n",
    "        matches = pattern.findall(line)\n",
    "\n",
    "        for font_family, font_size, text in matches:\n",
    "            print(\"Font family:\", font_family.strip(),\n",
    "                \"Font size:\", font_size,\n",
    "                \"Text:\", text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "c94f55a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting unidecode\n",
      "  Downloading Unidecode-1.4.0-py3-none-any.whl.metadata (13 kB)\n",
      "Downloading Unidecode-1.4.0-py3-none-any.whl (235 kB)\n",
      "   ---------------------------------------- 0.0/235.8 kB ? eta -:--:--\n",
      "   - -------------------------------------- 10.2/235.8 kB ? eta -:--:--\n",
      "   ----- --------------------------------- 30.7/235.8 kB 640.0 kB/s eta 0:00:01\n",
      "   --------------- ------------------------ 92.2/235.8 kB 1.0 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 235.8/235.8 kB 1.6 MB/s eta 0:00:00\n",
      "Installing collected packages: unidecode\n",
      "Successfully installed unidecode-1.4.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEPRECATION: Loading egg at c:\\users\\parv\\anaconda3\\envs\\pytorch_gpu\\lib\\site-packages\\ont_bonito-0.8.1-py3.12.egg is deprecated. pip 24.3 will enforce this behaviour change. A possible replacement is to use pip for package installation.. Discussion can be found at https://github.com/pypa/pip/issues/12330\n"
     ]
    }
   ],
   "source": [
    "!pip install unidecode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "2685b239",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "\n",
      "\n",
      "\n",
      "Coding Over Coupon Collector Channels\n",
      "for Combinatorial Motif-Based DNA Storage\n",
      "\n",
      "\n",
      "\n",
      "Roman Sokolovskii, Parv Agarwal, Luis Alberto Croquevielle, Zijian Zhou, and Thomas Heinis\n",
      "\n",
      "\n",
      "\n",
      "Abstract--Encoding\n",
      "information\n",
      "in\n",
      "combinations\n",
      "of\n",
      "pre-\n",
      "synthesised deoxyribonucleic acid (DNA) strands (referred to as\n",
      "motifs) is an interesting approach to DNA storage that could\n",
      "potentially circumvent the prohibitive costs of nucleotide-by-\n",
      "nucleotide DNA synthesis. Based on our analysis of an empirical\n",
      "data set from HelixWorks, we propose two channel models for\n",
      "this setup (with and without interference) and analyse their\n",
      "fundamental limits. We propose a coding scheme that approaches\n",
      "those limits by leveraging all information available at the output\n",
      "of the channel, in contrast to earlier schemes developed for\n",
      "a similar setup by Preuss et al. We highlight an important\n",
      "connection between channel capacity curves and the fundamental\n",
      "trade-off between synthesis (writing) and sequencing (reading)\n",
      "costs, and offer a way to mitigate an exponential growth in\n",
      "decoding complexity with the size of the motif library.\n",
      "\n",
      "\n",
      "\n",
      "Index Terms--DNA storage, coupon collector's problem, spa-\n",
      "tially coupled LDPC codes.\n",
      "\n",
      "\n",
      "\n",
      "I. INTRODUCTION\n",
      "\n",
      "\n",
      "\n",
      "The density and durability of deoxyribonucleic acid (DNA)\n",
      "make it an attractive medium for archival storage. However,\n",
      "despite numerous demonstrations of the feasibility of DNA\n",
      "storage (we refer the reader to [1] for a comprehensive\n",
      "overview), prohibitive cost of DNA synthesis hinders its\n",
      "widespread adoption. A promising alternative to the most com-\n",
      "monly used (and expensive) nucleotide-by-nucleotide DNA\n",
      "synthesis is encoding information in combinations of pre-\n",
      "synthesised oligonucleotides (referred to as motifs) from a\n",
      "fixed motif library [2]. We refer to those combinations of mo-\n",
      "tifs as combinatorial symbols. During synthesis (writing), the\n",
      "selected combinations of motifs are added to a pool of growing\n",
      "DNA strands, with no control over which of those motifs is\n",
      "attached to any given strand, and therefore over which of those\n",
      "motifs will be encountered during sequencing (reading). The\n",
      "exponential growth in the number of combinations allows for\n",
      "a linear growth in information density with the size of the\n",
      "motif library, at the cost of additional sequencing.\n",
      "Combinatorial motif-based DNA storage is a refreshing\n",
      "setup from a coding-theoretic perspective with little research\n",
      "conducted to date. In [3], Preuss et al. emphasise an important\n",
      "connection between the process of accumulating the motifs\n",
      "during sequencing with the classical Coupon Collector's Prob-\n",
      "lem. The authors employ maximum distance separable (MDS)\n",
      "codes and rely on fully accumulated symbols for decoding. In\n",
      "other words, the coding scheme requires only a subset of the\n",
      "\n",
      "\n",
      "\n",
      "This work was funded by MoSS (grant 101058035).\n",
      "The authors are with the Department of Computing, Imperial College\n",
      "London, SW7 2AZ London, UK (email: r.sokolovskii@imperial.ac.uk;\n",
      "p.agarwal23@imperial.ac.uk;\n",
      "a.croquevielle22@imperial.ac.uk;\n",
      "zijian.zhou18@imperial.ac.uk; t.heinis@imperial.ac.uk).\n",
      "\n",
      "\n",
      "\n",
      "combinatorial symbols to be recovered in order to decode a\n",
      "codeword, but these symbols must be recovered fully (i.e.,\n",
      "all motifs constituting these symbols must be encountered\n",
      "during sequencing) for decoding to be possible. This MDS-\n",
      "based scheme is subsequently analysed in detail in a follow-\n",
      "up paper [4] that proposes a computational framework for\n",
      "choosing the parameters of the coding scheme based on\n",
      "coverage and reliability constraints. For a similar setup, [5]\n",
      "gives the expected number of reads required to reconstruct\n",
      "information. We remark that in the case of standard (as op-\n",
      "posed to combinatorial motif-based) synthesis, the connection\n",
      "between DNA sequencing and the Coupon Collector's Problem\n",
      "is formally introduced in [6]; a random-access version of the\n",
      "problem is further investigated in [7], [8].\n",
      "The approach in [3], [4] has two important limitations: First,\n",
      "as we elaborate in Section III, using only fully recovered\n",
      "symbols during decoding reduces the pipeline to an erasure\n",
      "channel; processing partial combinations offers a promising\n",
      "way to improve decoding by leveraging all available informa-\n",
      "tion. Second, the setup in [4] mitigates possible interference\n",
      "(i.e., the detection of a motif that is not a part of the chosen\n",
      "combination) by requiring several copies of the candidate\n",
      "motifs to be encountered and declaring erasure otherwise. This\n",
      "further increases the required coverage--again, processing full\n",
      "available probabilistic information could significantly improve\n",
      "decoding performance.\n",
      "The first work to make use of partial combinations is [9],\n",
      "which studies adversarial code constructions that provably\n",
      "recover from up to e missing motifs in up to d symbols\n",
      "(the possibility of interference is not considered). The work\n",
      "provides bounds on the size of such codes and offers an\n",
      "explicit encoding scheme for correcting up to e = 1 missing\n",
      "motif (with d remaining a configurable parameter) when the\n",
      "block length does not exceed the size of the motif library.\n",
      "In this work, we consider error correction for combinatorial\n",
      "motif-based DNA storage beyond the MDS-based schemes\n",
      "proposed and analysed in [3], [4], aiming to use all information\n",
      "available at the output of the channel. Based on our analysis of\n",
      "the experimental data set provided by HelixWorks (Section II),\n",
      "we formally introduce a channel model that mimics the empir-\n",
      "ically observed behaviour, which we call the Coupon Collector\n",
      "Channel, either with or without interference (Section III). We\n",
      "derive the capacity for the interference-free Coupon Collector\n",
      "Channel (Section III-A) and its erasure-channel counterpart\n",
      "(Section III-B), as well as propose a computational method\n",
      "to estimate the capacity of the channel with interference\n",
      "(Section III-C). We demonstrate that relying only on fully\n",
      "accumulated combinations imposes a fundamental limit on\n",
      "\n",
      "\n",
      "\n",
      "arXiv:2406.04141v3  [cs.IT]  4 Dec 2024\n",
      "\n",
      "2\n",
      "\n",
      "\n",
      "\n",
      "8\n",
      "7\n",
      "6\n",
      "5\n",
      "4\n",
      "3\n",
      "2\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "Motif library\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "1\n",
      "2\n",
      "3\n",
      "1\n",
      "4\n",
      "2\n",
      "3\n",
      "\n",
      "\n",
      "\n",
      "2\n",
      "3\n",
      "5\n",
      "5\n",
      "4\n",
      "5\n",
      "4\n",
      "6\n",
      "\n",
      "\n",
      "\n",
      "5\n",
      "4\n",
      "6\n",
      "7\n",
      "6\n",
      "7\n",
      "5\n",
      "7\n",
      "\n",
      "\n",
      "\n",
      "8\n",
      "7\n",
      "8\n",
      "8\n",
      "7\n",
      "8\n",
      "7\n",
      "8\n",
      "A1\n",
      "\n",
      "\n",
      "\n",
      "Data block\n",
      "\n",
      "\n",
      "\n",
      "A0\n",
      "\n",
      "\n",
      "\n",
      "Encoding\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "4\n",
      "2\n",
      "5\n",
      "7\n",
      "4\n",
      "4\n",
      "3\n",
      "\n",
      "\n",
      "\n",
      "8\n",
      "1\n",
      "2\n",
      "8\n",
      "6\n",
      "4\n",
      "2\n",
      "8\n",
      "A1\n",
      "A0\n",
      "\n",
      "\n",
      "\n",
      "A1\n",
      "\n",
      "\n",
      "\n",
      "Strands\n",
      "\n",
      "\n",
      "\n",
      "A0\n",
      "\n",
      "\n",
      "\n",
      "...\n",
      "Synthesis\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "1\n",
      "2\n",
      "?\n",
      "?\n",
      "4\n",
      "2\n",
      "3\n",
      "\n",
      "\n",
      "\n",
      "?\n",
      "?\n",
      "?\n",
      "5\n",
      "?\n",
      "?\n",
      "4\n",
      "?\n",
      "\n",
      "\n",
      "\n",
      "?\n",
      "4\n",
      "?\n",
      "?\n",
      "6\n",
      "?\n",
      "?\n",
      "?\n",
      "\n",
      "\n",
      "\n",
      "8\n",
      "?\n",
      "?\n",
      "8\n",
      "7\n",
      "?\n",
      "?\n",
      "8\n",
      "A1\n",
      "\n",
      "\n",
      "\n",
      "Inferred block\n",
      "\n",
      "\n",
      "\n",
      "A0\n",
      "\n",
      "\n",
      "\n",
      "Sequencing +\n",
      "basecalling +\n",
      "motif detection\n",
      "\n",
      "\n",
      "\n",
      "Fig. 1.\n",
      "Combinatorial DNA data storage pipeline used in [2]. First, the data is encoded into blocks, each containing 2 address-carrying symbols (denoted\n",
      "by A0 as A1) and 8 payload-carrying symbols; a payload symbol (shown as a row of the data block) is a combination of 4 distinct motifs chosen out of a\n",
      "library of 8. Second, the DNA pool is synthesised: in each payload synthesis cycle, the 4 chosen motifs corresponding to the synthesised symbol are added\n",
      "to the reaction tube containing a set of growing strands (shown vertically as columns); a random motif is attached to each strand. The bioinformatics pipeline\n",
      "at the reading stage includes sequencing, basecalling, and motif detection. It produces the original data blocks with some motifs missing or interfering due\n",
      "to synthesis and detection errors.\n",
      "\n",
      "\n",
      "\n",
      "achievable performance. In Section III-D, we highlight an\n",
      "important connection between the capacity curves and the\n",
      "fundamental trade-off between the costs of reading and writing\n",
      "based on the total-cost optimisation framework introduced\n",
      "in [10]. Section IV offers a way to approach the fundamental\n",
      "limits of the Coupon Collector Channel using protograph-\n",
      "based non-binary spatially coupled low-density parity-check\n",
      "(SC-LDPC) codes, which we demonstrate numerically in\n",
      "Section V, and describes how to avoid an exponential growth\n",
      "in decoding complexity with the size of the motif library.\n",
      "\n",
      "\n",
      "\n",
      "II. MOTIF-BASED DNA STORAGE\n",
      "\n",
      "\n",
      "\n",
      "A. General Principle\n",
      "\n",
      "\n",
      "\n",
      "The starting point of our work is the combinatorial motif-\n",
      "based DNA storage pipeline used by HelixWorks and de-\n",
      "scribed in detail in [2]. We illustrate it schematically in\n",
      "Fig. 1. Instead of individual nucleotides, synthesis operates\n",
      "with a library of pre-synthesised 50-nucleotide motifs. The\n",
      "size of the library determines the information density; a library\n",
      "of 8 motifs is used in current real-world experiments by\n",
      "HelixWorks, and we use their parameters as our running\n",
      "example throughout this paper without loss of generality. The\n",
      "data is arranged into blocks, where each block contains 2\n",
      "symbols for addressing (denoted by A0 and A1 in Fig. 1)\n",
      "and 8 combinatorial symbols for the payload. Each payload\n",
      "symbol, in turn, is a combination of 4 distinct motifs out\n",
      "of 8 in the library.1 For example, the first payload symbol\n",
      "in the data block illustrated in Fig. 1 is {3, 6, 7, 8} (the\n",
      "symbols themselves are arranged from top to bottom). During\n",
      "synthesis, a block is represented by a set of growing strands.\n",
      "A synthesis cycle incorporates one combinatorial symbol;\n",
      "specifically, the 4 motifs encoding the symbol are added to\n",
      "the pool and appended to the strands. Carrying on with the\n",
      "example in Fig. 1, we show two synthesised strands (a strand\n",
      "is shown as a column): during the first payload synthesis\n",
      "cycle, multiple copies of each of the motifs in the symbol\n",
      "{3, 6, 7, 8} are added to the pool; during the second cycle, the\n",
      "\n",
      "\n",
      "\n",
      "1A symbol used for addressing comprises a single motif from the library;\n",
      "we leave the problem of addressing outside the scope of this paper.\n",
      "\n",
      "\n",
      "\n",
      "motifs {2, 4, 5, 7} are added, and so on. By design, only one\n",
      "out of the four added motifs gets attached to a strand in one\n",
      "cycle; however, there is no control over which one it will be.\n",
      "This fundamental randomness is the key property of motif-\n",
      "based DNA storage. In our running example in Fig. 1, motif 3\n",
      "happens to get attached to the first strand in the first payload\n",
      "synthesis cycle, and motif 8 is attached to the second strand.\n",
      "At the reading stage, the DNA strands are sequenced, and a\n",
      "bioinformatics pipeline that includes basecalling and motif de-\n",
      "tection attempts to reconstruct the stored data blocks. However,\n",
      "limited sequencing coverage and errors during basecalling and\n",
      "motif detection may lead to incomplete data retrieval and\n",
      "erroneous motif detection. Assuming that only the two strands\n",
      "displayed in Fig. 1 are sequenced and that all motifs are\n",
      "correctly detected, only motifs 3 and 8 from the first payload\n",
      "cycle are observed at the reading end of the pipeline--the\n",
      "other two motifs remain unknown and could in principle be\n",
      "any two distinct motifs from the set {1, 2, 4, 5, 6, 7}.\n",
      "\n",
      "\n",
      "\n",
      "B. Experimental Data Set\n",
      "\n",
      "\n",
      "\n",
      "To investigate the empirical behaviour of the synthesis-\n",
      "storage-sequencing pipeline, we analysed a data set gener-\n",
      "ated by a wet-lab experiment conducted by HelixWorks. The\n",
      "experiment synthesised 21 blocks. The resulting DNA pool\n",
      "was sequenced via Oxford Nanopore Technologies (ONT)\n",
      "nanopore sequencing. The generated reads were basecalled and\n",
      "processed by HelixWorks's software pipeline, which detected\n",
      "3.7 million motif occurrences. The pipeline assigns reads\n",
      "to blocks using ONT's barcoding feature based on the two\n",
      "address symbols, and detects payload motifs using exact search\n",
      "for the 50-nucleotide motif sequences in the library.2 Our\n",
      "analysis shows three distinct patterns in the input-output error\n",
      "behaviour, which we discuss in more detail in Section II-C:\n",
      "\n",
      "\n",
      "\n",
      "1) As was highlighted in Section II-A, there is a funda-\n",
      "mental uncertainty about which motif is attached during\n",
      "synthesis and read during sequencing. This means that\n",
      "accumulating the motifs at the reading stage resembles\n",
      "\n",
      "\n",
      "\n",
      "2We are aware that more elaborate motif detection methods are being\n",
      "developed by HelixWorks and collaborators.\n",
      "\n",
      "3\n",
      "\n",
      "\n",
      "\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "0.2\n",
      "\n",
      "\n",
      "\n",
      "0.4\n",
      "\n",
      "\n",
      "\n",
      "0.6\n",
      "\n",
      "\n",
      "\n",
      "0.8\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "Number of reads\n",
      "\n",
      "\n",
      "\n",
      "CDF\n",
      "\n",
      "\n",
      "\n",
      "Simulated\n",
      "\n",
      "\n",
      "\n",
      "Empirical\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "2\n",
      "4\n",
      "3\n",
      "5\n",
      "8\n",
      "6\n",
      "7\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "1 000\n",
      "\n",
      "\n",
      "\n",
      "2 000\n",
      "\n",
      "\n",
      "\n",
      "Motif label\n",
      "\n",
      "\n",
      "\n",
      "Number of occurrences\n",
      "\n",
      "\n",
      "\n",
      "Correct\n",
      "\n",
      "\n",
      "\n",
      "Interfered\n",
      "\n",
      "\n",
      "\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "0.2\n",
      "\n",
      "\n",
      "\n",
      "0.4\n",
      "\n",
      "\n",
      "\n",
      "0.6\n",
      "\n",
      "\n",
      "\n",
      "0.8\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "Number of reads\n",
      "\n",
      "\n",
      "\n",
      "CDF\n",
      "\n",
      "\n",
      "\n",
      "Simulated\n",
      "\n",
      "\n",
      "\n",
      "Empirical\n",
      "\n",
      "\n",
      "\n",
      "2\n",
      "5\n",
      "6\n",
      "4\n",
      "8\n",
      "1\n",
      "7\n",
      "3\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "1000\n",
      "\n",
      "\n",
      "\n",
      "2000\n",
      "\n",
      "\n",
      "\n",
      "Motif label\n",
      "\n",
      "\n",
      "\n",
      "Number of occurrences\n",
      "\n",
      "\n",
      "\n",
      "Correct\n",
      "\n",
      "\n",
      "\n",
      "Interfered\n",
      "\n",
      "\n",
      "\n",
      "Fig. 2.\n",
      "Empirical and simulated uncoded data recovery CDFs and motif histograms. The first two subplots correspond to block 21. The first subplot shows\n",
      "a close match between the simulated (blue curve, circle marker) and empirical (red curve, square marker) CDF. The representative interfered motif histogram\n",
      "for symbol 8 in this block (second subplot, red dashed bars) is relatively even. The last two plots correspond to block 1. A large discrepancy between the\n",
      "simulated and empirical CDF (third plot) is caused by imbalanced interfered motif histogram (last plot, representative symbol 8, red dashed bars).\n",
      "\n",
      "\n",
      "\n",
      "the classical Coupon Collector's Problem, as discussed\n",
      "in [3].\n",
      "2) In addition to the coupon-collecting dynamics of the\n",
      "reading process, we have also observed significant vari-\n",
      "ation in how often the motifs belonging to different\n",
      "locations within a block are detected, which we refer to as\n",
      "cycle visibility. In particular, the first motif in a block is\n",
      "detected 5.7 times more often than the last, with a general\n",
      "monotonically decreasing trend from the beginning to\n",
      "the end of the block. This visibility variation suggests\n",
      "that there is an accumulation of synthesis failures or of\n",
      "detection errors in the pipeline.\n",
      "3) Finally, we observe some interference in the form of mo-\n",
      "tifs which are erroneously detected in symbols that do not\n",
      "contain them. Different motifs are observed at different\n",
      "rates, and this pattern does not appear to be consistent\n",
      "across different blocks. We speculate that these errors are\n",
      "primarily caused by occasional incorrect assignment of\n",
      "entire reads to wrong blocks by ONT's barcoding.\n",
      "\n",
      "\n",
      "\n",
      "C. Channel Simulator\n",
      "To test whether the error patterns described in Section II-B\n",
      "are sufficient to model the dynamics of the sequencing process,\n",
      "we developed a simulator that replicates those patterns using\n",
      "parameters derived from empirical data. For a given block\n",
      "and a given number of reads, we compare the probability of\n",
      "recovering the entire block (with no error-correction coding)\n",
      "for the simulated reads and the reads obtained by sub-sampling\n",
      "the empirical data set down to the required coverage.\n",
      "For each read, the simulator first uses uniform sampling to\n",
      "generate a sequence of 8 motifs, each motif chosen from the\n",
      "4 motifs of the corresponding symbol. Then, it samples from\n",
      "the empirical cycle visibility distribution to determine which\n",
      "(if any) of these motifs are visible at the reading stage. Finally,\n",
      "based on the empirical average interference rate, we emulate\n",
      "interference by uniform sampling from all motifs in the library\n",
      "(in contrast to sampling only from the motifs belonging to the\n",
      "chosen symbol).\n",
      "Fig. 2 focuses on two representative blocks from the empir-\n",
      "ical data set: the first two plots correspond to block 21, and the\n",
      "last two plots correspond to block 1. For each block, we plot\n",
      "the cumulative distribution function (CDF) of the probability\n",
      "of fully recovering the entire block and a representative empir-\n",
      "ical histogram of the number of detected motifs for one of the\n",
      "\n",
      "\n",
      "\n",
      "symbols within the block. In the first (leftmost) subplot (block\n",
      "21), we observe a close match between the simulated (blue\n",
      "curve, circle marker) and empirical (red curve, square marker)\n",
      "CDFs. We picked symbol 8 from that block and plotted the\n",
      "number of times each motif in the library was detected in\n",
      "that location (second subplot), splitting the motifs into those\n",
      "that belong to the chosen symbol (blue bars) and those that\n",
      "do not (red dashed bars). The distribution of interfered motifs\n",
      "(red dashed bars) is relatively even. Conversely, for block 1,\n",
      "where there is a significant mismatch between the simulated\n",
      "and empirical CDFs (third plot), the typical distribution of\n",
      "interfered motifs (symbol 8, last plot, red dashed bars) is\n",
      "imbalanced. We observe this behaviour more generally, with\n",
      "around half of the blocks in the data set showing a close match\n",
      "to the simulated model and the other half showing a significant\n",
      "mismatch with the associated imbalance in interfered motif\n",
      "distributions. We have not discerned any systematic pattern in\n",
      "which blocks exhibit this mismatch.3\n",
      "\n",
      "\n",
      "\n",
      "We conclude that when the pattern of interference is rela-\n",
      "tively uniform, the channel simulator provides a good model\n",
      "of the synthesis-storage-sequencing pipeline. In this work, we\n",
      "focus on the coupon-collecting nature of the channel and as-\n",
      "sume uniform interference, leaving unequal cycle visibility and\n",
      "skewed interference outside the scope of the paper. Tackling\n",
      "these aspects of the pipeline is an interesting direction for\n",
      "future work. We remark that the channel models proposed in\n",
      "Section III aim to reflect the empirically observed behaviour\n",
      "(see Section II-B) while also being applicable to other motif-\n",
      "based DNA storage schemes beyond the particular pipeline\n",
      "employed by HelixWorks.\n",
      "\n",
      "\n",
      "\n",
      "III. COUPON COLLECTOR CHANNELS\n",
      "The simulator described in Section II-C serves as the\n",
      "starting point for channel modelling. We are now ready to give\n",
      "a formal definition of the channel model without interference\n",
      "(we discuss interference in Section III-C). Let there be a set M\n",
      "containing n different motifs and let k be a positive integer\n",
      "such that k < n. Then any subset x M with exactly k\n",
      "motifs is a combinatorial symbol. Let X be the set of possible\n",
      "combinatorial symbols. It is easy to see that |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      ".\n",
      "Transmission through the channel involves a parameter\n",
      "R, which represents the number of reads and must be a\n",
      "\n",
      "\n",
      "\n",
      "3We remark that when we artificially filter out interfered motifs, we observe\n",
      "a near-perfect match between the simulated and empirical CDFs for all blocks.\n",
      "\n",
      "4\n",
      "\n",
      "\n",
      "\n",
      "8\n",
      "7\n",
      "6\n",
      "5\n",
      "4\n",
      "3\n",
      "2\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "n = 8\n",
      "motifs\n",
      "\n",
      "\n",
      "\n",
      "7\n",
      "6\n",
      "3\n",
      "1\n",
      "k = 4\n",
      "chosen\n",
      "CC(n, k, R)\n",
      "\n",
      "\n",
      "\n",
      "Coupon Collector\n",
      "Channel\n",
      "\n",
      "\n",
      "\n",
      "x\n",
      "\n",
      "\n",
      "\n",
      "6\n",
      "1\n",
      "7\n",
      "6\n",
      "1\n",
      "6\n",
      "\n",
      "\n",
      "\n",
      "R = 6\n",
      "reads\n",
      "y\n",
      "\n",
      "\n",
      "\n",
      "?\n",
      "7\n",
      "6\n",
      "1\n",
      "l= 3\n",
      "unique\n",
      "\n",
      "\n",
      "\n",
      "Fig. 3. An example of transmission over the Coupon Collector Channel CC(n = 8, k = 4, R = 6). From a library of n = 8 motifs, a combination of k = 4\n",
      "is chosen for transmission. The Coupon Collector Channel generates R = 6 reads by uniform sampling with replacement among the k chosen motifs. In the\n",
      "illustrated example, only l= 3 motifs out of k = 4 are encountered.\n",
      "\n",
      "\n",
      "\n",
      "positive integer. In its simplest version, the channel is denoted\n",
      "as CC(n, k, R) and represents interference-free sequencing.\n",
      "In this case, when a symbol x is transmitted, the received\n",
      "sequence y is generated by sampling R motifs from x,\n",
      "independently and uniformly at random. Figure 3 illustrates\n",
      "this channel.\n",
      "Notice that the received sequence can be written as y =\n",
      "(y1, . . . , yR) where yi x for all i = 1, . . . , R. This means\n",
      "that the number of unique (i.e., distinct) motifs present in y\n",
      "can range from 1 to L = min{k, R}. We denote by Y the set\n",
      "of all possible received sequences--i.e., the set of R-tuples\n",
      "with at least 1 and at most L different motifs. Further, we say\n",
      "that x and y are compatible if yi x for all i = 1, . . . , R,\n",
      "and we use Y(x) to denote the set of all y that are compatible\n",
      "with x.\n",
      "\n",
      "\n",
      "\n",
      "A. Capacity of the CC(n, k, R) Channel\n",
      "\n",
      "\n",
      "\n",
      "We first give the capacity of the channel as discussed so far.\n",
      "\n",
      "\n",
      "\n",
      "Theorem 1. The capacity of CC(n, k, R) is given by\n",
      "\n",
      "\n",
      "\n",
      "CCC = log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "-k-R\n",
      "L\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "l=1\n",
      "\n",
      "\n",
      "\n",
      "\u0012k\n",
      "l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\u001aR\n",
      "l\n",
      "\n",
      "\n",
      "\n",
      "\u001b\n",
      "l! log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n -l\n",
      "k -l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      ", (1)\n",
      "\n",
      "\n",
      "\n",
      "where L = min{k, R}, and\n",
      "\bR\n",
      "l\n",
      "\t\n",
      "= 1/l! Pl\n",
      "i=0(-1)l-i\u0000l\n",
      "i\n",
      "\u0001\n",
      "iR\n",
      "\n",
      "\n",
      "\n",
      "are the Stirling numbers of the second kind.\n",
      "\n",
      "\n",
      "\n",
      "Before proving this result, we introduce important notation\n",
      "and a useful Lemma. Notice that when a symbol x is trans-\n",
      "mitted, only a compatible y Y(x) can be received, each\n",
      "being equally likely. Now, the set Y(x) can be partitioned\n",
      "into subsets that are easier to work with, defined as\n",
      "\n",
      "\n",
      "\n",
      "Yl(x) = {y Y(x) | y contains lunique motifs} .\n",
      "(2)\n",
      "\n",
      "\n",
      "\n",
      "Notice that {Yl(x)}L\n",
      "l=1 is a partition of Y(x). The following\n",
      "result gives the size of these sets.\n",
      "\n",
      "\n",
      "\n",
      "Lemma 2. Let x X and l{1, . . . , L} with L =\n",
      "min{k, R}. Then it holds that\n",
      "\n",
      "\n",
      "\n",
      "|Yl(x)| =\n",
      "\u0012k\n",
      "l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\u001aR\n",
      "l\n",
      "\n",
      "\n",
      "\n",
      "\u001b\n",
      "l! ,\n",
      "(3)\n",
      "\n",
      "\n",
      "\n",
      "where\n",
      "\bR\n",
      "l\n",
      "\t\n",
      "are Stirling numbers of the second kind.\n",
      "\n",
      "\n",
      "\n",
      "Proof. On the one hand, there are\n",
      "\u0000k\n",
      "l\n",
      "\u0001\n",
      "ways to choose lunique\n",
      "motifs out of k motifs in x. On the other, there are\n",
      "\bR\n",
      "l\n",
      "\t\n",
      "l! ways\n",
      "\n",
      "\n",
      "\n",
      "to distribute R distinct reads among ldistinct motifs (see case\n",
      "3 of The Twelvefold Way in [11, Section 1.9]). Multiplying\n",
      "both factors gives the result.\n",
      "\n",
      "\n",
      "\n",
      "Consider two random variables X, Y such that X repre-\n",
      "sents the transmitted combinatorial symbol and Y represents\n",
      "the corresponding received sequence. We use the following\n",
      "compact notation: P (x) denotes P (X = x), and analogously\n",
      "for P (y). P (x|y) denotes P (X = x|Y = y), and analogously\n",
      "for P (y|x). Finally, P (x, y) denotes P (X = x, Y = y).\n",
      "\n",
      "\n",
      "\n",
      "Proof of Theorem 1. We first prove that the mutual informa-\n",
      "tion I(X; Y ) under the uniform input distribution is equal to\n",
      "the right-hand side of (1). Second, we prove that the uniform\n",
      "distribution is capacity-achieving.\n",
      "\n",
      "\n",
      "\n",
      "Uniform input distribution. Assume X to be uniformly dis-\n",
      "tributed. We expand I(X; Y ) = H(X) -H(X|Y ) and\n",
      "calculate the two entropy terms. First, since |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "and\n",
      "P (x) = 1/|X| for all x, we get\n",
      "\n",
      "\n",
      "\n",
      "H(X) = -\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "P (x) log2 P (x) = log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      ".\n",
      "(4)\n",
      "\n",
      "\n",
      "\n",
      "Second, we need to compute H(X|Y ). By definition of the\n",
      "channel, it is easy to see that P (y|x) = k-R if y Y(x),\n",
      "and P (y|x) = 0 otherwise. Using this, we get\n",
      "\n",
      "\n",
      "\n",
      "H(X|Y ) = -\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "\n",
      "\n",
      "\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "yY\n",
      "P (x, y) log2 P (x|y)\n",
      "\n",
      "\n",
      "\n",
      "= -\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "P (x)\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "yY(x)\n",
      "P (y|x) log2 P (x|y)\n",
      "\n",
      "\n",
      "\n",
      "= -\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "P (x)\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "yY(x)\n",
      "k-R log2 P (x|y) .\n",
      "(5)\n",
      "\n",
      "\n",
      "\n",
      "Notice that P (x|y) only depends on the number of unique\n",
      "motifs within y. Also, it is clear that if y has lunique\n",
      "motifs, then there are\n",
      "\u0000n-l\n",
      "k-l\n",
      "\u0001\n",
      "possible symbols in X that are\n",
      "compatible. That is, P (x|y) = 1/\n",
      "\u0000n-l\n",
      "k-l\n",
      "\u0001\n",
      ". Replacing in (5):\n",
      "\n",
      "\n",
      "\n",
      "H(X|Y ) = k-R X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "P (x)\n",
      "\n",
      "\n",
      "\n",
      "L\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "l=1\n",
      "\n",
      "\n",
      "\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "yYl(x)\n",
      "log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n -l\n",
      "k -l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "\n",
      "\n",
      "\n",
      "= k-R X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "P (x)\n",
      "\n",
      "\n",
      "\n",
      "L\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "l=1\n",
      "|Yl(x)| log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n -l\n",
      "k -l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      ".\n",
      "(6)\n",
      "\n",
      "5\n",
      "\n",
      "\n",
      "\n",
      "We can substitute the value given in (3) for |Yl(x)| into (6).\n",
      "Since the inner sum does not depend on the value of x, we\n",
      "get4\n",
      "\n",
      "\n",
      "\n",
      "H(X|Y ) = k-R\n",
      "L\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "l=1\n",
      "\n",
      "\n",
      "\n",
      "\u0012k\n",
      "l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\u001aR\n",
      "l\n",
      "\n",
      "\n",
      "\n",
      "\u001b\n",
      "l! log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n -l\n",
      "k -l\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      ".\n",
      "(7)\n",
      "\n",
      "\n",
      "\n",
      "Combining (4) and (7), we conclude that under uniform input\n",
      "distribution I(X; Y ) is equal to the right-hand side of (1).\n",
      "#\n",
      "\n",
      "\n",
      "\n",
      "Symmetry. It remains to prove that the uniform input distri-\n",
      "bution maximises I(X; Y ). For this, we rely on the following\n",
      "result from [12]:\n",
      "\n",
      "\n",
      "\n",
      "Lemma 3 ([12, Theorem 4.5.2]). For a symmetric discrete\n",
      "memoryless channel, capacity is achieved by using the inputs\n",
      "with equal probability.\n",
      "\n",
      "\n",
      "\n",
      "We can therefore prove that the uniform input distribution\n",
      "achieves capacity by proving that CC(n, k, R) is symmetric--\n",
      "i.e., that the set of channel outputs can be partitioned in\n",
      "such a way that for each subset the corresponding matrix\n",
      "of transition probabilities P (y|x) has the property that each\n",
      "row is a permutation of each other row and each column is a\n",
      "permutation of each other column [12].\n",
      "For this, consider the partition {Yl}L\n",
      "l=1 of Y, where Ylis\n",
      "the set of all y Y such that y has lunique motifs. Now,\n",
      "let l{1, . . . , L} and consider the transition probabilities\n",
      "P (y|x) where y Yland x X. These probabilities can be\n",
      "written as a matrix Plwhere element Pl\n",
      "ij is the probability of\n",
      "receiving the j-th element of Ylwhen the i-th element of X\n",
      "is transmitted (consider any valid total order over Yland X).\n",
      "The elements Pl\n",
      "ij can take only one of two possible values:\n",
      "Pl\n",
      "ij = k-R if x and y are compatible, and Pl\n",
      "ij = 0 otherwise.\n",
      "For any given y Yl, there are\n",
      "\u0000n-l\n",
      "k-l\n",
      "\u0001\n",
      "compatible x and\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "-\n",
      "\u0000n-l\n",
      "k-l\n",
      "\u0001\n",
      "non-compatible x. On the other hand, for any\n",
      "given x X, there are |Yl(x)| compatible y and |Yl|-|Yl(x)|\n",
      "non-compatible y. In other words, and using Lemma 2 to get\n",
      "the value of |Yl(x)|:\n",
      "\n",
      "\n",
      "\n",
      "* Every column in Plhas\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "entries. Of these,\n",
      "\u0000n-l\n",
      "k-l\n",
      "\u0001\n",
      "\n",
      "\n",
      "\n",
      "have value k-R, and the rest are 0. This is the same\n",
      "for all columns, so it holds that all columns in Plare\n",
      "permutations of each other.\n",
      "\n",
      "\n",
      "\n",
      "* Every row in Plhas |Yl| entries. Of these,\n",
      "\u0000k\n",
      "l\n",
      "R\n",
      "l\n",
      "\t\n",
      "l!\n",
      "have value k-R, and the rest are 0. This is the same for\n",
      "all rows, so it holds that all rows in Plare permutations\n",
      "of each other.\n",
      "\n",
      "\n",
      "\n",
      "These last two conditions constitute the definition of a\n",
      "symmetric channel, since {Yl}L\n",
      "l=1 is a partition of Y. We\n",
      "conclude that the CC(n, k, R) channel is symmetric.\n",
      "#\n",
      "\n",
      "\n",
      "\n",
      "Together, the expression for I(X; Y ) under uniform input\n",
      "distribution and the symmetry of the channel prove Theorem 1.\n",
      "\n",
      "\n",
      "\n",
      "4We remark that (7) can be interpreted as averaging the conditional entropy\n",
      "of X over the distribution of the number of unique motifs lobserved in Y .\n",
      "\n",
      "\n",
      "\n",
      "0\n",
      "2\n",
      "4\n",
      "6\n",
      "8\n",
      "10\n",
      "12\n",
      "14\n",
      "16\n",
      "18\n",
      "20\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "2\n",
      "\n",
      "\n",
      "\n",
      "3\n",
      "\n",
      "\n",
      "\n",
      "4\n",
      "\n",
      "\n",
      "\n",
      "5\n",
      "\n",
      "\n",
      "\n",
      "6\n",
      "\n",
      "\n",
      "\n",
      "7\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "6.1\n",
      "\n",
      "\n",
      "\n",
      "Number of reads R\n",
      "\n",
      "\n",
      "\n",
      "Information density [bits/cycle]\n",
      "\n",
      "\n",
      "\n",
      "CCC(n = 8, k = 4)\n",
      "CNBEC(n = 8, k = 4)\n",
      "\n",
      "\n",
      "\n",
      "SC-LDPC, FER = 4 *10-4\n",
      "\n",
      "\n",
      "\n",
      "Fig. 4. Capacities of CC (blue, circle marker) from (1) and of NBEC (red,\n",
      "square marker) from (9) for (n = 8, k = 4). The horizontal dashed line shows\n",
      "the asymptotic limit for both curves. The region of operation enabled by full\n",
      "processing of partial information is shaded green. The black dot shows the\n",
      "operating point of the coding scheme proposed in Section IV, see Section V\n",
      "for details. The code operates well above the capacity of NBEC.\n",
      "\n",
      "\n",
      "\n",
      "B. Erasure Version of the Coupon Collector Channel\n",
      "\n",
      "\n",
      "\n",
      "Suppose now that the decoder relies only on outputs y that\n",
      "fully determine the corresponding input x--in other words, on\n",
      "y Yk. In this case, the channel effectively reduces to the\n",
      "non-binary erasure channel with erasure probability\n",
      "\n",
      "\n",
      "\n",
      "1 -F(k, R) = 1 -\n",
      "\u001aR\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u001b\n",
      "k! k-R ,\n",
      "(8)\n",
      "\n",
      "\n",
      "\n",
      "where F(k, R) is the CDF in the standard Coupon Collector's\n",
      "Problem, denoting the probability that R trials are sufficient\n",
      "to collect all k coupons [13, Example II.11]. We denote\n",
      "the resulting all-or-nothing version of the Coupon Collector\n",
      "Channel as NBEC(n, k, R).\n",
      "The capacity of NBEC(n, k, R) is a scaled CDF from the\n",
      "Coupon Collector's Problem F(k, R):\n",
      "\n",
      "\n",
      "\n",
      "CNBEC = log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "*\n",
      "\u001aR\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u001b\n",
      "k! k-R .\n",
      "(9)\n",
      "\n",
      "\n",
      "\n",
      "Ignoring partial information available at the output of the\n",
      "Coupon Collector Channel entails a loss in achievable infor-\n",
      "mation rate for a given R, as we illustrate in Fig. 4. The\n",
      "blue curve (circle marker) shows the capacity of CC(n =\n",
      "8, k = 4, R) calculated using (1), and the red curve (square\n",
      "marker) shows the corresponding CNBEC from (9). Both curves\n",
      "tend to log2\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "as R -(horizontal dashed line), where\n",
      "redundancy is no longer required. A key contribution of this\n",
      "paper (see Section IV below) is a coding scheme that uses all\n",
      "information available in y and is thus able to operate above\n",
      "CNBEC (i.e., in the area shaded green in Fig. 4), beyond the\n",
      "reach of the schemes that wait for all k motifs to be observed,\n",
      "including the MDS-based schemes proposed and analysed\n",
      "in [3], [4]. The black dot in Fig. 4 shows the operating point\n",
      "of a particular coding scheme described in detail in Section V,\n",
      "\n",
      "6\n",
      "\n",
      "\n",
      "\n",
      "demonstrating that it is indeed possible to operate in the green-\n",
      "shaded region in practice.\n",
      "\n",
      "\n",
      "\n",
      "C. Coupon Collector Channel with Interference\n",
      "As we have observed in the experimental data set (Sec-\n",
      "tion II-B), the bioinformatics pipeline may sometimes erro-\n",
      "neously detect a motif that was not present in the transmitted\n",
      "symbol x. Such errors are most likely caused by an incorrect\n",
      "assignment of a read to a wrong block and can therefore be\n",
      "though of as cross-block interference.\n",
      "We incorporate this type of error into the Coupon Collector\n",
      "Channel by introducing an auxiliary parameter r that controls\n",
      "the rate of interference. Specifically, given x X, each\n",
      "element yi in y = (y1, . . . , yR) is independently generated as\n",
      "follows: with probability 1-r, yi is sampled from x uniformly\n",
      "at random as for the standard CC(n, k, R) introduced earlier;\n",
      "otherwise, with probability r, yi is sampled uniformly at\n",
      "random from the entire motif library M without regard to\n",
      "x, modelling interference from a different block. Importantly,\n",
      "we do not sample from M\\x but from the entire M when we\n",
      "emulate interference because an interfered read from a wrong\n",
      "block may happen to contain a motif that is also present in\n",
      "x by coincidence. We denote the Coupon Collector Channel\n",
      "with interference by CC(n, k, R, r).\n",
      "We can estimate the capacity of CC(n, k, R, r) using uni-\n",
      "form input distribution (our model of interference does not\n",
      "break the symmetry of the channel). We rewrite I(X; Y ) as\n",
      "\n",
      "\n",
      "\n",
      "I(X; Y ) = H(X) -E\n",
      "h\n",
      "H(X|Y = y)\n",
      "i\n",
      ",\n",
      "(10)\n",
      "\n",
      "\n",
      "\n",
      "calculate H(X) as in (4), and estimate the expectation using\n",
      "samples y generated via Monte-Carlo simulations of the\n",
      "channel. Specifically, we generate a number of samples y,\n",
      "calculate H(X|Y = y) for each sample as described below,\n",
      "average the obtained values to estimate E\n",
      "\u0002\n",
      "H(X|Y = y)\n",
      "\u0003\n",
      ",\n",
      "and use it in (10) to obtain an estimation of the capacity of\n",
      "CC(n, k, R, r).\n",
      "It remains to show how to calculate\n",
      "\n",
      "\n",
      "\n",
      "H(X|Y = y) = -\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "xX\n",
      "P (x|y) log2 P (x|y) ,\n",
      "(11)\n",
      "\n",
      "\n",
      "\n",
      "which amounts to calculating P (x|y) for a given y and all\n",
      "x X. We observe that\n",
      "\n",
      "\n",
      "\n",
      "P (x|y) = P (y|x) P (x)\n",
      "\n",
      "\n",
      "\n",
      "P (y) P (y|x) ,\n",
      "(12)\n",
      "\n",
      "\n",
      "\n",
      "since both P (x) and P (y) are constant for uniform input\n",
      "distribution and a fixed y. This means that P (x|y) can be\n",
      "obtained by normalising P (y|x) by P\n",
      "\n",
      "\n",
      "\n",
      "x'X P (y|x'), and we\n",
      "are left with the task of calculating P (y|x). We have\n",
      "\n",
      "\n",
      "\n",
      "P (y|x) =\n",
      "\n",
      "\n",
      "\n",
      "R\n",
      "Y\n",
      "\n",
      "\n",
      "\n",
      "i=1\n",
      "P (yi|x)\n",
      "(13)\n",
      "\n",
      "\n",
      "\n",
      "with\n",
      "\n",
      "\n",
      "\n",
      "P (yi|x) = r * P (yi|x, int.) + (1 -r) * P (yi|x, no int.) (14)\n",
      "\n",
      "\n",
      "\n",
      "=\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pin = r * 1\n",
      "\n",
      "\n",
      "\n",
      "n + (1 -r) * 1\n",
      "\n",
      "\n",
      "\n",
      "k ,\n",
      "if yi x ;\n",
      "\n",
      "\n",
      "\n",
      "pout = r * 1\n",
      "\n",
      "\n",
      "\n",
      "n ,\n",
      "otherwise.\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "5\n",
      "10\n",
      "15\n",
      "20\n",
      "25\n",
      "30\n",
      "35\n",
      "40\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      "2\n",
      "\n",
      "\n",
      "\n",
      "3\n",
      "\n",
      "\n",
      "\n",
      "4\n",
      "\n",
      "\n",
      "\n",
      "5\n",
      "\n",
      "\n",
      "\n",
      "6\n",
      "\n",
      "\n",
      "\n",
      "7\n",
      "\n",
      "\n",
      "\n",
      "Information density [bits/cycle]\n",
      "\n",
      "\n",
      "\n",
      "CCC(r = 0)\n",
      "CCC(r = 0.078)\n",
      "CNBEC(t = 2)\n",
      "CNBEC(t = 3)\n",
      "SC-LDPC,\n",
      "FER = 3 *10-4\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "5\n",
      "10\n",
      "15\n",
      "20\n",
      "25\n",
      "30\n",
      "35\n",
      "40\n",
      "10-4\n",
      "\n",
      "\n",
      "\n",
      "10-3\n",
      "\n",
      "\n",
      "\n",
      "10-2\n",
      "\n",
      "\n",
      "\n",
      "10-1\n",
      "\n",
      "\n",
      "\n",
      "Number of reads R\n",
      "\n",
      "\n",
      "\n",
      "Symbol error probability\n",
      "\n",
      "\n",
      "\n",
      "t = 2\n",
      "t = 3\n",
      "\n",
      "\n",
      "\n",
      "Fig. 5.\n",
      "Top: capacity curves for the Coupon Collector Channel with\n",
      "(n = 8, k = 4), no interference (black curve, no marker) and r = 0.078\n",
      "(blue curve, circle marker). For the hard-decision decoding scheme that waits\n",
      "for t copies of the top k motifs before making a decision, the capacity of\n",
      "the associated non-binary erasure channel (neglecting interference) is shown\n",
      "for t = 2 (purple curve, square marker) and t = 3 (red curve, triangle\n",
      "marker). Hard-decision schemes that mitigate interference by accumulating\n",
      "t > 1 copies of top-k motifs cannot operate in the green-shaded region\n",
      "even when no interference is present. In contrast, the black dot shows the\n",
      "operating point of the coding scheme proposed in Section IV for r = 0.078,\n",
      "see Section V for details. Soft information processing allows the scheme to\n",
      "operate in the green-shaded region. Bottom: symbol-level substitution rates\n",
      "for the same hard-decision schemes with t = 2 and t = 3 as in the top figure\n",
      "(same line styles) and r = 0.078. These errors are not accounted for in the\n",
      "corresponding capacity curves, making the green-shaded region underestimate\n",
      "the potential gain from soft-decision information processing.\n",
      "\n",
      "\n",
      "\n",
      "Putting it all together, for a given y and each x, we need\n",
      "to count the number s = | {i : yi x} | of motifs in y that\n",
      "match the motifs in x, calculate\n",
      "\n",
      "\n",
      "\n",
      "P (y|x) = ps\n",
      "in pR-s\n",
      "out\n",
      ",\n",
      "(15)\n",
      "\n",
      "\n",
      "\n",
      "normalise the resulting vector of probabilities [P (y|x)]xX to\n",
      "sum up to 1, and use [P (x|y)]xX thus obtained when cal-\n",
      "culating (11) to estimate (10) using simulated samples y. We\n",
      "remark that this procedure does not scale well with increasing\n",
      "n and k because it requires considering all |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "possible\n",
      "x to calculate (11); providing a more efficient way to estimate\n",
      "(or compute exactly) the capacity of CC(n, k, R, r) is therefore\n",
      "an interesting research problem that falls outside the scope of\n",
      "this work.\n",
      "In Fig. 5 (top), we plot the resulting capacity estimation for\n",
      "CC(n = 8, k = 4, R, r) with r = 0.078, the level of interfer-\n",
      "ence observed in the empirical data set (see Section II-B),\n",
      "as the blue curve with the circle marker. Naturally, inter-\n",
      "\n",
      "7\n",
      "\n",
      "\n",
      "\n",
      "ference induces a penalty with respect to the corresponding\n",
      "interference-free channel (black curve with no marker).\n",
      "The MDS-based schemes proposed and analysed in [3],\n",
      "[4] do not attempt to optimally process information available\n",
      "in y. Instead, they make a hard decision on the symbol x\n",
      "by choosing the k most frequent motifs in y. To mitigate\n",
      "interference, they require at least t > 1 copies of each motif in\n",
      "the top k to be present in y and declare erasure if this condition\n",
      "is not met [4]. We estimate an upper bound on the capacity\n",
      "associated with this decoding approach by simulating the\n",
      "Coupon Collector Channel with no interference and declaring\n",
      "erasure whenever y does not contain at least t copies of each\n",
      "of the k motifs in x. Having simulated the associated erasure\n",
      "probability,5 which we denote by t, we calculate the capacity\n",
      "of the resulting non-binary erasure channel as\n",
      "\n",
      "\n",
      "\n",
      "CNBEC(n, k, t) = log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "* (1 -t) .\n",
      "(16)\n",
      "\n",
      "\n",
      "\n",
      "The resulting capacity estimations are shown in Fig. 5 (top)\n",
      "for t = 2 (purple curve with the square marker) and t = 3\n",
      "(red curve with a triangle marker). We observe that even\n",
      "without interference and for t = 2, the hard-decision approach\n",
      "analysed in [4] exhibits a gap to CC(n, k, R, r = 0.078),\n",
      "which we shade green in Fig. 5 (top). The coding scheme\n",
      "we propose in Section IV operates in this region (black dot).\n",
      "Moreover, the hard-decision schemes yield a symbol-level\n",
      "substitution--i.e., decide on a symbol ^x = x--when a motif\n",
      "from M \\ x is one of the top-k most encountered motifs in y\n",
      "and is present in at least t copies. We simulate the rate of such\n",
      "symbol-level substitutions for CC(n = 8, k = 4, R, r = 0.078)\n",
      "and plot the corresponding curves in Fig. 5 (bottom) for t = 2\n",
      "and t = 3 using the same line styles as in Fig. 5 (top). These\n",
      "errors are not included in the analysis in [4] because easily\n",
      "distinguishable motifs and the choice of t > 1 render them\n",
      "unlikely. While we agree with this in principle, we emphasise\n",
      "that r also depends on the probability of assigning a read to\n",
      "a wrong block, and we observe that the relatively high level\n",
      "of interference r = 0.078 in the empirical data set results in\n",
      "non-negligible symbol error rates of about 10-2 for t = 2 and\n",
      "10-3 for t = 3.\n",
      "\n",
      "\n",
      "\n",
      "D. Read-Write Cost Trade-Off\n",
      "\n",
      "\n",
      "\n",
      "The read-write cost trade-off (first introduced explicitly\n",
      "in [14]) is a fundamental property of DNA storage: the more\n",
      "redundancy is introduced to the system in the form of error-\n",
      "correcting codes (i.e., the lower the code rate), the fewer reads\n",
      "will suffice at the sequencing stage to reconstruct the data\n",
      "(i.e., the lower the required coverage). Before we proceed to\n",
      "designing specific error-correcting schemes for the Coupon\n",
      "Collector Channels (Section IV), we would like to highlight\n",
      "an important connection between the capacity curves and\n",
      "the read-write cost trade-off. We will use the notation and\n",
      "terminology of the Coupon Collector Channels and motif-\n",
      "based synthesis, although the same reasoning applies in equal\n",
      "measure to more standard nucleotide-by-nucleotide storage\n",
      "schemes.\n",
      "\n",
      "\n",
      "\n",
      "5An alternative way to compute this probability based on the Markov chain\n",
      "formulation of the coupon-collecting process is provided in [4].\n",
      "\n",
      "\n",
      "\n",
      "-l\n",
      "0\n",
      "R*\n",
      "\n",
      "\n",
      "\n",
      "C(R*)\n",
      "\n",
      "\n",
      "\n",
      "log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012n\n",
      "k\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "\n",
      "\n",
      "\n",
      "(l + R) C'(R*)\n",
      "\n",
      "\n",
      "\n",
      "C(R)\n",
      "\n",
      "\n",
      "\n",
      "(l + R*) C'(R*) = C(R*)\n",
      "\n",
      "\n",
      "\n",
      "R\n",
      "\n",
      "\n",
      "\n",
      "Fig. 6.\n",
      "Capacity and the read-write cost trade-off. The optimum point\n",
      "(R*, C(R*)) is obtained according to (19).\n",
      "\n",
      "\n",
      "\n",
      "The task of the system designer is to choose the code rate\n",
      "and the target coverage at the optimum point of the read-write\n",
      "cost trade-off. The location of the optimum point depends on\n",
      "the costs of synthesising and sequencing a symbol. Denote\n",
      "by pw [$/cycle] the cost of running one synthesis cycle and\n",
      "writing one combinatorial symbol, and by pr [$/read] the cost\n",
      "of sequencing one motif belonging to that symbol. Suppose\n",
      "C(R) is a capacity curve that shows the highest information\n",
      "density (in information bits per cycle) achievable for a given\n",
      "number of reads R--implying that it would take 1/C(R)\n",
      "cycles to store one information bit and R/C(R) reads to\n",
      "retrieve it. The goal is to minimise the total cost of DNA\n",
      "storage per information bit,\n",
      "\n",
      "\n",
      "\n",
      "P(R) = pw *\n",
      "1\n",
      "C(R) + pr *\n",
      "R\n",
      "C(R)\n",
      "[$/inf. bit]\n",
      "\n",
      "\n",
      "\n",
      "= pr * l + R\n",
      "\n",
      "\n",
      "\n",
      "C(R) ,\n",
      "(17)\n",
      "\n",
      "\n",
      "\n",
      "where l = pw/pr describes how much more expensive synthe-\n",
      "sis is relative to sequencing. This optimisation is introduced\n",
      "in [10]; in the following, we offer its visual interpretation\n",
      "based on the capacity curves. Assuming a well-behaved\n",
      "C(R),6 the optimum is obtained when\n",
      "\n",
      "\n",
      "\n",
      "P '(R) = pr * C(R) -(l + R) C'(R)\n",
      "\n",
      "\n",
      "\n",
      "C(R)2\n",
      "= 0 ,\n",
      "(18)\n",
      "\n",
      "\n",
      "\n",
      "implying that at the optimum R, denoted by R*,\n",
      "\n",
      "\n",
      "\n",
      "C(R*) = (l + R*) C'(R*) .\n",
      "(19)\n",
      "\n",
      "\n",
      "\n",
      "This optimisation is illustrated graphically in Fig. 6. For a\n",
      "candidate R*, we can visualise a line L(R) = (l+R) C'(R*)\n",
      "that crosses the horizontal axis at -l and has the same slope\n",
      "as C(R) at R*. The larger R*is, the more the capacity\n",
      "curve C(R) flattens as it tends to the theoretical maximum\n",
      "log2\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      ", and the line L(R) lowers down accordingly. The\n",
      "condition (19) means that the optimum R*corresponds to the\n",
      "case when the line just touches the capacity curve, and R*is\n",
      "equal to the point at which they meet.7 Naturally, as the cost of\n",
      "\n",
      "\n",
      "\n",
      "6A continuous version of CCC(R) can be obtained by modelling R as a\n",
      "random variable and plotting CCC against E [R].\n",
      "7The same procedure can be equivalently interpreted as finding a point at\n",
      "which the tangent line to C(R) crosses the horizontal axis at -l.\n",
      "\n",
      "8\n",
      "\n",
      "\n",
      "\n",
      "writing relative to reading increases and -l slides leftwards,\n",
      "L(R) touches C(R) at larger R and higher C(R).\n",
      "Overall, we believe that this framework elucidates the\n",
      "relationship between capacity curves and the read-write cost\n",
      "trade-off: for example, it can be used to visually estimate what\n",
      "kind of read-write cost ratio is needed for a code rate to be\n",
      "relevant.\n",
      "\n",
      "\n",
      "\n",
      "IV. NON-BINARY LDPC CODES OVER THE COUPON\n",
      "COLLECTOR CHANNELS\n",
      "\n",
      "\n",
      "\n",
      "To introduce error-correction coding into the motif-based\n",
      "DNA storage pipeline, we map motif combinations in X to\n",
      "elements of the prime field GF(q), where q is chosen to be\n",
      "the highest prime below |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      ". We use lexicographic\n",
      "ordering of motif combinations and leave the last\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "-q\n",
      "combinations unassigned.\n",
      "We use non-binary LDPC (NB-LDPC) codes. An NB-LDPC\n",
      "code can be defined via a parity-check matrix H = (hij)MxN\n",
      "over GF(q) as the set of vectors xcw = (x1, . . . , xN)T\n",
      "\n",
      "\n",
      "\n",
      "that satisfy Hxcw = 0. In order to describe the decoding\n",
      "algorithms, we define the sets Ni = {j : hij = 0} and\n",
      "Mj = {i : hij = 0}. The corresponding Tanner graph has an\n",
      "edge between variable node (VN) vj and check node (CN) ci if\n",
      "and only if j Ni (or, equivalently, i Mj). In other words,\n",
      "Ni and Mj represent the local neighbourhoods of CN ci and\n",
      "VN vj, respectively. For the version of the channel without\n",
      "interference, we decode using a set-based belief propagation\n",
      "algorithm. For the version of the channel with interference,\n",
      "we use a probability-based belief propagation method dubbed\n",
      "QSPA (for q-ary sum-product algorithm) [15], [16, Section\n",
      "14.2].\n",
      "We remark that using only a subset of possible combina-\n",
      "tions has two important implications: First, it incurs a loss\n",
      "in the overall information transmission rate by a factor of\n",
      "log2 q/ log2\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "--for our running example of (n = 8, k = 4)\n",
      "and\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "= 70, the closest prime is 67, and the (multiplicative)\n",
      "rate loss factor is 0.99, implying a 1% loss. Second, it breaks\n",
      "the symmetry of the channel--e.g., as we observe in our\n",
      "simulations, the all-zero codeword becomes more resilient to\n",
      "errors than other codewords. To circumvent this problem, we\n",
      "re-introduce symmetry in the channel by adding a pseudo-\n",
      "random mask modulo |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "to the codeword before\n",
      "transmission and undoing the effects of masking during de-\n",
      "coding by the appropriate re-labelling of symbol possibilities\n",
      "and symbol likelihoods. In the following description of the\n",
      "decoding algorithms, we omit the trivial details of such re-\n",
      "labelling for clarity.\n",
      "\n",
      "\n",
      "\n",
      "A. No Interference\n",
      "\n",
      "\n",
      "\n",
      "For the Coupon Collector Channel without interference\n",
      "(defined at the beginning of Section III), we employ a set-\n",
      "based decoder reminiscent of the message-passing Sudoku\n",
      "solver described in [17]. The decoder keeps track of the set of\n",
      "possibilities Sj for each symbol and iteratively updates these\n",
      "sets based on parity-check constraints, progressively reducing\n",
      "the uncertainty about the transmitted symbols. First, Sj is\n",
      "initialised as the set of symbols in X compatible with the\n",
      "\n",
      "\n",
      "\n",
      "corresponding channel output yj. If the number of unique\n",
      "motifs in yj is l(i.e., if yj Yl), then |Sj| =\n",
      "\u0000n-l\n",
      "k-l\n",
      "\u0001\n",
      ". The set\n",
      "of initial symbol possibilities Sj for each symbol is passed\n",
      "to the corresponding VN vj. We then iteratively apply the\n",
      "following procedure:\n",
      "\n",
      "\n",
      "\n",
      "1) CN update. For convenience, we define a pairwise sum\n",
      "operation over finite sets of elements in GF(q) as\n",
      "A + B = {a + b | a A, b B} and a scalar multiplica-\n",
      "tion operation as c * A = {c * a | a A}. For each VN vj,\n",
      "we generate the set of symbol possibilities Sij based on\n",
      "CN ci as\n",
      "Sij = -h-1\n",
      "ij\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "tNi\\{j}\n",
      "hitSt .\n",
      "(20)\n",
      "\n",
      "\n",
      "\n",
      "2) VN update. Using Sij computed above, we update the set\n",
      "of symbol possibilities for each VN vj as\n",
      "\n",
      "\n",
      "\n",
      "Sj -\n",
      "\\\n",
      "\n",
      "\n",
      "\n",
      "iMj\n",
      "Sij Sj .\n",
      "(21)\n",
      "\n",
      "\n",
      "\n",
      "3) Check for termination. We repeat the decoding iterations\n",
      "(20)-(21) until either all VNs are resolved--i.e., the set\n",
      "of symbol possibilities for each VN is reduced to one\n",
      "element--or the total number of symbol possibilities re-\n",
      "mains the same for each VN, indicating decoding failure.\n",
      "\n",
      "\n",
      "\n",
      "B. Interference\n",
      "\n",
      "\n",
      "\n",
      "For the Coupon Collector Channel with interference (Sec-\n",
      "tion III-C), the belief propagation algorithm aims to approxi-\n",
      "mate the symbol-wise posterior distribution of the transmitted\n",
      "codeword given the received information. This entails approxi-\n",
      "mating a length-q probability mass function for each codeword\n",
      "symbol. The output of the decoding algorithm is then given\n",
      "by the symbol-wise maximum a posteriori estimate, provided\n",
      "the resulting codeword satisfies all parity constraints.\n",
      "The QSPA works by performing message-passing over the\n",
      "Tanner graph corresponding to H. Messages passed along the\n",
      "edges of the Tanner graph consist of q-ary probability mass\n",
      "functions. Let Qij = (Qa\n",
      "ij)aGF(q) denote the message passed\n",
      "from vj to ci, and Sij = (Sa\n",
      "ij)aGF(q) denote the message in\n",
      "the other direction, from ci to vj.\n",
      "Decoding starts by initialising all Qij vectors. For all i, j\n",
      "such that hij = 0, we set Qij = Pj, where Pj = (P a\n",
      "j )aGF(q)\n",
      "denotes the vector of conditional probabilities P (xj = a|yj)\n",
      "for xj, the j-th symbol in the codeword xcw, and the corre-\n",
      "sponding channel output yj. The Pj vectors are computed as\n",
      "described in Section III-C using (12)-(15). After initialisation,\n",
      "decoding proceeds by repeatedly iterating the following steps:\n",
      "\n",
      "\n",
      "\n",
      "1) CN update. For each a GF(q), the update is\n",
      "\n",
      "\n",
      "\n",
      "Sa\n",
      "ij =\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "s\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      " X\n",
      "\n",
      "\n",
      "\n",
      "t\n",
      "hitst + hija = 0\n",
      "\n",
      "\n",
      "\n",
      "! Y\n",
      "\n",
      "\n",
      "\n",
      "t\n",
      "Qst\n",
      "it ,\n",
      "(22)\n",
      "\n",
      "\n",
      "\n",
      "where the iteration variable s = (st) runs over all q-ary\n",
      "sequences of length |Ni| -1, indexed by Ni \\ {j} for\n",
      "notational convenience, and t Ni \\ {j}.\n",
      "2) VN update. For each a GF(q), compute first\n",
      "\n",
      "\n",
      "\n",
      "Qa\n",
      "ij = P a\n",
      "j\n",
      "Y\n",
      "\n",
      "\n",
      "\n",
      "tMj\\{i}\n",
      "Sa\n",
      "it\n",
      "(23)\n",
      "\n",
      "9\n",
      "\n",
      "\n",
      "\n",
      "and then normalise each Qij vector so that it sums to 1.\n",
      "3) Check for termination. Compute z = (zj) via\n",
      "\n",
      "\n",
      "\n",
      "zj = arg\n",
      "max\n",
      "aGF(q) P a\n",
      "j\n",
      "Y\n",
      "\n",
      "\n",
      "\n",
      "tMj\n",
      "Sa\n",
      "it .\n",
      "(24)\n",
      "\n",
      "\n",
      "\n",
      "If Hz = 0, return z as the decoded codeword. Otherwise,\n",
      "go back to step 1.\n",
      "This procedure may fail to output a codeword, so we set\n",
      "a maximum number of iterations, after which the algorithm\n",
      "declares decoding failure.\n",
      "It is important to point out that step 1 can be implemented\n",
      "efficiently via the Fast Fourier Transform (FFT) algorithm.\n",
      "Assume for simplicity that H consists only of 0 and 1\n",
      "elements, as is the case in our experiments (Section V), but\n",
      "the argument can be generalised to any H (see [16, Section\n",
      "14.2]). Then, the CN update rule in (22) can be rewritten as\n",
      "\n",
      "\n",
      "\n",
      "Sa\n",
      "ij =\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "s\n",
      "1\n",
      "\n",
      "\n",
      "\n",
      " X\n",
      "\n",
      "\n",
      "\n",
      "t\n",
      "st = -a\n",
      "\n",
      "\n",
      "\n",
      "! Y\n",
      "\n",
      "\n",
      "\n",
      "t\n",
      "Qst\n",
      "it .\n",
      "(25)\n",
      "\n",
      "\n",
      "\n",
      "This can be computed by first evaluating\n",
      "\n",
      "\n",
      "\n",
      "~Sij =\n",
      "O\n",
      "\n",
      "\n",
      "\n",
      "tNi\\{j}\n",
      "Qit ,\n",
      "(26)\n",
      "\n",
      "\n",
      "\n",
      "where denotes the circular convolution, and then re-indexing\n",
      "~Sij to obtain Sa\n",
      "ij =\n",
      "~S-a\n",
      "ij\n",
      "for all a GF(q). The FFT\n",
      "can be used to efficiently implement the series of circular\n",
      "convolutions in (26).\n",
      "\n",
      "\n",
      "\n",
      "C. Protograph-Based Non-Binary SC-LDPC Codes\n",
      "\n",
      "\n",
      "\n",
      "Spatial coupling is a powerful technique that allows the\n",
      "construction of capacity-achieving codes under suboptimal but\n",
      "computationally efficient belief propagation (BP) decoding.\n",
      "It involves interconnecting a sequence of Tanner graphs of\n",
      "the underlying uncoupled codes and creating a coupled chain\n",
      "where the termination boundaries contain CNs of lower av-\n",
      "erage degrees. The CNs at the boundaries help ignite the\n",
      "decoding process, and reliable information propagates from\n",
      "the boundaries of the chain inwards in a wave-like fash-\n",
      "ion. Importantly, spatially coupled low-density parity-check\n",
      "(SC-LDPC) codes [18], [19] have been proven to universally\n",
      "achieve capacity over the broad class of binary-input memory-\n",
      "less symmetric channels [20], which means that a sequence of\n",
      "codes that achieves capacity for one such channel is proven to\n",
      "achieve capacity for all of them. Even though universality has\n",
      "not been formally proven for non-binary channels, it motivates\n",
      "us to choose non-binary SC-LDPC codes for Coupon Collec-\n",
      "tor Channels and achieve excellent performance (Section V)\n",
      "without channel-tailored code optimisation.\n",
      "We use the classical protograph-based regular SC-LDPC\n",
      "code construction proposed in [21]. To obtain the protograph\n",
      "of a terminated (dv, dc, Lp, Np) SC-LDPC code, we start from\n",
      "a \"spatial\" sequence of Lp copies of the protograph of the\n",
      "(dv, dc)-regular LDPC code and rearrange their edges such that\n",
      "a variable node (VN) at position i {1, . . . , Lp} is connected\n",
      "to dv consecutive positions (i, i + 1, . . . , i + dv -1), and a\n",
      "check node (CN) at position i is connected to dc/dv VNs at\n",
      "each of the dv positions (i -dv + 1, . . . , i), where we assume\n",
      "\n",
      "\n",
      "\n",
      "0\n",
      "50\n",
      "100\n",
      "150\n",
      "200\n",
      "0\n",
      "\n",
      "\n",
      "\n",
      "20\n",
      "\n",
      "\n",
      "\n",
      "40\n",
      "\n",
      "\n",
      "\n",
      "60\n",
      "\n",
      "\n",
      "\n",
      "80\n",
      "\n",
      "\n",
      "\n",
      "Number of reads R\n",
      "\n",
      "\n",
      "\n",
      "Information density [bits/cycle]\n",
      "\n",
      "\n",
      "\n",
      "CC(n = 80, k = 40)\n",
      "10x CC(n = 8, k = 4)\n",
      "\n",
      "\n",
      "\n",
      "Fig. 7. Capacity of the CC(n = 80, k = 40, R) channel (blue curve, circle\n",
      "marker) and of the library-splitting scheme with a = 10 (red curve, square\n",
      "marker). The gap between the curves illustrates the price paid for manageable\n",
      "computational complexity.\n",
      "\n",
      "\n",
      "\n",
      "dc/dv is an integer. To connect the overhanging edges, dv -1\n",
      "spatial positions that contain CNs only are appended to the\n",
      "sequence. The Tanner graph is generated from the protograph\n",
      "using the standard copy-and-permute operation, whereby the\n",
      "SC-LDPC protograph is copied\n",
      "dv\n",
      "dc Np times and the copies\n",
      "of the same edge in the original protograph are permuted.\n",
      "The resulting Tanner graph contains N = LpNp VNs and\n",
      "M = dv\n",
      "\n",
      "\n",
      "\n",
      "dc Np(Lp + dv -1) CNs, yielding the design rate of\n",
      "\n",
      "\n",
      "\n",
      "RSC = 1-dv\n",
      "\n",
      "\n",
      "\n",
      "dc\n",
      "*(Lp + dv -1)\n",
      "\n",
      "\n",
      "\n",
      "Lp\n",
      "= 1-dv\n",
      "\n",
      "\n",
      "\n",
      "dc\n",
      "\n",
      "\n",
      "\n",
      "\u0012\n",
      "1 + dv -1\n",
      "\n",
      "\n",
      "\n",
      "Lp\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      ", (27)\n",
      "\n",
      "\n",
      "\n",
      "which tends to the rate of the underlying LDPC ensemble,\n",
      "1 -dv/dc , as Lp -. We refer the reader to [21], [22]\n",
      "for a more general description of protograph-based SC-LDPC\n",
      "codes.\n",
      "\n",
      "\n",
      "\n",
      "D. Combinatorial Explosion in Decoding Complexity\n",
      "\n",
      "\n",
      "\n",
      "One of the advantages of combinatorial motif-based DNA\n",
      "storage is that it enables a nearly linear growth in information\n",
      "density with the size of the motif library, in contrast with\n",
      "the more standard nucleotide-by-nucleotide synthesis where\n",
      "doubling the size of the alphabet only adds one bit to the\n",
      "maximum information density. Specifically, let us assume in\n",
      "this section that k = n/2 so that |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "is maximised\n",
      "for a given n. As n -, the maximum information density\n",
      "grows as\n",
      "\n",
      "\n",
      "\n",
      "log2 |X| = log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012 n\n",
      "n/2\n",
      "\n",
      "\n",
      "\n",
      "\u0013\n",
      "~log2\n",
      "\n",
      "\n",
      "\n",
      "\"\n",
      "2n\n",
      "p\n",
      "\n",
      "\n",
      "\n",
      "np/2\n",
      "\n",
      "\n",
      "\n",
      "#\n",
      "\n",
      "\n",
      "\n",
      "= n -log2\n",
      "p\n",
      "\n",
      "\n",
      "\n",
      "np/2 = n -O(log2 n) .\n",
      "\n",
      "\n",
      "\n",
      "(28)\n",
      "\n",
      "\n",
      "\n",
      "Furthermore, increasing k = n/2 requires a higher R at\n",
      "the reading stage and makes the system more relevant for\n",
      "higher synthesis-sequencing cost ratios l, as we elaborated\n",
      "in Section III-D.\n",
      "However, increasing k and n presents a significant challenge\n",
      "to the decoding methods proposed in Sections IV-A and IV-B\n",
      "above: their decoding complexity depends on |X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "\n",
      "10\n",
      "\n",
      "\n",
      "\n",
      "that grows exponentially with increasing (n, k = n/2) as\n",
      "shown in (28)--in other words, the very mechanism that\n",
      "makes combinatorial motif-based storage attractive renders\n",
      "straightforward application of the proposed coding methods\n",
      "impractical.\n",
      "We propose the following approach to circumvent this\n",
      "problem: Suppose the largest alphabet the decoder can handle\n",
      "is a choice of ~k out of ~n motifs. We split the target library\n",
      "of n = a~n motifs into a groups of ~n and restrict the choice\n",
      "of k = a~k motifs to ~k motifs within each group instead of a\n",
      "choice of any k motifs out of n. Such restriction lowers the\n",
      "number of possible motifs and hence information density, but\n",
      "on the other hand it allows to operate the decoder as if each\n",
      "cycle contains a independent symbols from a smaller alphabet.\n",
      "Crucially, since the total number of possible symbols in a cycle\n",
      "becomes\n",
      "\n",
      "\n",
      "\n",
      "|X| =\n",
      "\u0012~n\n",
      "~k\n",
      "\n",
      "\n",
      "\n",
      "\u0013n/~n\n",
      ",\n",
      "(29)\n",
      "\n",
      "\n",
      "\n",
      "the linear growth in maximum information density is retained:\n",
      "\n",
      "\n",
      "\n",
      "log2 |X| = log2\n",
      "\n",
      "\n",
      "\n",
      "\u0012~n\n",
      "~k\n",
      "\n",
      "\n",
      "\n",
      "\u0013n/~n\n",
      "= n * log2\n",
      "\u0000~n\n",
      "~k\n",
      "\u0001\n",
      "\n",
      "\n",
      "\n",
      "~n\n",
      ",\n",
      "(30)\n",
      "\n",
      "\n",
      "\n",
      "albeit with a slower growth rate of log2\n",
      "\u0000~n\n",
      "~k\n",
      "\u0001\n",
      "/~n instead of 1 as\n",
      "in (28). For example, for ~n = 8 and ~k = 4 as in our examples\n",
      "earlier, the penalty term in (30) evaluates to about 0.77, which\n",
      "means that asymptotically the scheme loses around 23% in\n",
      "maximum information density relative to the computationally\n",
      "infeasible processing of all\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "possible combinations.\n",
      "The capacity of the library-splitting scheme is obtained\n",
      "by averaging the capacity of an individual sub-channel over\n",
      "the binomial distribution of the effective number of reads.\n",
      "Specifically, taking the interference-free Coupon Collector\n",
      "Channel as an example, we get\n",
      "\n",
      "\n",
      "\n",
      "CaxCC = a\n",
      "\n",
      "\n",
      "\n",
      "R\n",
      "X\n",
      "\n",
      "\n",
      "\n",
      "r=1\n",
      "CCC(~n, ~k, r)\n",
      "\u0012R\n",
      "i\n",
      "\n",
      "\n",
      "\n",
      "\u0013 \u00121\n",
      "\n",
      "\n",
      "\n",
      "a\n",
      "\n",
      "\n",
      "\n",
      "\u0013r \u0012\n",
      "1 -1\n",
      "\n",
      "\n",
      "\n",
      "a\n",
      "\n",
      "\n",
      "\n",
      "\u0013R-r\n",
      "(31)\n",
      "\n",
      "\n",
      "\n",
      "with CCC(~n, ~k, r) denoting the capacity of CC(n = ~n, k =\n",
      "~k, R = r) evaluated via (1).\n",
      "Fig. 7 compares the capacity curve for the CC(n = 80, k =\n",
      "40, R) channel (blue curve with the circle marker) obtained\n",
      "using (1) with the capacity curve for the case where the\n",
      "library of n = 80 motifs is split into a = 10 groups of\n",
      "~n = 8 motifs each with ~k = 4 (red curve with the square\n",
      "marker) obtained using (31). We observe that maintaining a\n",
      "manageable decoding complexity using the proposed scheme\n",
      "induces a penalty on information density. The penalty grows\n",
      "for increasing R and eventually stabilises at the asymptotic\n",
      "value dictated by (30)--in this case at around a quarter of\n",
      "the maximum information density. It is an interesting open\n",
      "problem to reduce this gap by, e.g., constructing a coding\n",
      "scheme that allows decoding without keeping track of all\n",
      "|X| =\n",
      "\u0000n\n",
      "k\n",
      "\u0001\n",
      "possible symbols and thus avoids combinatorial\n",
      "explosion in decoding complexity altogether.\n",
      "\n",
      "\n",
      "\n",
      "V. NUMERICAL RESULTS\n",
      "To demonstrate the potential of the proposed coding scheme,\n",
      "we use the protograph-based regular (dv = 4, dc = 12, Lp =\n",
      "\n",
      "\n",
      "\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "10-4\n",
      "\n",
      "\n",
      "\n",
      "10-3\n",
      "\n",
      "\n",
      "\n",
      "10-2\n",
      "\n",
      "\n",
      "\n",
      "10-1\n",
      "\n",
      "\n",
      "\n",
      "100\n",
      "\n",
      "\n",
      "\n",
      "Number of reads R\n",
      "\n",
      "\n",
      "\n",
      "FER\n",
      "\n",
      "\n",
      "\n",
      "Sim., no int.\n",
      "Sim., int.\n",
      "Lim, no int.\n",
      "Lim, int.\n",
      "\n",
      "\n",
      "\n",
      "Fig. 8.\n",
      "FER for the (dv = 4, dc = 12, Lp = 50, Np = 1002) SC-LDPC\n",
      "code ensemble with q = 67 over the Coupon Collector Channel with (n =\n",
      "8, k = 4), with no interference (blue, circle markers) and with r = 0.078\n",
      "(red, square markers). For the employed overall information transmission rate\n",
      "of 3.92 bits/cycle, the minimum achievable R (based on the capacity curves\n",
      "in Figs. 4 and 5) for the channel with and without interference is shown as\n",
      "the dash-dotted and dashed vertical line, respectively.\n",
      "\n",
      "\n",
      "\n",
      "50, Np = 1002) SC-LDPC code ensemble (see Section IV-C).\n",
      "For simplicity, we only consider binary entries in the parity-\n",
      "check matrix H. We use q = 67, targeting (n = 8, k = 4),\n",
      "which results in an overall information rate of RSC * log2 q \n",
      "3.92, with RSC calculated in (27).\n",
      "Fig. 8 shows the simulated frame error rate (FER) of this\n",
      "code ensemble over CC(n = 8, k = 4, R, r = 0) (blue line\n",
      "with circle markers) and CC(n = 8, k = 4, R, r = 0.078)\n",
      "(red line with square markers). The minimum achievable R\n",
      "for these two channels and the employed code rate of 3.92 is\n",
      "shown as the dashed and dash-dotted line, respectively.8\n",
      "\n",
      "\n",
      "\n",
      "We observe that the ensemble operates at a FER below\n",
      "10-3 for R = 6 over the channel without interference and for\n",
      "R = 11 over the channel with r = 0.078. The corresponding\n",
      "points are shown in Figs. 4 and 5 (black dots), well within\n",
      "the green-shaded regions. Notably, the capacity-approaching\n",
      "performance (especially for the interference-free case) was\n",
      "achieved without ensemble optimisation, which demonstrates\n",
      "the power of spatial coupling in conjunction with channel-\n",
      "tailored message-passing decoding.\n",
      "\n",
      "\n",
      "\n",
      "VI. CONCLUSION\n",
      "\n",
      "\n",
      "\n",
      "We explore a practically interesting DNA storage scheme\n",
      "based on combinatorial motif-based synthesis. We analyse a\n",
      "data set from a real-world experiment and devise a model of\n",
      "the channel that mimics the empirically observed behaviour.\n",
      "At the core of the channel model is the Coupon Collector's\n",
      "Problem, which was highlighted in a similar setting earlier by\n",
      "Preuss et al. in [3], [4]. In contrast to their work, however,\n",
      "here we propose to process partial information available out\n",
      "of the channel instead of limiting decoding only to fully recov-\n",
      "ered symbols, which allows us to approach the fundamental\n",
      "\n",
      "\n",
      "\n",
      "8The minimum achievable R is the smallest R for which C(R) > 3.92.\n",
      "\n",
      "11\n",
      "\n",
      "\n",
      "\n",
      "capacity limits that are outside of the reach of the earlier\n",
      "schemes. We show how to calculate those fundamental limits\n",
      "and how to reason about the associated trade-off between the\n",
      "read and write costs. We devise an error-correcting scheme\n",
      "that approaches these limits in practice and show how to\n",
      "circumvent the combinatorial complexity explosion in our\n",
      "decoding algorithm at the cost of some additional redundancy.\n",
      "Overall, the paper offers a powerful framework to reason about\n",
      "and incorporate error correcting coding in one of the most\n",
      "promising approaches to DNA data storage.\n",
      "\n",
      "\n",
      "\n",
      "REFERENCES\n",
      "\n",
      "\n",
      "\n",
      "[1] L. Ceze, J. Nivala, and K. Strauss, \"Molecular digital data storage using\n",
      "DNA,\" Nature Reviews Genetics, vol. 20, no. 8, pp. 456-466, May 2019.\n",
      "[2] Y. Yan, N. Pinnamaneni, S. Chalapati, C. Crosbie, and R. Appuswamy,\n",
      "\"Scaling logical density of DNA storage with enzymatically-ligated\n",
      "composite motifs,\" Scientific Reports, vol. 13, no. 1, p. 15978, Sep.\n",
      "2023.\n",
      "[3] I. Preuss, M. Rosenberg, Z. Yakhini, and L. Anavy, \"Efficient DNA-\n",
      "based data storage using shortmer combinatorial encoding,\" Scientific\n",
      "Reports, vol. 14, no. 1, p. 7731, Apr. 2024.\n",
      "[4] I. Preuss, B. Galili, Z. Yakhini, and L. Anavy, \"Sequencing coverage\n",
      "analysis for combinatorial DNA-based storage systems,\" IEEE Trans.\n",
      "Mol. Biol. Multi-Scale Commun., vol. 10, no. 2, pp. 297-316, Jun. 2024.\n",
      "[5] T. Cohen and E. Yaakobi, \"Optimizing the decoding probability and\n",
      "coverage ratio of composite DNA,\" in Proc. IEEE Int. Symp. Inf. Theory\n",
      "(ISIT), Athens, Greece, Jul. 2024.\n",
      "[6] D. Bar-Lev, O. Sabary, R. Gabrys, and E. Yaakobi, \"Cover your bases:\n",
      "How to minimize the sequencing coverage in DNA storage systems,\" in\n",
      "Proc. IEEE Int. Symp. Inf. Theory (ISIT), Taipei, Taiwan, Jun. 2023, pp.\n",
      "370-375.\n",
      "[7] A. Gruica, D. Bar-Lev, A. Ravagnani, and E. Yaakobi, \"A combinatorial\n",
      "perspective on random access efficiency for DNA storage,\" in Proc.\n",
      "IEEE Int. Symp. Inf. Theory (ISIT), Athens, Greece, Jul. 2024.\n",
      "[8] H. Abraham, R. Gabrys, and E. Yaakobi, \"Covering all bases: The next\n",
      "inning in DNA sequencing efficiency,\" in Proc. IEEE Int. Symp. Inf.\n",
      "Theory (ISIT), Athens, Greece, Jul. 2024.\n",
      "\n",
      "\n",
      "\n",
      "[9] O. Sabary, I. Preuss, R. Gabrys, Z. Yakhini, L. Anavy, and E. Yaakobi,\n",
      "\"Error-correcting codes for combinatorial composite DNA,\" in Proc.\n",
      "IEEE Int. Symp. Inf. Theory (ISIT), Athens, Greece, Jul. 2024.\n",
      "[10] R. Heckel, I. Shomorony, K. Ramchandran, and D. N. C. Tse, \"Funda-\n",
      "mental limits of DNA storage systems,\" in Proc. IEEE Int. Symp. Inf.\n",
      "Theory, Aachen, Germany, Jun. 2017, pp. 3130-3134.\n",
      "[11] R. P. Stanley, Enumerative Combinatorics, 2nd ed.\n",
      "Cambridge Univer-\n",
      "sity Press, 2011, vol. 1.\n",
      "[12] R. G. Gallager, Information Theory and Reliable Communication. John\n",
      "Wiley & Sons, Inc., 1968.\n",
      "[13] P. Flajolet and R. Sedgewick, Analytic combinatorics.\n",
      "Cambridge\n",
      "University Press, 2009.\n",
      "[14] S. Chandak, K. Tatwawadi, B. Lau, M. Kubit, J. Mardia, J. Neu,\n",
      "P. Griffin, M. Wootters, T. Weissman, and H. Ji, \"Improved read/write\n",
      "cost tradeoff in DNA-based data storage using LDPC codes,\" in Proc.\n",
      "Annu. Allerton Conf. Commun., Control, Comput., Monticello, IL, USA,\n",
      "Sep. 2019, pp. 147-156.\n",
      "[15] M. C. Davey and D. J. C. MacKay, \"Low density parity check codes\n",
      "over GF(q),\" IEEE Commun. Lett., vol. 2, no. 6, pp. 165-167, Jun. 1998.\n",
      "[16] W. E. Ryan and S. Lin, Channel Codes.\n",
      "Cambridge University Press,\n",
      "2009.\n",
      "[17] C. Atkins and J. Sayir, \"Density evolution for SUDOKU codes on the\n",
      "erasure channel,\" in 2014 8th International Symposium on Turbo Codes\n",
      "and Iterative Information Processing (ISTC), Bremen, Germany, Aug.\n",
      "2014, pp. 233-237.\n",
      "[18] A. Jimen'ez Feltstr\"om and K. S. Zigangirov, \"Time-varying periodic\n",
      "convolutional codes with low-density parity-check matrix,\" IEEE Trans.\n",
      "Inf. Theory, vol. 45, no. 6, pp. 2181-2191, Sep. 1999.\n",
      "[19] M. Lentmaier, A. Sridharan, D. J. Costello, and K. S. Zigangirov,\n",
      "\"Iterative decoding threshold analysis for LDPC convolutional codes,\"\n",
      "IEEE Trans. Inf. Theory, vol. 56, no. 10, pp. 5274-5289, Oct. 2010.\n",
      "[20] S. Kudekar, T. Richardson, and R. L. Urbanke, \"Spatially coupled\n",
      "ensembles universally achieve capacity under belief propagation,\" IEEE\n",
      "Trans. Inf. Theory, vol. 59, no. 12, pp. 7761-7813, Dec. 2013.\n",
      "[21] A. Sridharan, D. Truhachev, M. Lentmaier, D. J. Costello, and K. S.\n",
      "Zigangirov, \"Distance bounds for an ensemble of LDPC convolutional\n",
      "codes,\" IEEE Trans. Inf. Theory, vol. 53, no. 12, pp. 4537-4555, Dec.\n",
      "2007.\n",
      "[22] A. R. Iyengar, M. Papaleo, P. H. Siegel, J. K. Wolf, A. Vanelli-Coralli,\n",
      "and G. E. Corazza, \"Windowed decoding of protograph-based LDPC\n",
      "convolutional codes over erasure channels,\" IEEE Trans. Inf. Theory,\n",
      "vol. 58, no. 4, pp. 2303-2320, Apr. 2012.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from unidecode import unidecode\n",
    "\n",
    "output = []\n",
    "for page in doc:\n",
    "    output += page.get_text(\"blocks\")\n",
    "previous_block_id = 0 # Set a variable to mark the block id\n",
    "for block in output:\n",
    "    if block[6] == 0: # We only take the text\n",
    "        if previous_block_id != block[5]: # Compare the block number\n",
    "            print(\"\\n\")\n",
    "        plain_text = unidecode(block[4])\n",
    "        print(plain_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "43d93557",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:19: SyntaxWarning: invalid escape sequence '\\('\n",
      "<>:19: SyntaxWarning: invalid escape sequence '\\('\n",
      "C:\\Users\\Parv\\AppData\\Local\\Temp\\ipykernel_27480\\2512976593.py:19: SyntaxWarning: invalid escape sequence '\\('\n",
      "  if re.sub(\"[\\(\\[].*?[\\)\\]]\", \"\", text).isupper():\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pandas as pd\n",
    "\n",
    "import re\n",
    "spans = pd.DataFrame(columns=['xmin', 'ymin', 'xmax', 'ymax', 'text', 'tag'])\n",
    "rows = []\n",
    "for page_num, blocks in block_dict.items():\n",
    "    for block in blocks:\n",
    "        if block['type'] == 0:\n",
    "            for line in block['lines']:\n",
    "                for span in line['spans']:\n",
    "                    xmin, ymin, xmax, ymax = list(span['bbox'])\n",
    "                    font_size = span['size']\n",
    "                    text = unidecode(span['text'])\n",
    "                    span_font = span['font']\n",
    "                    is_upper = False\n",
    "                    is_bold = False\n",
    "                    if \"bold\" in span_font.lower():\n",
    "                        is_bold = True\n",
    "                    if re.sub(\"[\\(\\[].*?[\\)\\]]\", \"\", text).isupper():\n",
    "                        is_upper = True\n",
    "                    if text.replace(\" \",\"\") !=  \"\":\n",
    "                        rows.append((xmin, ymin, xmax, ymax, text,                              is_upper, is_bold, span_font, font_size))\n",
    "                        span_df = pd.DataFrame(rows, columns=['xmin','ymin','xmax','ymax', 'text', 'is_upper','is_bold','span_font', 'font_size'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "32a8af78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>xmin</th>\n",
       "      <th>ymin</th>\n",
       "      <th>xmax</th>\n",
       "      <th>ymax</th>\n",
       "      <th>text</th>\n",
       "      <th>is_upper</th>\n",
       "      <th>is_bold</th>\n",
       "      <th>span_font</th>\n",
       "      <th>font_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>135.619019</td>\n",
       "      <td>369.739471</td>\n",
       "      <td>150.722321</td>\n",
       "      <td>379.702087</td>\n",
       "      <td>I. I</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NimbusRomNo9L-Regu</td>\n",
       "      <td>9.962600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>151.221024</td>\n",
       "      <td>371.250580</td>\n",
       "      <td>213.371811</td>\n",
       "      <td>379.220673</td>\n",
       "      <td>NTRODUCTION</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NimbusRomNo9L-Regu</td>\n",
       "      <td>7.970100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>209.111603</td>\n",
       "      <td>72.000130</td>\n",
       "      <td>216.001755</td>\n",
       "      <td>80.021255</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CMMI7</td>\n",
       "      <td>8.021126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>209.111603</td>\n",
       "      <td>60.589195</td>\n",
       "      <td>216.001755</td>\n",
       "      <td>68.610321</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CMMI7</td>\n",
       "      <td>8.021126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>329.744781</td>\n",
       "      <td>72.000130</td>\n",
       "      <td>336.634918</td>\n",
       "      <td>80.021255</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CMMI7</td>\n",
       "      <td>8.021126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4448</th>\n",
       "      <td>521.108093</td>\n",
       "      <td>739.905334</td>\n",
       "      <td>527.484192</td>\n",
       "      <td>747.875488</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CMMI8</td>\n",
       "      <td>7.970100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4466</th>\n",
       "      <td>146.598022</td>\n",
       "      <td>189.884598</td>\n",
       "      <td>153.243073</td>\n",
       "      <td>199.847198</td>\n",
       "      <td>R</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NimbusRomNo9L-Regu</td>\n",
       "      <td>9.962600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4467</th>\n",
       "      <td>153.741028</td>\n",
       "      <td>191.395691</td>\n",
       "      <td>202.398483</td>\n",
       "      <td>199.365784</td>\n",
       "      <td>EFERENCES</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NimbusRomNo9L-Regu</td>\n",
       "      <td>7.970100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4469</th>\n",
       "      <td>67.223030</td>\n",
       "      <td>215.156677</td>\n",
       "      <td>89.180649</td>\n",
       "      <td>223.126770</td>\n",
       "      <td>DNA,\"</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NimbusRomNo9L-Regu</td>\n",
       "      <td>7.970100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4565</th>\n",
       "      <td>543.293884</td>\n",
       "      <td>345.176483</td>\n",
       "      <td>563.036377</td>\n",
       "      <td>353.146576</td>\n",
       "      <td>IEEE</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NimbusRomNo9L-ReguItal</td>\n",
       "      <td>7.970100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>600 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            xmin        ymin        xmax        ymax         text  is_upper  \\\n",
       "31    135.619019  369.739471  150.722321  379.702087         I. I      True   \n",
       "32    151.221024  371.250580  213.371811  379.220673  NTRODUCTION      True   \n",
       "174   209.111603   72.000130  216.001755   80.021255            A      True   \n",
       "177   209.111603   60.589195  216.001755   68.610321            A      True   \n",
       "196   329.744781   72.000130  336.634918   80.021255            A      True   \n",
       "...          ...         ...         ...         ...          ...       ...   \n",
       "4448  521.108093  739.905334  527.484192  747.875488            R      True   \n",
       "4466  146.598022  189.884598  153.243073  199.847198            R      True   \n",
       "4467  153.741028  191.395691  202.398483  199.365784    EFERENCES      True   \n",
       "4469   67.223030  215.156677   89.180649  223.126770        DNA,\"      True   \n",
       "4565  543.293884  345.176483  563.036377  353.146576         IEEE      True   \n",
       "\n",
       "      is_bold               span_font  font_size  \n",
       "31      False      NimbusRomNo9L-Regu   9.962600  \n",
       "32      False      NimbusRomNo9L-Regu   7.970100  \n",
       "174     False                   CMMI7   8.021126  \n",
       "177     False                   CMMI7   8.021126  \n",
       "196     False                   CMMI7   8.021126  \n",
       "...       ...                     ...        ...  \n",
       "4448    False                   CMMI8   7.970100  \n",
       "4466    False      NimbusRomNo9L-Regu   9.962600  \n",
       "4467    False      NimbusRomNo9L-Regu   7.970100  \n",
       "4469    False      NimbusRomNo9L-Regu   7.970100  \n",
       "4565    False  NimbusRomNo9L-ReguItal   7.970100  \n",
       "\n",
       "[600 rows x 9 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "span_df.loc[span_df['is_upper'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "bd02843c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>xmin</th>\n",
       "      <th>ymin</th>\n",
       "      <th>xmax</th>\n",
       "      <th>ymax</th>\n",
       "      <th>text</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [xmin, ymin, xmax, ymax, text, tag]\n",
       "Index: []"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "7313f862",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Quad(Point(266.5853576660156, 86.52171325683594), Point(528.1879272460938, 86.52171325683594), Point(266.5853576660156, 110.43201446533203), Point(528.1879272460938, 110.43201446533203)),\n",
       " Quad(Point(122.23931884765625, 599.9053344726562), Point(231.31983947753906, 599.9053344726562), Point(122.23931884765625, 609.867919921875), Point(231.31983947753906, 609.867919921875)),\n",
       " Quad(Point(311.97796630859375, 594.8651733398438), Point(419.82305908203125, 594.8651733398438), Point(311.97796630859375, 604.8277587890625), Point(419.82305908203125, 604.8277587890625))]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.search('Motif-Based DNA Storage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "90555c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "page = doc.load_page(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "b392fb16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<div id=\"page0\" style=\"width:612.0pt;height:792.0pt\">\n",
      "<p style=\"top:25.6pt;left:559.5pt;line-height:7.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:7.0pt;color:#000000\">1</span></p>\n",
      "<p style=\"top:57.6pt;left:103.5pt;line-height:23.9pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:23.9pt;color:#000000\">Coding Over Coupon Collector Channels</span></p>\n",
      "<p style=\"top:85.5pt;left:83.8pt;line-height:23.9pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:23.9pt;color:#000000\">for Combinatorial Motif-Based DNA Storage</span></p>\n",
      "<p style=\"top:119.2pt;left:87.7pt;line-height:11.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:11.0pt;color:#000000\">Roman Sokolovskii, Parv Agarwal, Luis Alberto Croquevielle, Zijian Zhou, and Thomas Heinis</span></p>\n",
      "<p style=\"top:176.8pt;left:58.9pt;line-height:9.0pt\"><b><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">Abstract</span></i></b><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">&#x2014;Encoding</span></b></p>\n",
      "<p style=\"top:176.8pt;left:142.6pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">information</span></b></p>\n",
      "<p style=\"top:176.8pt;left:195.7pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">in</span></b></p>\n",
      "<p style=\"top:176.8pt;left:210.6pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">combinations</span></b></p>\n",
      "<p style=\"top:176.8pt;left:269.3pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">of</span></b></p>\n",
      "<p style=\"top:176.8pt;left:284.3pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">pre-</span></b></p>\n",
      "<p style=\"top:186.8pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">synthesised deoxyribonucleic acid (DNA) strands (referred to as</span></b></p>\n",
      "<p style=\"top:196.7pt;left:49.0pt;line-height:9.0pt\"><b><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">motifs</span></i></b><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">) is an interesting approach to DNA storage that could</span></b></p>\n",
      "<p style=\"top:206.7pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">potentially circumvent the prohibitive costs of nucleotide-by-</span></b></p>\n",
      "<p style=\"top:216.7pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">nucleotide DNA synthesis. Based on our analysis of an empirical</span></b></p>\n",
      "<p style=\"top:226.6pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">data set from HelixWorks, we propose two channel models for</span></b></p>\n",
      "<p style=\"top:236.6pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">this setup (with and without interference) and analyse their</span></b></p>\n",
      "<p style=\"top:246.6pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">fundamental limits. We propose a coding scheme that approaches</span></b></p>\n",
      "<p style=\"top:256.5pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">those limits by leveraging all information available at the output</span></b></p>\n",
      "<p style=\"top:266.5pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">of the channel, in contrast to earlier schemes developed for</span></b></p>\n",
      "<p style=\"top:276.4pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">a similar setup by Preuss</span></b><b><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\"> et al.</span></i></b><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\"> We highlight an important</span></b></p>\n",
      "<p style=\"top:286.4pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">connection between channel capacity curves and the fundamental</span></b></p>\n",
      "<p style=\"top:296.4pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">trade-off between synthesis (writing) and sequencing (reading)</span></b></p>\n",
      "<p style=\"top:306.3pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">costs, and offer a way to mitigate an exponential growth in</span></b></p>\n",
      "<p style=\"top:316.3pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">decoding complexity with the size of the motif library.</span></b></p>\n",
      "<p style=\"top:333.7pt;left:58.9pt;line-height:9.0pt\"><b><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">Index Terms&#x2014;</span></i></b><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">DNA storage, coupon collector&#x2019;s problem, spa-</span></b></p>\n",
      "<p style=\"top:345.7pt;left:49.0pt;line-height:9.0pt\"><b><span style=\"font-family:NimbusRomNo9L,serif;font-size:9.0pt;color:#000000\">tially coupled LDPC codes.</span></b></p>\n",
      "<p style=\"top:369.3pt;left:135.6pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">I. I</span><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">NTRODUCTION</span></p>\n",
      "<p style=\"top:384.4pt;left:58.9pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">The density and durability of deoxyribonucleic acid (DNA)</span></p>\n",
      "<p style=\"top:396.3pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">make it an attractive medium for archival storage. However,</span></p>\n",
      "<p style=\"top:408.3pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">despite numerous demonstrations of the feasibility of DNA</span></p>\n",
      "<p style=\"top:420.2pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">storage (we refer the reader to [1] for a comprehensive</span></p>\n",
      "<p style=\"top:432.2pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">overview), prohibitive cost of DNA synthesis hinders its</span></p>\n",
      "<p style=\"top:444.1pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">widespread adoption. A promising alternative to the most com-</span></p>\n",
      "<p style=\"top:456.1pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">monly used (and expensive) nucleotide-by-nucleotide DNA</span></p>\n",
      "<p style=\"top:468.0pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">synthesis is encoding information in combinations of pre-</span></p>\n",
      "<p style=\"top:480.0pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">synthesised oligonucleotides (referred to as</span><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> motifs</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">) from a</span></p>\n",
      "<p style=\"top:492.0pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">fixed motif library [2]. We refer to those combinations of mo-</span></p>\n",
      "<p style=\"top:503.9pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">tifs as</span><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> combinatorial symbols</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">. During synthesis (writing), the</span></p>\n",
      "<p style=\"top:515.9pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">selected combinations of motifs are added to a pool of growing</span></p>\n",
      "<p style=\"top:527.8pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">DNA strands, with no control over which of those motifs is</span></p>\n",
      "<p style=\"top:539.8pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">attached to any given strand, and therefore over which of those</span></p>\n",
      "<p style=\"top:551.7pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">motifs will be encountered during sequencing (reading). The</span></p>\n",
      "<p style=\"top:563.7pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">exponential growth in the number of combinations allows for</span></p>\n",
      "<p style=\"top:575.6pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">a linear growth in information density with the size of the</span></p>\n",
      "<p style=\"top:587.6pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">motif library, at the cost of additional sequencing.</span></p>\n",
      "<p style=\"top:599.5pt;left:58.9pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">Combinatorial motif-based DNA storage is a refreshing</span></p>\n",
      "<p style=\"top:611.4pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">setup from a coding-theoretic perspective with little research</span></p>\n",
      "<p style=\"top:623.4pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">conducted to date. In [3], Preuss</span><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> et al.</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> emphasise an important</span></p>\n",
      "<p style=\"top:635.4pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">connection between the process of accumulating the motifs</span></p>\n",
      "<p style=\"top:647.3pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">during sequencing with the classical Coupon Collector&#x2019;s Prob-</span></p>\n",
      "<p style=\"top:659.3pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">lem. The authors employ maximum distance separable (MDS)</span></p>\n",
      "<p style=\"top:671.2pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">codes and rely on fully accumulated symbols for decoding. In</span></p>\n",
      "<p style=\"top:683.2pt;left:49.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">other words, the coding scheme requires only a subset of the</span></p>\n",
      "<p style=\"top:703.6pt;left:56.9pt;line-height:8.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">This work was funded by MoSS (grant 101058035).</span></p>\n",
      "<p style=\"top:712.6pt;left:56.9pt;line-height:8.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">The authors are with the Department of Computing, Imperial College</span></p>\n",
      "<p style=\"top:721.6pt;left:49.0pt;line-height:8.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">London, SW7 2AZ London, UK (email: r.sokolovskii@imperial.ac.uk;</span></p>\n",
      "<p style=\"top:730.5pt;left:49.0pt;line-height:8.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">p.agarwal23@imperial.ac.uk;</span></p>\n",
      "<p style=\"top:730.5pt;left:192.2pt;line-height:8.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">a.croquevielle22@imperial.ac.uk;</span></p>\n",
      "<p style=\"top:739.5pt;left:49.0pt;line-height:8.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:8.0pt;color:#000000\">zijian.zhou18@imperial.ac.uk; t.heinis@imperial.ac.uk).</span></p>\n",
      "<p style=\"top:176.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">combinatorial symbols to be recovered in order to decode a</span></p>\n",
      "<p style=\"top:188.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">codeword, but these symbols must be recovered fully (i.e.,</span></p>\n",
      "<p style=\"top:199.9pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">all motifs constituting these symbols must be encountered</span></p>\n",
      "<p style=\"top:211.9pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">during sequencing) for decoding to be possible. This MDS-</span></p>\n",
      "<p style=\"top:223.8pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">based scheme is subsequently analysed in detail in a follow-</span></p>\n",
      "<p style=\"top:235.8pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">up paper [4] that proposes a computational framework for</span></p>\n",
      "<p style=\"top:247.8pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">choosing the parameters of the coding scheme based on</span></p>\n",
      "<p style=\"top:259.7pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">coverage and reliability constraints. For a similar setup, [5]</span></p>\n",
      "<p style=\"top:271.7pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">gives the expected number of reads required to reconstruct</span></p>\n",
      "<p style=\"top:283.6pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">information. We remark that in the case of standard (as op-</span></p>\n",
      "<p style=\"top:295.6pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">posed to combinatorial motif-based) synthesis, the connection</span></p>\n",
      "<p style=\"top:307.5pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">between DNA sequencing and the Coupon Collector&#x2019;s Problem</span></p>\n",
      "<p style=\"top:319.5pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">is formally introduced in [6]; a random-access version of the</span></p>\n",
      "<p style=\"top:331.4pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">problem is further investigated in [7], [8].</span></p>\n",
      "<p style=\"top:343.4pt;left:321.9pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">The approach in [3], [4] has two important limitations: First,</span></p>\n",
      "<p style=\"top:355.3pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">as we elaborate in Section III, using only fully recovered</span></p>\n",
      "<p style=\"top:367.3pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">symbols during decoding reduces the pipeline to an erasure</span></p>\n",
      "<p style=\"top:379.3pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">channel; processing partial combinations offers a promising</span></p>\n",
      "<p style=\"top:391.2pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">way to improve decoding by leveraging all available informa-</span></p>\n",
      "<p style=\"top:403.2pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">tion. Second, the setup in [4] mitigates possible</span><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> interference</span></i></p>\n",
      "<p style=\"top:415.1pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">(i.e., the detection of a motif that is not a part of the chosen</span></p>\n",
      "<p style=\"top:427.1pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">combination) by requiring several copies of the candidate</span></p>\n",
      "<p style=\"top:439.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">motifs to be encountered and declaring erasure otherwise. This</span></p>\n",
      "<p style=\"top:451.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">further increases the required coverage&#x2014;again, processing full</span></p>\n",
      "<p style=\"top:462.9pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">available probabilistic information could significantly improve</span></p>\n",
      "<p style=\"top:474.9pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">decoding performance.</span></p>\n",
      "<p style=\"top:486.9pt;left:321.9pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">The first work to make use of partial combinations is [9],</span></p>\n",
      "<p style=\"top:498.8pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">which studies adversarial code constructions that provably</span></p>\n",
      "<p style=\"top:510.8pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">recover from up to</span><i><span style=\"font-family:CMMI10,serif;font-size:10.0pt;color:#000000\"> e</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> missing motifs in up to</span><i><span style=\"font-family:CMMI10,serif;font-size:10.0pt;color:#000000\"> d</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> symbols</span></p>\n",
      "<p style=\"top:522.7pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">(the possibility of interference is not considered). The work</span></p>\n",
      "<p style=\"top:534.7pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">provides bounds on the size of such codes and offers an</span></p>\n",
      "<p style=\"top:546.6pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">explicit encoding scheme for correcting up to</span><i><span style=\"font-family:CMMI10,serif;font-size:10.0pt;color:#000000\"> e</span></i><span style=\"font-family:CMR10,serif;font-size:10.0pt;color:#000000\"> = 1</span><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> missing</span></p>\n",
      "<p style=\"top:558.6pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">motif (with</span><i><span style=\"font-family:CMMI10,serif;font-size:10.0pt;color:#000000\"> d</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> remaining a configurable parameter) when the</span></p>\n",
      "<p style=\"top:570.5pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">block length does not exceed the size of the motif library.</span></p>\n",
      "<p style=\"top:582.5pt;left:321.9pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">In this work, we consider error correction for combinatorial</span></p>\n",
      "<p style=\"top:594.5pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">motif-based DNA storage beyond the MDS-based schemes</span></p>\n",
      "<p style=\"top:606.4pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">proposed and analysed in [3], [4], aiming to use</span><i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> all</span></i><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\"> information</span></p>\n",
      "<p style=\"top:618.4pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">available at the output of the channel. Based on our analysis of</span></p>\n",
      "<p style=\"top:630.3pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">the experimental data set provided by HelixWorks (Section II),</span></p>\n",
      "<p style=\"top:642.3pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">we formally introduce a channel model that mimics the empir-</span></p>\n",
      "<p style=\"top:654.2pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">ically observed behaviour, which we call the Coupon Collector</span></p>\n",
      "<p style=\"top:666.2pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">Channel, either with or without interference (Section III). We</span></p>\n",
      "<p style=\"top:678.1pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">derive the capacity for the interference-free Coupon Collector</span></p>\n",
      "<p style=\"top:690.1pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">Channel (Section III-A) and its erasure-channel counterpart</span></p>\n",
      "<p style=\"top:702.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">(Section III-B), as well as propose a computational method</span></p>\n",
      "<p style=\"top:714.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">to estimate the capacity of the channel with interference</span></p>\n",
      "<p style=\"top:726.0pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">(Section III-C). We demonstrate that relying only on fully</span></p>\n",
      "<p style=\"top:737.9pt;left:312.0pt;line-height:10.0pt\"><span style=\"font-family:NimbusRomNo9L,serif;font-size:10.0pt;color:#000000\">accumulated combinations imposes a fundamental limit on</span></p>\n",
      "<p style=\"top:539.0pt;left:10.9pt;line-height:20.0pt\"><span style=\"font-family:Times New Roman,serif;font-size:20.0pt;color:#7f807f\">arXiv:2406.04141v3  [cs.IT]  4 Dec 2024</span></p>\n",
      "</div>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(page.get_text(\"html\", sort=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "cc55ce5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FontInfos',\n",
       " 'FormFonts',\n",
       " 'Graftmaps',\n",
       " 'InsertedImages',\n",
       " 'ShownPages',\n",
       " '__class__',\n",
       " '__contains__',\n",
       " '__delattr__',\n",
       " '__delitem__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__enter__',\n",
       " '__eq__',\n",
       " '__exit__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getitem__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__len__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__slots__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '_addFormFont',\n",
       " '_delToC',\n",
       " '_deleteObject',\n",
       " '_delete_page',\n",
       " '_do_links',\n",
       " '_do_widgets',\n",
       " '_embeddedFileGet',\n",
       " '_embeddedFileIndex',\n",
       " '_embfile_add',\n",
       " '_embfile_del',\n",
       " '_embfile_info',\n",
       " '_embfile_names',\n",
       " '_embfile_upd',\n",
       " '_extend_toc_items',\n",
       " '_forget_page',\n",
       " '_getMetadata',\n",
       " '_getOLRootNumber',\n",
       " '_getPDFfileid',\n",
       " '_getPageInfo',\n",
       " '_get_char_widths',\n",
       " '_get_page_labels',\n",
       " '_graft_id',\n",
       " '_insert_font',\n",
       " '_loadOutline',\n",
       " '_make_page_map',\n",
       " '_move_copy_page',\n",
       " '_name',\n",
       " '_newPage',\n",
       " '_outline',\n",
       " '_page_refs',\n",
       " '_pages_ret',\n",
       " '_remove_links_to',\n",
       " '_remove_toc_item',\n",
       " '_reset_page_refs',\n",
       " '_set_page_labels',\n",
       " '_update_toc_item',\n",
       " 'add_layer',\n",
       " 'add_ocg',\n",
       " 'authenticate',\n",
       " 'bake',\n",
       " 'can_save_incrementally',\n",
       " 'chapter_count',\n",
       " 'chapter_page_count',\n",
       " 'close',\n",
       " 'convert_to_pdf',\n",
       " 'copy_page',\n",
       " 'del_toc_item',\n",
       " 'del_xml_metadata',\n",
       " 'delete_page',\n",
       " 'delete_pages',\n",
       " 'embfile_add',\n",
       " 'embfile_count',\n",
       " 'embfile_del',\n",
       " 'embfile_get',\n",
       " 'embfile_info',\n",
       " 'embfile_names',\n",
       " 'embfile_upd',\n",
       " 'extract_font',\n",
       " 'extract_image',\n",
       " 'ez_save',\n",
       " 'find_bookmark',\n",
       " 'fullcopy_page',\n",
       " 'get_char_widths',\n",
       " 'get_layer',\n",
       " 'get_layers',\n",
       " 'get_new_xref',\n",
       " 'get_oc',\n",
       " 'get_ocgs',\n",
       " 'get_ocmd',\n",
       " 'get_outline_xrefs',\n",
       " 'get_page_fonts',\n",
       " 'get_page_images',\n",
       " 'get_page_labels',\n",
       " 'get_page_numbers',\n",
       " 'get_page_pixmap',\n",
       " 'get_page_text',\n",
       " 'get_page_xobjects',\n",
       " 'get_sigflags',\n",
       " 'get_toc',\n",
       " 'get_xml_metadata',\n",
       " 'has_annots',\n",
       " 'has_links',\n",
       " 'init_doc',\n",
       " 'insert_file',\n",
       " 'insert_page',\n",
       " 'insert_pdf',\n",
       " 'is_closed',\n",
       " 'is_dirty',\n",
       " 'is_encrypted',\n",
       " 'is_fast_webaccess',\n",
       " 'is_form_pdf',\n",
       " 'is_pdf',\n",
       " 'is_reflowable',\n",
       " 'is_repaired',\n",
       " 'is_stream',\n",
       " 'journal_can_do',\n",
       " 'journal_enable',\n",
       " 'journal_is_enabled',\n",
       " 'journal_load',\n",
       " 'journal_op_name',\n",
       " 'journal_position',\n",
       " 'journal_redo',\n",
       " 'journal_save',\n",
       " 'journal_start_op',\n",
       " 'journal_stop_op',\n",
       " 'journal_undo',\n",
       " 'language',\n",
       " 'last_location',\n",
       " 'layer_ui_configs',\n",
       " 'layout',\n",
       " 'load_page',\n",
       " 'location_from_page_number',\n",
       " 'make_bookmark',\n",
       " 'markinfo',\n",
       " 'metadata',\n",
       " 'move_page',\n",
       " 'name',\n",
       " 'need_appearances',\n",
       " 'needs_pass',\n",
       " 'new_page',\n",
       " 'next_location',\n",
       " 'outline',\n",
       " 'page_annot_xrefs',\n",
       " 'page_count',\n",
       " 'page_count2',\n",
       " 'page_cropbox',\n",
       " 'page_number_from_location',\n",
       " 'page_xref',\n",
       " 'pagelayout',\n",
       " 'pagemode',\n",
       " 'pages',\n",
       " 'pdf_catalog',\n",
       " 'pdf_trailer',\n",
       " 'permissions',\n",
       " 'prev_location',\n",
       " 'recolor',\n",
       " 'reload_page',\n",
       " 'resolve_link',\n",
       " 'resolve_names',\n",
       " 'rewrite_images',\n",
       " 'save',\n",
       " 'saveIncr',\n",
       " 'save_snapshot',\n",
       " 'scrub',\n",
       " 'search_page_for',\n",
       " 'select',\n",
       " 'set_language',\n",
       " 'set_layer',\n",
       " 'set_layer_ui_config',\n",
       " 'set_markinfo',\n",
       " 'set_metadata',\n",
       " 'set_oc',\n",
       " 'set_ocmd',\n",
       " 'set_page_labels',\n",
       " 'set_pagelayout',\n",
       " 'set_pagemode',\n",
       " 'set_toc',\n",
       " 'set_toc_item',\n",
       " 'set_xml_metadata',\n",
       " 'stream',\n",
       " 'subset_fonts',\n",
       " 'switch_layer',\n",
       " 'this',\n",
       " 'this_is_pdf',\n",
       " 'thisown',\n",
       " 'tobytes',\n",
       " 'update_object',\n",
       " 'update_stream',\n",
       " 'version_count',\n",
       " 'write',\n",
       " 'xref',\n",
       " 'xref_copy',\n",
       " 'xref_get_key',\n",
       " 'xref_get_keys',\n",
       " 'xref_is_font',\n",
       " 'xref_is_image',\n",
       " 'xref_is_stream',\n",
       " 'xref_is_xobject',\n",
       " 'xref_length',\n",
       " 'xref_object',\n",
       " 'xref_set_key',\n",
       " 'xref_stream',\n",
       " 'xref_stream_raw',\n",
       " 'xref_xml_metadata']"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(doc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2085d3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch_gpu",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
